<!doctype html><html lang=en class=no-js> <head><script type=application/ld+json>
{
  "@context": "https://schema.org",
  "@id": "https://hds-sandbox.github.io/scRNASeq_course/",
  "@type": "Course",
  "coursePrerequisites": [
    "Knowledge of R, Rstudio and Rmarkdown or Python and Jupyter Notebooks",
    "Basic knowledge of scRNAseq technology",
    "Basic knowledge of data science and statistics such as PCA, clustering and statistical testing"
  ],
  "dct:conformsTo": "https://bioschemas.org/profiles/Course/1.0-RELEASE",
  "description": "This workshop contains a basic tutorial on how to approach scRNAseq experiments, starting from fastq files out of the sequencer. Thus, the workshop does not include any information of laboratory protocols, library preparation or any wet-lab related procedures.",
  "isBasedOn": [
    "https://nf-co.re/scrnaseq",
    "https://satijalab.org/seurat/",
    "https://cran.r-project.org/web/packages/gprofiler2/vignettes/gprofiler2.html",
    "https://bioconductor.org/packages/release/bioc/vignettes/slingshot/inst/doc/vignette.html#constructing-smooth-curves-and-ordering-cells"
  ],
  "keywords": "single cell, RNAseq, python, R",
  "license": {
    "@id": "http://www.apache.org/licenses/",
    "@type": "CreativeWork"
  },
  "mentions": [
    {
      "@id": "#DOI of zenodo repo",
      "@type": "Thing"
    },
    {
      "@id": "#DOI of zenodo data and materials",
      "@type": "Thing"
    }
  ],
  "name": "single-cell RNAseq workshop"
}
</script><title>Normalization and dimensionality reduction - Single cell sequencing data analysis - an introductiory but comprehensive course</title><link rel=stylesheet href=../assets/stylesheets/main.975780f9.min.css><link rel=stylesheet href=../assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../stylesheets/extra.css><script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#normalization-and-dimensionality-reduction class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Single cell sequencing data analysis - an introductiory but comprehensive course" class="md-header__button md-logo" aria-label="Single cell sequencing data analysis - an introductiory but comprehensive course" data-md-component=logo> <img src=../img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Single cell sequencing data analysis - an introductiory but comprehensive course </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Normalization and dimensionality reduction </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/scRNASeq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/scRNASeq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=../python/installation.html class=md-tabs__link> Python </a> </li> <li class=md-tabs__item> <a href=Part01_read_the_data.html class="md-tabs__link md-tabs__link--active"> R </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Single cell sequencing data analysis - an introductiory but comprehensive course" class="md-nav__button md-logo" aria-label="Single cell sequencing data analysis - an introductiory but comprehensive course" data-md-component=logo> <img src=../img/logo.png alt=logo> </a> Single cell sequencing data analysis - an introductiory but comprehensive course </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/scRNASeq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/scRNASeq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../index.html class=md-nav__link> Course Introduction </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_2 type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2> Python <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Python data-md-level=1> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Python </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../python/installation.html class=md-nav__link> Setup and installation </a> </li> <li class=md-nav__item> <a href=../python/Part01_read_the_data.html class=md-nav__link> Read the data </a> </li> <li class=md-nav__item> <a href=../python/Part02_filtering_sample2.html class=md-nav__link> Quality Control (QC) and filtering </a> </li> <li class=md-nav__item> <a href=../python/Part03_normalize_and_integrate.html class=md-nav__link> Normalize and Integrate </a> </li> <li class=md-nav__item> <a href=../python/Part04_clustering.html class=md-nav__link> Cell clustering and Differential Expression (DE) </a> </li> <li class=md-nav__item> <a href=../python/Part05_cell_fate.html class=md-nav__link> Pseudotimes and cell fates </a> </li> <li class=md-nav__item> <a href=../python/Part06_Condition_analysis.html class=md-nav__link> Cross-data analysis </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_2_8 type=checkbox id=__nav_2_8> <label class=md-nav__link for=__nav_2_8> Extra <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Extra data-md-level=2> <label class=md-nav__title for=__nav_2_8> <span class="md-nav__icon md-icon"></span> Extra </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../python/Part02_filtering_sample3.html class=md-nav__link> Filtering another sample </a> </li> <li class=md-nav__item> <a href=../python/Part02_filtering_sample1.html class=md-nav__link> Filtering a low quality sample </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_3 type=checkbox id=__nav_3 checked> <label class=md-nav__link for=__nav_3> R <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=R data-md-level=1> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> R </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=Part01_read_the_data.html class=md-nav__link> Read the data </a> </li> <li class=md-nav__item> <a href=Part02_filtering_sample2.html class=md-nav__link> Preprocessing </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Normalization <span class="md-nav__icon md-icon"></span> </label> <a href=Part03_normalize_and_dim_reduc.html class="md-nav__link md-nav__link--active"> Normalization </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#setup class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=#standard-approach-for-normalization-and-scaling class=md-nav__link> Standard approach for normalization and scaling </a> <nav class=md-nav aria-label="Standard approach for normalization and scaling"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#normalizing-the-data class=md-nav__link> Normalizing the data </a> </li> <li class=md-nav__item> <a href=#identification-of-highly-variable-features-feature-selection class=md-nav__link> Identification of highly variable features (feature selection) </a> </li> <li class=md-nav__item> <a href=#scaling-the-data class=md-nav__link> Scaling the data </a> </li> <li class=md-nav__item> <a href=#perform-linear-dimensional-reduction class=md-nav__link> Perform linear dimensional reduction </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#determine-the-dimensionality-of-the-dataset class=md-nav__link> Determine the ‘dimensionality’ of the dataset </a> </li> <li class=md-nav__item> <a href=#run-non-linear-dimensional-reduction-umaptsne class=md-nav__link> Run non-linear dimensional reduction (UMAP/tSNE) </a> </li> <li class=md-nav__item> <a href=#recommended-approach-using-sctransform class=md-nav__link> Recommended approach using scTransform </a> </li> <li class=md-nav__item> <a href=#apply-sctransform-normalization class=md-nav__link> Apply sctransform normalization </a> </li> <li class=md-nav__item> <a href=#wrapping-up class=md-nav__link> Wrapping up </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=Part04_integration.html class=md-nav__link> Integration </a> </li> <li class=md-nav__item> <a href=Part05_clustering.html class=md-nav__link> Clustering and Differential Expression </a> </li> <li class=md-nav__item> <a href=Part06_pseudotime.html class=md-nav__link> Pseudotimes </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_3_7 type=checkbox id=__nav_3_7> <label class=md-nav__link for=__nav_3_7> Extra <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Extra data-md-level=2> <label class=md-nav__title for=__nav_3_7> <span class="md-nav__icon md-icon"></span> Extra </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=Part02_filtering_sample3.html class=md-nav__link> Preprocessing another sample </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#setup class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=#standard-approach-for-normalization-and-scaling class=md-nav__link> Standard approach for normalization and scaling </a> <nav class=md-nav aria-label="Standard approach for normalization and scaling"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#normalizing-the-data class=md-nav__link> Normalizing the data </a> </li> <li class=md-nav__item> <a href=#identification-of-highly-variable-features-feature-selection class=md-nav__link> Identification of highly variable features (feature selection) </a> </li> <li class=md-nav__item> <a href=#scaling-the-data class=md-nav__link> Scaling the data </a> </li> <li class=md-nav__item> <a href=#perform-linear-dimensional-reduction class=md-nav__link> Perform linear dimensional reduction </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#determine-the-dimensionality-of-the-dataset class=md-nav__link> Determine the ‘dimensionality’ of the dataset </a> </li> <li class=md-nav__item> <a href=#run-non-linear-dimensional-reduction-umaptsne class=md-nav__link> Run non-linear dimensional reduction (UMAP/tSNE) </a> </li> <li class=md-nav__item> <a href=#recommended-approach-using-sctransform class=md-nav__link> Recommended approach using scTransform </a> </li> <li class=md-nav__item> <a href=#apply-sctransform-normalization class=md-nav__link> Apply sctransform normalization </a> </li> <li class=md-nav__item> <a href=#wrapping-up class=md-nav__link> Wrapping up </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/hds-sandbox/scRNASeq_course/edit/master/docs/R/Part03_normalize_and_dim_reduc.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1 id=normalization-and-dimensionality-reduction>Normalization and dimensionality reduction<a class=headerlink href=#normalization-and-dimensionality-reduction title="Permanent link">&para;</a></h1> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>&#128368; <strong>Time Estimation:</strong> 40 minutes if using at least 2 cores (that is, at least 16 virtual-CPUs). Expect longer time with slower hardware, and a RAM usage of some 40-100GB.</p> <p>&#128172; <strong>Learning Objectives:</strong> </p> <ol> <li>Normalization of datasets</li> <li>Perform dimensionality reduction</li> <li>Select an adequate number of principal components</li> <li>Perform UMAP and visualization</li> <li>Understand and apply scTransformation of the data.</li> </ol> </div> <p>Biologically similar cells are not necessarily directly comparable in a dataset because of different technical biases, amongst many the different percentage of captured transcripts (capture efficiency), the presence of technical replicates, the presence of noisy transcripts. The capture efficiency can be influenced by many factors, i.e. the different transcript tags leading to different capture efficiency, the type of protocol used in the laboratory, the amount of PCR performed on different transcripts.</p> <p>To avoid these differences, a normalization approach is needed. Normalization is one of the main topics of scRNAseq data preprocessing, and many advanced techniques takes into account the statistical distribution of counts and the presence of technical/biological features of interest.</p> <p>The most standard approach is the TMP (Transcript Per Million) normalization followed by logarithmization and standardization. We have applied this technique to have a double-check on our data filtering in our previous notebook.</p> <p>As a rule of thumb, TPM+log+standardization is no longer consider a very good normalization technique, especially when you are integrating multiple datasets together. Instead, it is suggested to use more advanced methods for considering technical and biological covariates as part of a statistical model for the transcripts. One of the current state-of-the-art method is scTransform. We will apply it in this notebook.</p> <p>The difference between the two methods will be easily observed when performing dimensionality reduction methods and visualizing the dataset. It is therefore very important to select an adequate number of Principal Components for other more advanced techniques such as dataset integration and clustering.</p> <h2 id=setup>Setup<a class=headerlink href=#setup title="Permanent link">&para;</a></h2> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=nf>library</span><span class=p>(</span><span class=n>tidyverse</span><span class=p>)</span>
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a><span class=nf>library</span><span class=p>(</span><span class=n>patchwork</span><span class=p>)</span>
<a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a><span class=nf>library</span><span class=p>(</span><span class=n>Seurat</span><span class=p>)</span>
<a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a><span class=nf>library</span><span class=p>(</span><span class=n>SeuratDisk</span><span class=p>)</span>
<a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a><span class=nf>library</span><span class=p>(</span><span class=n>sctransform</span><span class=p>)</span>
<a id=__codelineno-0-6 name=__codelineno-0-6 href=#__codelineno-0-6></a>
<a id=__codelineno-0-7 name=__codelineno-0-7 href=#__codelineno-0-7></a><span class=nf>LoadH5Seurat</span><span class=p>(</span><span class=s>&quot;../../Data/notebooks_data/sample_2.filt.h5Seurat&quot;</span><span class=p>)</span>
</code></pre></div> <h2 id=standard-approach-for-normalization-and-scaling>Standard approach for normalization and scaling<a class=headerlink href=#standard-approach-for-normalization-and-scaling title="Permanent link">&para;</a></h2> <h3 id=normalizing-the-data>Normalizing the data<a class=headerlink href=#normalizing-the-data title="Permanent link">&para;</a></h3> <p>After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method "LogNormalize" that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. Normalized values are stored in <code>sample_2[["RNA"]]@data</code>.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>NormalizeData</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>normalization.method</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;LogNormalize&quot;</span><span class=p>,</span><span class=w> </span><span class=n>scale.factor</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1e4</span><span class=p>)</span>
</code></pre></div> <p>For clarity, in this previous line of code (and in future commands), we provide the default values for certain parameters in the function call. However, this isn’t required and the same behavior can be achieved with:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>NormalizeData</span><span class=p>(</span><span class=n>sample_2</span><span class=p>)</span>
</code></pre></div> <h3 id=identification-of-highly-variable-features-feature-selection>Identification of highly variable features (feature selection)<a class=headerlink href=#identification-of-highly-variable-features-feature-selection title="Permanent link">&para;</a></h3> <p>We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). Seurat authors and others <a href=https://www.nature.com/articles/nmeth.2645>others</a> have found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.</p> <p>The procedure in Seurat is described in detail <a href=https://doi.org/10.1016/j.cell.2019.05.031>here</a>, and improves on previous versions by directly modeling the mean-variance relationship inherent in single-cell data, and is implemented in the <code>FindVariableFeatures()</code> function. By default, it returns 2,000 features per dataset. These will be used in downstream analysis, like PCA.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindVariableFeatures</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>selection.method</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&#39;vst&#39;</span><span class=p>,</span><span class=w> </span><span class=n>nfeatures</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>2000</span><span class=p>)</span>
</code></pre></div> <p>We can access a dataframe with our variable features using the function <code>VariableFeatures()</code>:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=c1># Identify the 10 most highly variable genes</span>
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a><span class=n>top10</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>head</span><span class=p>(</span><span class=nf>VariableFeatures</span><span class=p>(</span><span class=n>sample_2</span><span class=p>),</span><span class=w> </span><span class=m>10</span><span class=p>)</span>
</code></pre></div> <p>We can plot variable features with the function <code>VariableFeaturePlot()</code>, and add labels using <code>LabelPoints()</code> function.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=n>plot1</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>VariableFeaturePlot</span><span class=p>(</span><span class=n>sample_2</span><span class=p>)</span><span class=w> </span><span class=c1># Plot without labels</span>
<a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a><span class=n>plot2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>LabelPoints</span><span class=p>(</span><span class=n>plot</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>plot1</span><span class=p>,</span><span class=w> </span><span class=n>points</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>top10</span><span class=p>,</span><span class=w> </span><span class=n>repel</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span><span class=w> </span><span class=c1># Plot with labels. repel = TRUE will jitter the dots</span>
<a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a><span class=n>plot1</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=n>plot2</span>
</code></pre></div> <h3 id=scaling-the-data>Scaling the data<a class=headerlink href=#scaling-the-data title="Permanent link">&para;</a></h3> <p>Next, we apply a linear transformation (‘scaling’) that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The <code>ScaleData()</code> function:</p> <ul> <li>Shifts the expression of each gene, so that the mean expression across cells is 0</li> <li>Scales the expression of each gene, so that the variance across cells is 1</li> <li>This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate</li> <li>The results of this are stored in <code>sample_2[["RNA"]]@scale.data</code></li> </ul> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=n>all.genes</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>rownames</span><span class=p>(</span><span class=n>sample_2</span><span class=p>)</span>
<a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>ScaleData</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>all.genes</span><span class=p>)</span>
</code></pre></div> <details class=tip> <summary>This step takes too long! Can I make it faster?</summary> <p>Scaling is an essential step in the Seurat workflow, but only on genes that will be used as input to PCA. Therefore, the default in <code>ScaleData()</code> is only to perform scaling on the previously identified variable features (2,000 by default). To do this, omit the <code>features</code> argument in the previous function call, i.e.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>ScaleData</span><span class=p>(</span><span class=n>sample_2</span><span class=p>)</span>
</code></pre></div> <p>Your PCA and clustering results will be unaffected. However, Seurat heatmaps (produced as shown below with <code>DoHeatmap()</code>) require genes in the heatmap to be scaled, to make sure highly-expressed genes don't dominate the heatmap. To make sure we don't leave any genes out of the heatmap later, we are scaling all genes in this tutorial. </p> </details> <div class="admonition info"> <p class=admonition-title>Remove unwanted sources of variation**</p> <p>In Seurat we can also use the <code>ScaleData()</code> function to remove unwanted sources of variation from a single-cell dataset. For example, we could 'regress out' heterogeneity associated with (for example) cell cycle stage, or mitochondrial contamination:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>ScaleData</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>vars.to.regress</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&#39;percent.mt&#39;</span><span class=p>)</span>
</code></pre></div> </div> <h3 id=perform-linear-dimensional-reduction>Perform linear dimensional reduction<a class=headerlink href=#perform-linear-dimensional-reduction title="Permanent link">&para;</a></h3> <p>Next we perform PCA on the scaled data. By default, only the previously determined variable features are used as input, but can be defined using <code>features</code> argument if you wish to choose a different subset.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>RunPCA</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>VariableFeatures</span><span class=p>(</span><span class=n>object</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>))</span>
</code></pre></div> <p>Seurat provides several useful ways of visualizing both cells and features that define the PCA, including <code>VizDimReduction()</code>, <code>DimPlot()</code>, and <code>DimHeatmap()</code></p> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=c1># Examine and visualize PCA results a few different ways</span>
<a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a><span class=nf>print</span><span class=p>(</span><span class=n>sample_2</span><span class=p>[[</span><span class=s>&#39;pca&#39;</span><span class=p>]],</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>5</span><span class=p>,</span><span class=w> </span><span class=n>nfeatures</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>5</span><span class=p>)</span>
<a id=__codelineno-10-3 name=__codelineno-10-3 href=#__codelineno-10-3></a><span class=nf>VizDimLoadings</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>2</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&#39;pca&#39;</span><span class=p>)</span>
<a id=__codelineno-10-4 name=__codelineno-10-4 href=#__codelineno-10-4></a><span class=nf>DimPlot</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&#39;pca&#39;</span><span class=p>)</span>
</code></pre></div> <p>In particular <code>DimHeatmap()</code> allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses. Both cells and features are ordered according to their PCA scores. Setting <code>cells</code> to a number plots the ‘extreme’ cells on both ends of the spectrum, which dramatically speeds plotting for large datasets. Though clearly a supervised analysis, we find this to be a valuable tool for exploring correlated feature sets.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=nf>DimHeatmap</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=p>,</span><span class=w> </span><span class=n>cells</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>500</span><span class=p>,</span>
<a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a><span class=w>           </span><span class=n>nfeatures</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>30</span><span class=p>,</span><span class=w> </span><span class=c1># number of genes to plot</span>
<a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a><span class=w>           </span><span class=n>balanced</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span><span class=w> </span><span class=c1># balanced = TRUE will show equal number of positive and negative loadings</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=nf>DimHeatmap</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>15</span><span class=p>,</span><span class=w> </span><span class=n>cells</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>500</span><span class=p>,</span><span class=w> </span><span class=n>nfeatures</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>30</span><span class=p>,</span><span class=w> </span><span class=n>balanced</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span>
</code></pre></div> <h2 id=determine-the-dimensionality-of-the-dataset>Determine the ‘dimensionality’ of the dataset<a class=headerlink href=#determine-the-dimensionality-of-the-dataset title="Permanent link">&para;</a></h2> <p>To overcome the extensive technical noise in any single feature for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a ‘metafeature’ that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset. However, how many components should we choose to include? 10? 20? 100?</p> <p>In <a href=http://www.cell.com/abstract/S0092-8674(15)00549-8>Macosko <em>et al</em></a>, Seurat implemented a resampling test inspired by the JackStraw procedure. It randomly permutes a subset of the data (1% by default) and rerun PCA, constructing a ‘null distribution’ of feature scores, and repeat this procedure. Thus, it identifies ‘significant’ PCs as those who have a strong enrichment of low p-value features.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=c1># NOTE: This process can take a long time for big datasets, comment out for expediency. More approximate techniques such as those implemented in ElbowPlot() can be used to reduce computation time</span>
<a id=__codelineno-13-2 name=__codelineno-13-2 href=#__codelineno-13-2></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>JackStraw</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>num.replicate</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>100</span><span class=p>)</span>
<a id=__codelineno-13-3 name=__codelineno-13-3 href=#__codelineno-13-3></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>ScoreJackStraw</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>20</span><span class=p>)</span>
</code></pre></div> <p>The <code>JackStrawPlot()</code> function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). ‘Significant’ PCs will show a strong enrichment of features with low p-values (solid curve above the dashed line). In this case it appears that there is a sharp drop-off in significance after the first 10-12 PCs.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a><span class=nf>JackStrawPlot</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>15</span><span class=p>)</span>
</code></pre></div> <p>An alternative heuristic method generates an ‘Elbow plot’: a ranking of principle components based on the percentage of variance explained by each one (<code>ElbowPlot()</code> function). In this example, we can observe an ‘elbow’ around PC9-10, suggesting that the majority of true signal is captured in the first 10 PCs.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=nf>ElbowPlot</span><span class=p>(</span><span class=n>sample_2</span><span class=p>)</span>
</code></pre></div> <p>Identifying the true dimensionality of a dataset – can be challenging/uncertain for the user. It is therefore suggested to consider three approaches. The first is more supervised, exploring PCs to determine relevant sources of heterogeneity, and could be used in conjunction with GSEA for example. The second implements a statistical test based on a random null model, but is time-consuming for large datasets, and may not return a clear PC cutoff. The third is a heuristic that is commonly used, and can be calculated instantly. In this example, all three approaches yielded similar results, but we might have been justified in choosing anything between PC 7-12 as a cutoff.</p> <h2 id=run-non-linear-dimensional-reduction-umaptsne>Run non-linear dimensional reduction (UMAP/tSNE)<a class=headerlink href=#run-non-linear-dimensional-reduction-umaptsne title="Permanent link">&para;</a></h2> <p>Seurat offers several non-linear dimensional reduction techniques, such as tSNE and UMAP, to visualize and explore these datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP and tSNE, we suggest using the same PCs we selected in the step before.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-16-1 name=__codelineno-16-1 href=#__codelineno-16-1></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>RunUMAP</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>10</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-17-1 name=__codelineno-17-1 href=#__codelineno-17-1></a><span class=nf>DimPlot</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&#39;umap&#39;</span><span class=p>)</span>
</code></pre></div> <p>You can save the object at this point so that it can easily be loaded back in without having to rerun the computationally intensive steps performed above, or easily shared with collaborators.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-18-1 name=__codelineno-18-1 href=#__codelineno-18-1></a><span class=nf>SaveH5Seurat</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>filename</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;../../Data/notebooks_data/sample_2.filt.norm.h5Seurat&quot;</span><span class=p>)</span>
</code></pre></div> <h2 id=recommended-approach-using-sctransform>Recommended approach using scTransform<a class=headerlink href=#recommended-approach-using-sctransform title="Permanent link">&para;</a></h2> <p>Biological heterogeneity in single-cell RNA-seq data is often confounded by technical factors, including sequencing depth. The number of molecules detected in each cell can vary significantly between cells, even within the same celltype. Interpretation of scRNA-seq data requires effective pre-processing and normalization to remove this technical variability. In <a href=https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1>Hafemeister and Satija, 2019</a> Seurat introduces a modeling framework for the normalization and variance stabilization of molecular count data from scRNA-seq experiment. This procedure omits the need for heuristic steps including pseudocount addition or log-transformation and improves common downstream analytical tasks such as variable gene selection, dimensional reduction, and differential expression.</p> <p>By using <a href=https://github.com/ChristophH/sctransform/ >sctransform</a> based normalization, we enable recovering sharper biological distinction compared to log-normalization.</p> <p>We will reload first the filtered data</p> <div class=highlight><pre><span></span><code><a id=__codelineno-19-1 name=__codelineno-19-1 href=#__codelineno-19-1></a><span class=nf>LoadH5Seurat</span><span class=p>(</span><span class=s>&quot;../../Data/notebooks_data/sample_2.filt.h5Seurat&quot;</span><span class=p>)</span>
</code></pre></div> <h2 id=apply-sctransform-normalization>Apply sctransform normalization<a class=headerlink href=#apply-sctransform-normalization title="Permanent link">&para;</a></h2> <ul> <li>Note that this single command replaces <code>NormalizeData()</code>, <code>ScaleData()</code>, and <code>FindVariableFeatures()</code>.</li> <li>Transformed data will be available in the SCT assay, which is set as the default after running sctransform</li> <li>During normalization, we can also remove confounding sources of variation, for example, mitochondrial mapping percentage</li> </ul> <div class=highlight><pre><span></span><code><a id=__codelineno-20-1 name=__codelineno-20-1 href=#__codelineno-20-1></a><span class=c1># run sctransform</span>
<a id=__codelineno-20-2 name=__codelineno-20-2 href=#__codelineno-20-2></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>SCTransform</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>vars.to.regress</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;percent.mt&quot;</span><span class=p>,</span><span class=w> </span><span class=n>verbose</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
</code></pre></div> <p>The latest version of <code>sctransform</code> also supports using <a href=https://bioconductor.org/packages/release/bioc/html/glmGamPoi.html>glmGamPoi</a> package which substantially improves the speed of the learning procedure. It can be invoked by specifying <code>method="glmGamPoi"</code>.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-21-1 name=__codelineno-21-1 href=#__codelineno-21-1></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>SCTransform</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>method</span><span class=o>=</span><span class=s>&quot;glmGamPoi&quot;</span><span class=p>,</span><span class=w> </span><span class=n>vars.to.regress</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;percent.mt&quot;</span><span class=p>,</span><span class=w> </span><span class=n>verbose</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
</code></pre></div> <p>Perform dimensionality reduction by PCA and UMAP embedding</p> <div class=highlight><pre><span></span><code><a id=__codelineno-22-1 name=__codelineno-22-1 href=#__codelineno-22-1></a><span class=c1># These are now standard steps in the Seurat workflow for visualization and clustering</span>
<a id=__codelineno-22-2 name=__codelineno-22-2 href=#__codelineno-22-2></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>RunPCA</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>verbose</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
<a id=__codelineno-22-3 name=__codelineno-22-3 href=#__codelineno-22-3></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>RunUMAP</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>30</span><span class=p>,</span><span class=w> </span><span class=n>verbose</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
<a id=__codelineno-22-4 name=__codelineno-22-4 href=#__codelineno-22-4></a>
<a id=__codelineno-22-5 name=__codelineno-22-5 href=#__codelineno-22-5></a><span class=nf>DimPlot</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>label</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>NoLegend</span><span class=p>()</span>
</code></pre></div> <p>Finally, we shoud can save the object at this point so that it can easily be loaded back in without having to rerun the computationally intensive steps performed above, or easily shared with collaborators.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-23-1 name=__codelineno-23-1 href=#__codelineno-23-1></a><span class=nf>SaveH5Seurat</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>filename</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;../../Data/notebooks_data/sample_2.filt.norm.h5Seurat&quot;</span><span class=p>)</span>
</code></pre></div> <details class=info> <summary>Why can we choose more PCs when using sctransform?</summary> <p>In the standard Seurat workflow], we focused on 10 PCs for this dataset, though we highlight that the results are similar with higher settings for this parameter. Interestingly, we've found that when using sctransform, we often benefit by pushing this parameter even higher. We believe this is because the sctransform workflow performs more effective normalization, strongly removing technical effects from the data. </p> <p>Even after standard log-normalization, variation in sequencing depth is still a confounding factor (see <a href=https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1>Figure 1</a>), and this effect can subtly influence higher PCs. In sctransform, this effect is substantially mitigated (see <a href=https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1>Figure 3</a>). This means that higher PCs are more likely to represent subtle, but biologically relevant, sources of heterogeneity -- so including them may improve downstream analysis.</p> <p>In addition, sctransform returns 3,000 variable features by default, instead of 2,000. The rationale is similar, the additional variable features are less likely to be driven by technical differences across cells, and instead may represent more subtle biological fluctuations. In general, we find that results produced with sctransform are less dependent on these parameters (indeed, we achieve nearly identical results when using all genes in the transcriptome, though this does reduce computational efficiency). This can help users generate more robust results, and in addition, enables the application of standard analysis pipelines with identical parameter settings that can quickly be applied to new datasets:</p> <p>For example, the following code replicates the full end-to-end workflow, in a single command:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-24-1 name=__codelineno-24-1 href=#__codelineno-24-1></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>CreateSeuratObject</span><span class=p>(</span><span class=n>sample_2</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>SCTransform</span><span class=p>(</span><span class=n>vars.to.regress</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&#39;percent.mt&#39;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-24-2 name=__codelineno-24-2 href=#__codelineno-24-2></a><span class=w>  </span><span class=nf>RunPCA</span><span class=p>()</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>RunUMAP</span><span class=p>(</span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>30</span><span class=p>)</span>
</code></pre></div> </details> <details class=info> <summary>Where are normalized values stored for sctransform?</summary> <p>sctransform calculates a model of technical noise in scRNA-seq data using 'regularized negative binomial regression'. The residuals for this model are normalized values, and can be positive or negative. Positive residuals for a given gene in a given cell indicate that we observed more UMIs than expected given the gene’s average expression in the population and cellular sequencing depth, while negative residuals indicate the converse. </p> <p>The results of sctransfrom are stored in the "SCT" assay: You can learn more about multi-assay data and commands in Seurat the <a href=https://github.com/satijalab/seurat/wiki/Assay>developer guide</a>.</p> <ul> <li><code>sample_2[["SCT"]]@scale.data</code> contains the residuals (normalized values), and is used directly as input to PCA. Please note that this matrix is non-sparse, and can therefore take up a lot of memory if stored for all genes. To save memory, it stores these values only for variable genes, by setting the return.only.var.genes = TRUE by default in the <code>SCTransform()</code> function call.</li> <li>To assist with visualization and interpretation, it also converts Pearson residuals back to ‘corrected’ UMI counts. You can interpret these as the UMI counts we would expect to observe if all cells were sequenced to the same depth. If you want to see exactly how it is done, please look at the correct function <a href=https://github.com/ChristophH/sctransform/blob/master/R/denoise.R>here</a>.</li> <li>The 'corrected' UMI counts are stored in <code>sample_2[["SCT"]]@counts</code>. It stores log-normalized versions of these corrected counts in <code>sample_2[["SCT"]]@data</code>, which are very helpful for visualization.</li> <li>You can use the corrected log-normalized counts for differential expression and integration.</li> </ul> </details> <h2 id=wrapping-up>Wrapping up<a class=headerlink href=#wrapping-up title="Permanent link">&para;</a></h2> <p>In this notebook we learned two ways of normalizing scRNAseq data using the <code>scale.data()</code> and the <code>SCTransform()</code> functions. We have also learned how to determine the dimensionality of the dataset and how to select a proper number of PCs. Finally, we have introduced the concept of UMAPs to visualize scRNAseq datasets.</p> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=Part02_filtering_sample2.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Preprocessing" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Preprocessing </div> </div> </a> <a href=Part04_integration.html class="md-footer__link md-footer__link--next" aria-label="Next: Integration" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Integration </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox/ target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "..", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.footer", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "../assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script> <script src=../assets/javascripts/bundle.5a2dcb6a.min.js></script> <script src=../javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> <script src=../javascripts/embed-amd.js></script> </body> </html>