<!doctype html><html lang=en class=no-js> <head><script type=application/ld+json>
    {
      "@context": "https://schema.org",
      "@id": "https://hds-sandbox.github.io/scRNASeq_course/",
      "@type": "LearningResource",
      "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
      "description": "This workshop contains a basic tutorial on how to approach scRNAseq experiments, starting from fastq files out of the sequencer. Thus, the workshop does not include any information of laboratory protocols, library preparation or any wet-lab related procedures.",
      "keywords": [
        "bioinformatics, RNAseq, single cell, NGS, python, R"
      ],
      "learningResourceType": [
        "e-Learning"
      ],
      "license": [
        {
          "@id": "https://creativecommons.org/licenses/by/4.0/",
          "@type": "CreativeWork"
        }
      ],
      "mentions": [
        {
          "@id": "10.5281/zenodo.7660225",
          "@type": "Thing"
        },
        {
          "@id": "10.5281/zenodo.7565997",
          "@type": "Thing"
        }
      ],
      "name": "Introduction to single cell RNAseq analysis workshop"
    }
    </script><title>Quality Control (QC) and filtering - Single cell sequencing data analysis - an introductiory but comprehensive course</title><link rel=stylesheet href=../assets/stylesheets/main.975780f9.min.css><link rel=stylesheet href=../assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../stylesheets/extra.css><script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#qc-and-cell-filtering class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Single cell sequencing data analysis - an introductiory but comprehensive course" class="md-header__button md-logo" aria-label="Single cell sequencing data analysis - an introductiory but comprehensive course" data-md-component=logo> <img src=../img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Single cell sequencing data analysis - an introductiory but comprehensive course </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Quality Control (QC) and filtering </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/scRNASeq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/scRNASeq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=../python/installation.html class=md-tabs__link> Python </a> </li> <li class=md-tabs__item> <a href=Part01_read_the_data.html class="md-tabs__link md-tabs__link--active"> R </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Single cell sequencing data analysis - an introductiory but comprehensive course" class="md-nav__button md-logo" aria-label="Single cell sequencing data analysis - an introductiory but comprehensive course" data-md-component=logo> <img src=../img/logo.png alt=logo> </a> Single cell sequencing data analysis - an introductiory but comprehensive course </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/scRNASeq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/scRNASeq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../index.html class=md-nav__link> Course Introduction </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_2 type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2> Python <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Python data-md-level=1> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Python </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../python/installation.html class=md-nav__link> Setup and installation </a> </li> <li class=md-nav__item> <a href=../python/Part01_read_the_data.html class=md-nav__link> Read the data </a> </li> <li class=md-nav__item> <a href=../python/Part02_filtering_sample2.html class=md-nav__link> Quality Control (QC) and filtering </a> </li> <li class=md-nav__item> <a href=../python/Part03_normalize_and_integrate.html class=md-nav__link> Normalize and Integrate </a> </li> <li class=md-nav__item> <a href=../python/Part04_clustering.html class=md-nav__link> Cell clustering and Differential Expression (DE) </a> </li> <li class=md-nav__item> <a href=../python/Part05_cell_fate.html class=md-nav__link> Pseudotimes and cell fates </a> </li> <li class=md-nav__item> <a href=../python/Part06_Condition_analysis.html class=md-nav__link> Cross-data analysis </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_2_8 type=checkbox id=__nav_2_8> <label class=md-nav__link for=__nav_2_8> Extra <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Extra data-md-level=2> <label class=md-nav__title for=__nav_2_8> <span class="md-nav__icon md-icon"></span> Extra </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../python/Part02_filtering_sample3.html class=md-nav__link> Filtering another sample </a> </li> <li class=md-nav__item> <a href=../python/Part02_filtering_sample1.html class=md-nav__link> Filtering a low quality sample </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_3 type=checkbox id=__nav_3 checked> <label class=md-nav__link for=__nav_3> R <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=R data-md-level=1> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> R </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=Part01_read_the_data.html class=md-nav__link> Read the data </a> </li> <li class=md-nav__item> <a href=Part02_filtering_sample2.html class=md-nav__link> Preprocessing </a> </li> <li class=md-nav__item> <a href=Part03_normalize_and_dim_reduc.html class=md-nav__link> Normalization </a> </li> <li class=md-nav__item> <a href=Part04_integration.html class=md-nav__link> Integration </a> </li> <li class=md-nav__item> <a href=Part05_clustering.html class=md-nav__link> Clustering and Differential Expression </a> </li> <li class=md-nav__item> <a href=Part06_pseudotime.html class=md-nav__link> Pseudotimes </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_3_7 type=checkbox id=__nav_3_7 checked> <label class=md-nav__link for=__nav_3_7> Extra <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Extra data-md-level=2> <label class=md-nav__title for=__nav_3_7> <span class="md-nav__icon md-icon"></span> Extra </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Preprocessing another sample <span class="md-nav__icon md-icon"></span> </label> <a href=Part02_filtering_sample3.html class="md-nav__link md-nav__link--active"> Preprocessing another sample </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#qc-metrics class=md-nav__link> QC metrics </a> </li> <li class=md-nav__item> <a href=#visualize-and-evaluate-quality-measures class=md-nav__link> Visualize and evaluate quality measures </a> <nav class=md-nav aria-label="Visualize and evaluate quality measures"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#violin-plots class=md-nav__link> Violin plots </a> </li> <li class=md-nav__item> <a href=#counts-vs-genes class=md-nav__link> Counts vs Genes </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#filtering-cells-and-choosing-thresholds class=md-nav__link> Filtering cells and choosing thresholds </a> <nav class=md-nav aria-label="Filtering cells and choosing thresholds"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#doublet-filtering class=md-nav__link> Doublet filtering </a> </li> <li class=md-nav__item> <a href=#filtering-evaluation class=md-nav__link> Filtering evaluation </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#wrapping-up class=md-nav__link> Wrapping up </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#qc-metrics class=md-nav__link> QC metrics </a> </li> <li class=md-nav__item> <a href=#visualize-and-evaluate-quality-measures class=md-nav__link> Visualize and evaluate quality measures </a> <nav class=md-nav aria-label="Visualize and evaluate quality measures"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#violin-plots class=md-nav__link> Violin plots </a> </li> <li class=md-nav__item> <a href=#counts-vs-genes class=md-nav__link> Counts vs Genes </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#filtering-cells-and-choosing-thresholds class=md-nav__link> Filtering cells and choosing thresholds </a> <nav class=md-nav aria-label="Filtering cells and choosing thresholds"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#doublet-filtering class=md-nav__link> Doublet filtering </a> </li> <li class=md-nav__item> <a href=#filtering-evaluation class=md-nav__link> Filtering evaluation </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#wrapping-up class=md-nav__link> Wrapping up </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/hds-sandbox/scRNASeq_course/edit/master/docs/R/Part02_filtering_sample3.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1 id=qc-and-cell-filtering>QC and cell filtering<a class=headerlink href=#qc-and-cell-filtering title="Permanent link">&para;</a></h1> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>&#128368; <strong>Time Estimation:</strong> 40 minutes </p> <p>&#128172; <strong>Learning Objectives:</strong> </p> <ol> <li>Understand and discuss QC issues and measures from single cell data</li> <li>Explore QC graphs and set filtering tools and thresholds</li> <li>Analyze the results of QC filters and evaluate necessity for different filtering </li> </ol> </div> <p>Quality control and filtering is the most important steps of single cell data analysis. Allowing low quality cells into your analysis will compromise/mislead your conclusions by adding hundreds of meaningless data points to your workflow. The main sources of low quality cells are:</p> <ul> <li>broken cells for which some of their transcripts get lost</li> <li>cells isolated together with too much ambient RNA</li> <li>missing cell during isolation (e.g.Â empty droplet in microfluidic machines)</li> <li>multiple cells isolated together (multiplets, usually only two cells - doublets)</li> </ul> <h2 id=qc-metrics>QC metrics<a class=headerlink href=#qc-metrics title="Permanent link">&para;</a></h2> <p>Seurat allows you to easily explore QC metrics and filter cells based on any user-defined criteria. A few QC metrics <a href=https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4758103/ >commonly used</a> by the community include:</p> <ul> <li> <p>The number of unique genes detected in each cell.</p> </li> <li> <p>Low-quality cells or empty droplets will often have very few genes</p> </li> <li> <p>Cell doublets or multiplets may exhibit an aberrantly high gene count</p> </li> <li> <p>Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)</p> </li> <li> <p>The percentage of reads that map to the mitochondrial genome</p> </li> <li> <p>Low-quality / dying cells often exhibit extensive mitochondrial contamination</p> </li> </ul> <p>The number of unique genes and total molecules are automatically calculated during <code>CreateSeuratObject()</code>. However, we calculate mitochondrial QC metrics ourselves, since we need an identifier for mitochondrial genes, and there is not a standard one. In our case, we can use the <code>PercentageFeatureSet()</code> function, which calculates the percentage of counts originating from a set of features, using all genes starting with <code>MT-</code> as a set of mitochondrial genes</p> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=c1># The [[ operator can add columns to object metadata. This is a great place to stash QC stats</span>
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a><span class=n>sample_3</span><span class=p>[[</span><span class=s>&quot;percent.mt&quot;</span><span class=p>]]</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>PercentageFeatureSet</span><span class=p>(</span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>pattern</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;^MT-&quot;</span><span class=p>)</span>
</code></pre></div> <details class=info> <summary>Where are QC metrics stored in Seurat?</summary> <p>Cell QC metrics are stored in the @meta.data slot of a Seurat Object</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=nf>head</span><span class=p>(</span><span class=n>sample_3</span><span class=o>@</span><span class=n>meta.data</span><span class=p>,</span><span class=w> </span><span class=m>5</span><span class=p>)</span>
</code></pre></div> </details> <h2 id=visualize-and-evaluate-quality-measures>Visualize and evaluate quality measures<a class=headerlink href=#visualize-and-evaluate-quality-measures title="Permanent link">&para;</a></h2> <h3 id=violin-plots>Violin plots<a class=headerlink href=#violin-plots title="Permanent link">&para;</a></h3> <p>We can make use of violin plots to show the distribution of values for different metrics in our data. This is done using the <code>VlnPlot()</code> function. The function can take several metrics all together and produce a combined plot with our metrics of interest:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=nf>VlnPlot</span><span class=p>(</span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;nFeature_RNA&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;nCount_RNA&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;percent.mt&quot;</span><span class=p>),</span><span class=w> </span><span class=n>ncol</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>3</span><span class=p>)</span>
</code></pre></div> <p>Here we simply look at the distribution of transcripts per cells, detected genes per cell and mithocondrial %. Note how the distribution is bimodal. This usually denotes a cluster of low-quality cells and viable cells. Sometimes filtering out the data points on the bottom-most sides of those graphs removes a lot of cells from a dataset, but this is quite a normal thing, and there is no need to be worried about it. The top side of the distributions show a tail with few cells having a lot of transcripts and genes. It is also good to filter out some of those extreme values - for technical reasons, it will also help in having a better normalization of the data later on.</p> <p>In this dataset there are few cell with a high percentage of mitocondrial content. Those are precisely 245 if we set 0.1 (that is 10%) as a threshold. A value between 10% and 20% is the usual standard when filtering single cell datasets.</p> <p>We can do some plots to have a look at quality measures combined together.</p> <h3 id=counts-vs-genes>Counts vs Genes<a class=headerlink href=#counts-vs-genes title="Permanent link">&para;</a></h3> <p>This is a typical plot, where you look at the total transcripts per cells (x axis) and detected genes per cell (y axis). Usually, those two measures grow together. Points with a lot of transcripts and genes might be <strong>multiplets</strong> (multiple cells sequenced together as one), while very few transcripts and genes denote the presence of only ambient RNA or <strong>very low quality sequencing</strong> of a cell.</p> <p>In order to create this plot, we make use of the <code>FeatureScatter()</code> function:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=c1># FeatureScatter is typically used to visualize feature-feature relationships, but can be used for anything calculated by the object, i.e. columns in object metadata, PC scores etc.</span>
<a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a>
<a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a><span class=nf>FeatureScatter</span><span class=p>(</span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>feature1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;nCount_RNA&quot;</span><span class=p>,</span><span class=w> </span><span class=n>feature2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;nFeature_RNA&quot;</span><span class=p>)</span><span class=w> </span>
</code></pre></div> <p>??? How to color dots with a continuous variable?</p> <div class=highlight><pre><span></span><code>Unfortunately, Seurat does not support yet that functionality. We will need to use the function `FetchData()` (works with gene expression values as well) or select the @meta.data object. Then, we can create a custom plot with ggplot2.


```r
data &lt;- FetchData(sample_3, vars = c(&quot;nCount_RNA&quot;, &quot;nFeature_RNA&quot;, &quot;percent.mt&quot;))
data &lt;- sample_3@meta.data

data %&gt;% ggplot(aes(x = nCount_RNA, y = nFeature_RNA, color = percent.mt)) + geom_point() + scale_color_viridis_c()
```
</code></pre></div> <h2 id=filtering-cells-and-choosing-thresholds>Filtering cells and choosing thresholds<a class=headerlink href=#filtering-cells-and-choosing-thresholds title="Permanent link">&para;</a></h2> <p><strong>Mitocondrial content</strong>: In this dataset there are few cells with a high percentage of mitocondrial content. Those are precisely 245 if we set 10% as a threshold. A value between 10% and 20% is the usual standard when filtering single cell datasets.</p> <p>In addition, we will filter cells that are in both extremes of the distributions for <strong>number of detected genes</strong> and <strong>number of total transcripts per cell</strong>. This this case, it seems reasonable to exclude:</p> <ul> <li>Cells that have unique counts over 30000 or less than 5000</li> <li>Cells that have unique detected genes over 6000 or below 2000.</li> </ul> <p>Once we have chosen our thresholds, we can filter our dataset using the <code>subset()</code> function. The argument <code>subset</code> allows us to create a filtering parameter using our QC metrics.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=n>MIN_COUNTS</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>5000</span><span class=w>  </span><span class=c1>#minimum number of transcripts per cell</span>
<a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a><span class=n>MAX_COUNTS</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>30000</span><span class=w> </span><span class=c1>#maximum number of transcripts per cell</span>
<a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a><span class=n>MIN_GENES</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>2000</span><span class=w>   </span><span class=c1>#minimum number of genes per cell</span>
<a id=__codelineno-5-4 name=__codelineno-5-4 href=#__codelineno-5-4></a><span class=n>MAX_GENES</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>6000</span><span class=w>   </span><span class=c1>#maximum number of genes per cell</span>
<a id=__codelineno-5-5 name=__codelineno-5-5 href=#__codelineno-5-5></a><span class=n>MAX_MITO</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>10</span><span class=w>      </span><span class=c1>#mitocondrial percentage treshold)</span>
<a id=__codelineno-5-6 name=__codelineno-5-6 href=#__codelineno-5-6></a>
<a id=__codelineno-5-7 name=__codelineno-5-7 href=#__codelineno-5-7></a><span class=n>sample_3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>subset</span><span class=p>(</span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>subset</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>nFeature_RNA</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=n>MIN_GENES</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=n>nFeature_RNA</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=n>MAX_GENES</span><span class=w> </span><span class=o>&amp;</span>
<a id=__codelineno-5-8 name=__codelineno-5-8 href=#__codelineno-5-8></a><span class=w>                     </span><span class=n>nCounts_RNA</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=n>MIN_COUNTS</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=n>nCounts_RNA</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=n>MAX_COUNTS</span><span class=w> </span><span class=o>&amp;</span>
<a id=__codelineno-5-9 name=__codelineno-5-9 href=#__codelineno-5-9></a><span class=w>                     </span><span class=n>percent.mt</span><span class=w> </span><span class=o>&lt;</span><span class=w> </span><span class=m>10</span><span class=p>)</span>
</code></pre></div> <p>We can now recheck our QC plots:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=nf>VlnPlot</span><span class=p>(</span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;nFeature_RNA&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;nCount_RNA&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;percent.mt&quot;</span><span class=p>),</span><span class=w> </span><span class=n>ncol</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>3</span><span class=p>)</span>
<a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a><span class=nf>FeatureScatter</span><span class=p>(</span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>feature1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;nCount_RNA&quot;</span><span class=p>,</span><span class=w> </span><span class=n>feature2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;nFeature_RNA&quot;</span><span class=p>)</span><span class=w> </span>
</code></pre></div> <h3 id=doublet-filtering>Doublet filtering<a class=headerlink href=#doublet-filtering title="Permanent link">&para;</a></h3> <p>Another important step consists in filtering out multiplets. Those are in the almost totality of the cases doublets, because triplets and above multiplets are extremely rare. Read <a href=https://liorpachter.wordpress.com/2019/02/07/sub-poisson-loading-for-single-cell-rna-seq/ >this more technical blog post</a> for more explanations about this.</p> <p>The external tool <code>scrublet</code> simulates doublets by putting together the transcripts of random pairs of cells from the dataset. Then, it assigns a score to each cell in the data, based on the similarity with the simulated doublets. An <code>expected_doublet_rate</code> of 0.06 (6%) is quite a typical value for single cell data, but if you have a better estimate from laboratory work, microscope imaging or a specific protocol/sequencing machine, you can also tweak the value.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=n>sample_3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>scrublet_R</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>python_home</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;/opt/miniconda3/bin/python&quot;</span><span class=p>)</span>
</code></pre></div> <p>It seems that the doublet rate is likely to be lower than 6%, meaning that in this regard the data has been produced pretty well. We now plot the doublet scores assigned to each cell by the algorithm. We can see that most cells have a low score (the score is a value between 0 and 1). Datasets with many doublets show a more bimodal distribution, while here we just have a light tail beyond 0.1.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=nf>VlnPlot</span><span class=p>(</span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;doublet_score&quot;</span><span class=p>)</span>
</code></pre></div> <p>We can choose 0.1 as filtering treshold for the few detected doublets or alternatively use the automatic selection of doublets by the algorithm. We will choose the last option and use the automatically chosen doublets.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=n>sample_3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>subset</span><span class=p>(</span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>subset</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>doublet_score</span><span class=w> </span><span class=o>==</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
</code></pre></div> <h3 id=filtering-evaluation>Filtering evaluation<a class=headerlink href=#filtering-evaluation title="Permanent link">&para;</a></h3> <p>A quite basic but easy way to look at the results of our filtering is to normalize and plot the dataset on some projections. We will see these steps in the next lessons with more detail, but just to be able to follow up, we use a standard normalization technique:</p> <ul> <li><strong>TPM normalization</strong>: the transcripts of each cell are normalized, so that their total amounts to the same value in each cell. This should make cells more comparable independently of how many transcripts has been retained during cell isolation.</li> <li><strong>Logarithmization</strong>: the logarithm of the normalized transcripts is calculated. This reduce the variability of transcripts values and highlights variations due to biological factors.</li> <li><strong>Scaling</strong>: Each gene is scaled across all cells. This is useful, for example, for projecting the data onto a PCA.</li> </ul> <p>TPM normalization and its logarithm are done within a single step, using the <code>NormalizeData()</code> function, while scaling is done using the <code>ScaleData()</code> function:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=n>sample_3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>NormalizeData</span><span class=p>(</span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>normalization.method</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;LogNormalize&quot;</span><span class=p>)</span>
<a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a><span class=n>sample_3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>ScaleData</span><span class=p>(</span><span class=n>sample_3</span><span class=p>)</span>
</code></pre></div> <p>Then, in order to reduce computational times, we select the <em>most variable genes (MVG)</em> to perform PCA and visualization. MVG are calculated based on the LogNormalized reads. We will select the top 2000 MVGs</p> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=n>sample_3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindVariableFeatures</span><span class=p>(</span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>nfeatures</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>2000</span><span class=p>)</span>
</code></pre></div> <p>Now we perform PCA. We select our calculated variable genes using the <code>VariableFeatures()</code> function, and visualize the PCA projection with the <code>FeaturePlot()</code> function. We can color the dots using one of our QC metrics using the <code>features</code> parameter:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=n>sample_3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>RunPCA</span><span class=p>(</span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>VariableFeatures</span><span class=p>(</span><span class=n>object</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_3</span><span class=p>))</span>
<a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&#39;pca&#39;</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;nCount_RNA&quot;</span><span class=p>)</span>
</code></pre></div> <p>We can already see how the PCA has a clear structure with only a few dots sparsed around. It seems the filtering gave a good result.</p> <p>Next, we need to select how many Principal Components we should use for visualizing the data using UMAP. We plot the variance ratio to see how each component of the PCA changes in variability using the Elbow method. The explained variance of each component can be plot using the function <code>ElbowPlot()</code>:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=nf>ElbowPlot</span><span class=p>(</span><span class=n>sample_3</span><span class=p>)</span>
</code></pre></div> <p>We can choose a threshold to be used in all algorithms that use PCA to calculate any quantity. for example, we can select 15 PCA components, since the amount of variability explained dramatically drops after 10 PCs, more or less.</p> <p>Finally, We project the data using the UMAP algorithm. This is very good in preserving the structure of a dataset in low dimension, if any is present. We first calculate the neighbors of each cell (that is, its most similar cells), those are then used for the UMAP. The neighbors are calculated using the PCA matrix instead of the full data matrix, so we can choose the number of PCA components to use. Many algorithms work on the PCA, so you will see the parameter used again in other places.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a><span class=n>sample_3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>RunUMAP</span><span class=p>(</span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>10</span><span class=p>)</span>
<a id=__codelineno-14-2 name=__codelineno-14-2 href=#__codelineno-14-2></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;umap&quot;</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;nCount_RNA&quot;</span><span class=p>)</span>
</code></pre></div> <p>The UMAP plot gives a pretty well-structured output for this dataset. We will keep working further with this filtering.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=nf>SaveH5Seurat</span><span class=p>(</span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>filename</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;../../Data/notebooks_data/sample_3.filt.h5Seurat&quot;</span><span class=p>)</span>
</code></pre></div> <h2 id=wrapping-up>Wrapping up<a class=headerlink href=#wrapping-up title="Permanent link">&para;</a></h2> <p>We have succesfully gone through the filtering of a single cell dataset with good results that can be used further in the data analysis. In the next notebook <code>Normalize and Integrate</code>, we will integrate this dataset (testis cells from a healthy adult man) with the same type of sample from another man. Filtering of the other dataset is in the notebook <code>Part02_filtering_sample3.ipynb</code>. <strong>Run the notebook to generate the filtered dataset</strong>. The procedure follows tightly what happens for the dataset we just filtered.</p> <div class="admonition note"> <p class=admonition-title>Filtering a low quality sample</p> <p>As you could see, this dataset seemed pretty ok to handle. Another dataset of much lower quality is present, and is not going to be integrated in the coming data analysis. Its preprocessing is shown as the submenu <code>Filtering a low quality sample</code> in the section <code>Extra</code> of the course webpage. In it we will also show an aggressive filtering done using a combination of PCA technique and automatic outliers detection.</p> </div> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=Part06_pseudotime.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Pseudotimes" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Pseudotimes </div> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox/ target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "..", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.footer", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "../assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script> <script src=../assets/javascripts/bundle.5a2dcb6a.min.js></script> <script src=../javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> <script src=../javascripts/embed-amd.js></script> </body> </html>