<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="This course is an introduction for how to approach single cell RNAseq data, starting from the count matrix.  It will provide an overview of the fundamentals of scRNAseq analysis, including data QC, normalization, exploration with PCAs and heatmaps, performing differential expression analysis and annotation of the differentially expressed genes.  Participants will also learn how to evaluate confounding and batch effects in the data. "><meta name=author content="Samuele Soraggi and Jose Alejandro Romero Herrera"><link rel=icon href=../img/favicon.png><meta name=generator content="mkdocs-1.4.2, mkdocs-material-8.5.10"><title>Read the data - Single cell sequencing data analysis - an introductiory but comprehensive course</title><link rel=stylesheet href=../assets/stylesheets/main.975780f9.min.css><link rel=stylesheet href=../assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../stylesheets/extra.css><script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#read-the-data class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Single cell sequencing data analysis - an introductiory but comprehensive course" class="md-header__button md-logo" aria-label="Single cell sequencing data analysis - an introductiory but comprehensive course" data-md-component=logo> <img src=../img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Single cell sequencing data analysis - an introductiory but comprehensive course </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Read the data </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/scRNASeq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/scRNASeq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=../python/installation.html class=md-tabs__link> Python </a> </li> <li class=md-tabs__item> <a href=Part01_read_the_data.html class="md-tabs__link md-tabs__link--active"> R </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Single cell sequencing data analysis - an introductiory but comprehensive course" class="md-nav__button md-logo" aria-label="Single cell sequencing data analysis - an introductiory but comprehensive course" data-md-component=logo> <img src=../img/logo.png alt=logo> </a> Single cell sequencing data analysis - an introductiory but comprehensive course </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/scRNASeq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/scRNASeq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../index.html class=md-nav__link> Course Introduction </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_2 type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2> Python <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Python data-md-level=1> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Python </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../python/installation.html class=md-nav__link> Setup and installation </a> </li> <li class=md-nav__item> <a href=../python/Part01_read_the_data.html class=md-nav__link> Read the data </a> </li> <li class=md-nav__item> <a href=../python/Part02_filtering_sample2.html class=md-nav__link> Quality Control (QC) and filtering </a> </li> <li class=md-nav__item> <a href=../python/Part03_normalize_and_integrate.html class=md-nav__link> Normalize and Integrate </a> </li> <li class=md-nav__item> <a href=../python/Part04_clustering.html class=md-nav__link> Cell clustering and Differential Expression (DE) </a> </li> <li class=md-nav__item> <a href=../python/Part05_cell_fate.html class=md-nav__link> Pseudotimes and cell fates </a> </li> <li class=md-nav__item> <a href=../python/Part06_Condition_analysis.html class=md-nav__link> Cross-data analysis </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_2_8 type=checkbox id=__nav_2_8> <label class=md-nav__link for=__nav_2_8> Extra <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Extra data-md-level=2> <label class=md-nav__title for=__nav_2_8> <span class="md-nav__icon md-icon"></span> Extra </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../python/Part02_filtering_sample3.html class=md-nav__link> Filtering another sample </a> </li> <li class=md-nav__item> <a href=../python/Part02_filtering_sample1.html class=md-nav__link> Filtering a low quality sample </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_3 type=checkbox id=__nav_3 checked> <label class=md-nav__link for=__nav_3> R <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=R data-md-level=1> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> R </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Read the data <span class="md-nav__icon md-icon"></span> </label> <a href=Part01_read_the_data.html class="md-nav__link md-nav__link--active"> Read the data </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#setup class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=#loading-datasets class=md-nav__link> Loading datasets </a> </li> <li class=md-nav__item> <a href=#seurat-object-structure class=md-nav__link> Seurat object structure </a> <nav class=md-nav aria-label="Seurat object structure"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#exploring-the-seurat-object class=md-nav__link> Exploring the Seurat Object </a> <nav class=md-nav aria-label="Exploring the Seurat Object"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#assay-object class=md-nav__link> Assay object </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#accessing-data-from-the-seurat-object class=md-nav__link> Accessing data from the Seurat Object </a> </li> <li class=md-nav__item> <a href=#subsetting-data class=md-nav__link> Subsetting data </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#saving-and-loading-a-seurat-object class=md-nav__link> Saving and loading a Seurat object </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=Part02_filtering_sample2.html class=md-nav__link> Preprocessing </a> </li> <li class=md-nav__item> <a href=Part03_normalize_and_dim_reduc.html class=md-nav__link> Normalization </a> </li> <li class=md-nav__item> <a href=Part04_integration.html class=md-nav__link> Integration </a> </li> <li class=md-nav__item> <a href=Part05_clustering.html class=md-nav__link> Clustering and Differential Expression </a> </li> <li class=md-nav__item> <a href=Part06_pseudotime.html class=md-nav__link> Pseudotimes </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_3_7 type=checkbox id=__nav_3_7> <label class=md-nav__link for=__nav_3_7> Extra <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Extra data-md-level=2> <label class=md-nav__title for=__nav_3_7> <span class="md-nav__icon md-icon"></span> Extra </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=Part02_filtering_sample3.html class=md-nav__link> Preprocessing another sample </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#setup class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=#loading-datasets class=md-nav__link> Loading datasets </a> </li> <li class=md-nav__item> <a href=#seurat-object-structure class=md-nav__link> Seurat object structure </a> <nav class=md-nav aria-label="Seurat object structure"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#exploring-the-seurat-object class=md-nav__link> Exploring the Seurat Object </a> <nav class=md-nav aria-label="Exploring the Seurat Object"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#assay-object class=md-nav__link> Assay object </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#accessing-data-from-the-seurat-object class=md-nav__link> Accessing data from the Seurat Object </a> </li> <li class=md-nav__item> <a href=#subsetting-data class=md-nav__link> Subsetting data </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#saving-and-loading-a-seurat-object class=md-nav__link> Saving and loading a Seurat object </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/hds-sandbox/scRNASeq_course/edit/master/docs/R/Part01_read_the_data.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1 id=read-the-data>Read the data<a class=headerlink href=#read-the-data title="Permanent link">&para;</a></h1> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>&#128368; <strong>Time Estimation:</strong> 30-60 minutes </p> <p>&#128172; <strong>Learning Objectives:</strong> </p> <ol> <li>Get an overview of the <code>Seurat</code> package and the <code>R</code> language syntax</li> <li>Learn and explore the data structure containing a single cell dataset</li> <li>Understand and apply basic interactions with the transcript matrix and the components of a dataset</li> </ol> </div> <p>We will use <code>Seurat</code> as the main tool for the analysis, where we will also apply some other packages. Seurat has a comprehensive <a href=https://satijalab.org/seurat/ >manual webpage</a> that includes many different tutorials you can use for further practicing. Needed packages for this workshop are loaded with the function <code>library()</code>, which will make the tools available for us.</p> <p>An alternative and well-established tool for <code>R</code> users is <a href=https://scanpy.readthedocs.io/en/stable/ >scanpy</a>. This is used in the <code>python</code> version of this course.</p> <h2 id=setup>Setup<a class=headerlink href=#setup title="Permanent link">&para;</a></h2> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=nf>library</span><span class=p>(</span><span class=n>tidyverse</span><span class=p>)</span>
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a><span class=nf>library</span><span class=p>(</span><span class=n>Seurat</span><span class=p>)</span>
<a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a><span class=nf>library</span><span class=p>(</span><span class=n>SeuratDisk</span><span class=p>)</span>
<a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a><span class=nf>library</span><span class=p>(</span><span class=n>patchwork</span><span class=p>)</span>
</code></pre></div> <p>The Seurat functions will help us run all the necessary steps of the analysis from preprocessing, to clustering and visualization. There are also several helper and wrapper functions from other libraries that we will use to run the analysis. Use the <code>help()</code> function to see what another function does, and it should appear on the <code>help tab</code></p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=nf>help</span><span class=p>(</span><span class=n>FindClusters</span><span class=p>)</span>
</code></pre></div> <h2 id=loading-datasets>Loading datasets<a class=headerlink href=#loading-datasets title="Permanent link">&para;</a></h2> <p>Data can be loaded from different possible formats, which usually has as a dedicated function for loading. For example, the <code>Read10X()</code> function reads in the output of the <a href=https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger>cellranger</a> pipeline from 10X, returning a unique molecular identified (UMI) count matrix. The values in this matrix represent the number of molecules for each feature (i.e. gene; row) that are detected in each cell (column).</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>Read10X</span><span class=p>(</span><span class=n>data.dir</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;../../../sandbox_scRNA_testAndFeedback/scRNASeq_course/Data/cellranger_sample2/outs/filtered_feature_bc_matrix/&quot;</span><span class=p>)</span>
<a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a><span class=n>sample_3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>Read10X</span><span class=p>(</span><span class=n>data.dir</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;../../../sandbox_scRNA_testAndFeedback/scRNASeq_course/Data/cellranger_sample3/outs/filtered_feature_bc_matrix/&quot;</span><span class=p>)</span>
</code></pre></div> <p>We next use the count matrix to create a <code>Seurat</code> object, which is the main data structure used in the study.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=c1># Initialize the Seurat object with the raw (non-normalized data).</span>
<a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a><span class=n>sample_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>CreateSeuratObject</span><span class=p>(</span><span class=n>counts</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>project</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;spermatogenesis&quot;</span><span class=p>)</span>
<a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a><span class=n>sample_3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>CreateSeuratObject</span><span class=p>(</span><span class=n>counts</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_3</span><span class=p>,</span><span class=w> </span><span class=n>project</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;spermatogenesis&quot;</span><span class=p>)</span>
</code></pre></div> <details class=info> <summary>What does data in a count matrix look like?</summary> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=c1># Lets examine a few genes in the first thirty cells</span>
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a><span class=n>sample_2</span><span class=p>[</span><span class=nf>c</span><span class=p>(</span><span class=m>1</span><span class=o>:</span><span class=m>3</span><span class=p>),</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>30</span><span class=p>]</span>
</code></pre></div> <p>The <code>.</code> values in the matrix represent 0s (no molecules detected). Since most values in an scRNA-seq matrix are 0, Seurat uses a sparse-matrix representation whenever possible. This results in significant memory and speed savings for Drop-seq/inDrop/10x data.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=n>dense.size</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>object.size</span><span class=p>(</span><span class=nf>as.matrix</span><span class=p>(</span><span class=n>sample_2</span><span class=p>))</span><span class=w> </span>
<a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a><span class=n>dense.size</span><span class=w> </span>
<a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a><span class=n>sparse.size</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>object.size</span><span class=p>(</span><span class=n>sample_2</span><span class=p>)</span><span class=w> </span>
<a id=__codelineno-5-4 name=__codelineno-5-4 href=#__codelineno-5-4></a><span class=n>sparse.size</span><span class=w> </span>
<a id=__codelineno-5-5 name=__codelineno-5-5 href=#__codelineno-5-5></a><span class=n>dense.size</span><span class=w> </span><span class=o>/</span><span class=w> </span><span class=n>sparse.size</span>
</code></pre></div> </details> <h2 id=seurat-object-structure>Seurat object structure<a class=headerlink href=#seurat-object-structure title="Permanent link">&para;</a></h2> <div class="admonition info"> <p class=admonition-title>Info</p> <p>For a technical discussion of the <code>Seurat</code> object structure, check out the <a href=https://github.com/satijalab/seurat/wiki>Seurat GitHub Wiki</a>.</p> </div> <p>The <code>Seurat</code> object structure is a bit technical, but in a nutshell, it serves as a container that holds separate parts of single cell analyses:</p> <ul> <li>The <code>Assay</code> sub-object contains data (like the count matrix).</li> <li>The <code>DimReduc</code> sub-object holds the analyses performed on the data (like PCA, or clustering results).</li> </ul> <p>In addition, each of these objects (<code>Seurat</code>, <code>Assay</code> and <code>DimReduc</code>) are also divided in separate <code>Slots</code> that contain specific information within them. These slots can be accessed using the <code>@</code> operator (e.g <code>sample_2@assays</code>).</p> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=nf>slotNames</span><span class=p>(</span><span class=n>sample_2</span><span class=p>)</span>
<a id=__codelineno-6-2 name=__codelineno-6-2 href=#__codelineno-6-2></a><span class=n>sample_2</span><span class=o>@</span><span class=n>assays</span>
</code></pre></div> <p>For example, the raw count matrix is stored in <code>sample_2[["RNA"]]@counts</code>. <code>sample[["RNA"]]</code> directs to the <code>Assay</code> sub-object called "RNA", which contains the slot <code>@counts</code>. Normalized and scaled counts will be saved in their specific slots (<code>data</code> and <code>scale.data</code>, respectively).</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=nf>head</span><span class=p>(</span><span class=n>sample_2</span><span class=p>[[</span><span class=s>&quot;RNA&quot;</span><span class=p>]]</span><span class=o>@</span><span class=n>counts</span><span class=p>)</span>
</code></pre></div> <p>We will look into these slots as they are needed through the course, so do not worry about it! Nevertheless, here are some of the most useful ones for the <code>Seurat</code> object:</p> <table> <thead> <tr> <th>Slot</th> <th>Function</th> </tr> </thead> <tbody> <tr> <td><code>assays</code></td> <td>A list of assays within this object, such as RNAseq</td> </tr> <tr> <td><code>meta.data</code></td> <td>Contains cell-level meta data</td> </tr> <tr> <td><code>active.assay</code></td> <td>Name of active, or default, assay</td> </tr> <tr> <td><code>active.ident</code></td> <td>Identity classes for the current object</td> </tr> <tr> <td><code>reductions</code></td> <td>A list of DimReduc objects, like PCA or UMAP</td> </tr> <tr> <td><code>project.name</code></td> <td>User-defined project name (optional)</td> </tr> </tbody> </table> <h3 id=exploring-the-seurat-object>Exploring the Seurat Object<a class=headerlink href=#exploring-the-seurat-object title="Permanent link">&para;</a></h3> <p>Summary information about <code>Seurat</code> objects can be had quickly and easily using standard R functions. Object shape/dimensions can be found using the <code>dim</code>, <code>ncol</code>, and <code>nrow</code> functions; cell and feature names can be found using the <code>colnames</code> and <code>rownames</code> functions, respectively, or the <code>dimnames</code> function. A vector of names of <code>Assay</code> and <code>DimReduc</code> objects contained in a <code>Seurat</code> object can be obtained by using <code>names</code>.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=n>sample_2</span>
</code></pre></div> <p><code>nrow()</code> and <code>ncol()</code> provide the number of features and cells in the active assay, respectively. <code>dim()</code> provides both nrow and ncol at the same time:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=nf>dim</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=nf>head</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>rownames</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>))</span>
</code></pre></div> <p>In addition to rownames and colnames, one can use <code>dimnames()</code>, which provides a two-length list with both rownames and colnames:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=nf>head</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>colnames</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>))</span>
</code></pre></div> <h4 id=assay-object>Assay object<a class=headerlink href=#assay-object title="Permanent link">&para;</a></h4> <p>For typical scRNA-seq experiments, a Seurat object will have a single Assay ("RNA"). This assay will also store multiple ‘transformations’ of the data, including raw counts (@counts slot), normalized data (@data slot), and scaled data for dimensional reduction (@scale.data slot).</p> <p>For more complex experiments, an object could contain multiple assays. These could include multi-modal data types (CITE-seq antibody-derived tags, ADTs), or imputed/batch-corrected measurements. Each of those assays has the option to store the same data transformations as well.</p> <p>The slots of the <code>Assay</code> object contain the count matrix and its transformations, but it can also contain gene metadata</p> <table> <thead> <tr> <th>Slot</th> <th>Function</th> </tr> </thead> <tbody> <tr> <td><code>counts</code></td> <td>Stores unnormalized data such as raw counts or TPMs</td> </tr> <tr> <td><code>data</code></td> <td>Normalized data matrix</td> </tr> <tr> <td><code>scale.data</code></td> <td>Scaled data matrix</td> </tr> <tr> <td><code>meta.features</code></td> <td>Feature-level meta data</td> </tr> </tbody> </table> <h3 id=accessing-data-from-the-seurat-object>Accessing data from the Seurat Object<a class=headerlink href=#accessing-data-from-the-seurat-object title="Permanent link">&para;</a></h3> <p>Pulling specific <code>Assay</code>, objects can be done with the double <code>[[</code> <a href=https://www.rdocumentation.org/packages/base/versions/3.5.1/topics/Extract>extract operator</a>. Adding new objects to a <code>Seurat</code> object is also done with the double <code>[[</code> <a href=https://www.rdocumentation.org/packages/base/versions/3.5.1/topics/Extract>extract operator</a>; Seurat will figure out where in the <code>Seurat</code> object a new associated object belongs.</p> <p>If you want to know all the objects inside the <code>Seurat</code> object, you can use the <code>names()</code> function. These can be passed to the double <code>[[</code> extract operator to pull them from the Seurat object:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=nf>names</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>)</span>
<a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a><span class=n>sample_2</span><span class=p>[[</span><span class=s>&quot;RNA&quot;</span><span class=p>]]</span>
</code></pre></div> <p>Accessing data from an <code>Seurat</code> object is done with the <code>GetAssayData</code> function. Adding expression data to either the <code>counts</code>, <code>data</code>, or <code>scale.data</code> slots can be done with <code>SetAssayData</code>. New data must have the same cells in the same order as the current expression data. Data added to <code>counts</code> or <code>data</code> must have the same features as the current expression data.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=nf>GetAssayData</span><span class=p>(</span><span class=n>object</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>slot</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&#39;scale.data&#39;</span><span class=p>)[</span><span class=m>1</span><span class=o>:</span><span class=m>3</span><span class=p>,</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>3</span><span class=p>]</span>
</code></pre></div> <p>Cell-level meta data can be accessed with the single <code>[[</code> <a href=https://www.rdocumentation.org/packages/base/versions/3.5.1/topics/Extract>extract operator</a> or using the <code>$</code> sigil. Pulling with the <code>$</code> sigil means only one bit of meta data can be pulled at a time, though tab-autocompletion has been enabled for it, making it ideal for interactive use. Adding cell-level meta data can be set using the single <code>[[</code> <a href=https://www.rdocumentation.org/packages/base/versions/3.5.1/topics/Extract>extract operator</a> as well, or by using <code>AddMetaData</code>.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a><span class=c1># Cell-level meta data is stored as a data frame</span>
<a id=__codelineno-14-2 name=__codelineno-14-2 href=#__codelineno-14-2></a><span class=c1># Standard data frame functions work on the meta data data frame</span>
<a id=__codelineno-14-3 name=__codelineno-14-3 href=#__codelineno-14-3></a><span class=nf>colnames</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>[[]])</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=c1># The $ sigil can only pull bit of meta data at a time; however, tab-autocompletion</span>
<a id=__codelineno-15-2 name=__codelineno-15-2 href=#__codelineno-15-2></a><span class=c1># has been enabled for the $ sigil, making it ideal for interactive use</span>
<a id=__codelineno-15-3 name=__codelineno-15-3 href=#__codelineno-15-3></a><span class=nf>head</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>)</span>
</code></pre></div> <div class="admonition note"> <p class=admonition-title>Note</p> <p>We will take a look at the <code>DimReduc</code> object in the next sessions!</p> </div> <h3 id=subsetting-data>Subsetting data<a class=headerlink href=#subsetting-data title="Permanent link">&para;</a></h3> <p>You can subset Seurat objects using the function <code>subset()</code>. You can subset based on metadata, expression level of a gene or an identity class. Here you can find some examples:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-16-1 name=__codelineno-16-1 href=#__codelineno-16-1></a><span class=c1># Subset Seurat object based on a cell identity class</span>
<a id=__codelineno-16-2 name=__codelineno-16-2 href=#__codelineno-16-2></a><span class=nf>subset</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>idents</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;B cells&quot;</span><span class=p>)</span>
<a id=__codelineno-16-3 name=__codelineno-16-3 href=#__codelineno-16-3></a><span class=nf>subset</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>idents</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;CD4 T cells&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;CD8 T cells&quot;</span><span class=p>),</span><span class=w> </span><span class=n>invert</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span>
<a id=__codelineno-16-4 name=__codelineno-16-4 href=#__codelineno-16-4></a>
<a id=__codelineno-16-5 name=__codelineno-16-5 href=#__codelineno-16-5></a><span class=c1># Subset on the expression level of a gene/feature</span>
<a id=__codelineno-16-6 name=__codelineno-16-6 href=#__codelineno-16-6></a><span class=nf>subset</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>subset</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>MS4A1</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=m>3</span><span class=p>)</span>
<a id=__codelineno-16-7 name=__codelineno-16-7 href=#__codelineno-16-7></a>
<a id=__codelineno-16-8 name=__codelineno-16-8 href=#__codelineno-16-8></a><span class=c1># Subset on a combination of criteria</span>
<a id=__codelineno-16-9 name=__codelineno-16-9 href=#__codelineno-16-9></a><span class=nf>subset</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>subset</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>MS4A1</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=m>3</span><span class=w> </span><span class=o>&amp;</span><span class=w> </span><span class=n>PC1</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=m>5</span><span class=p>)</span>
<a id=__codelineno-16-10 name=__codelineno-16-10 href=#__codelineno-16-10></a><span class=nf>subset</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>subset</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>MS4A1</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=m>3</span><span class=p>,</span><span class=w> </span><span class=n>idents</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;B cells&quot;</span><span class=p>)</span>
<a id=__codelineno-16-11 name=__codelineno-16-11 href=#__codelineno-16-11></a>
<a id=__codelineno-16-12 name=__codelineno-16-12 href=#__codelineno-16-12></a><span class=c1># Subset on a value in the object meta data</span>
<a id=__codelineno-16-13 name=__codelineno-16-13 href=#__codelineno-16-13></a><span class=nf>subset</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>subset</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>orig.ident</span><span class=w> </span><span class=o>==</span><span class=w> </span><span class=s>&quot;Replicate1&quot;</span><span class=p>)</span>
<a id=__codelineno-16-14 name=__codelineno-16-14 href=#__codelineno-16-14></a>
<a id=__codelineno-16-15 name=__codelineno-16-15 href=#__codelineno-16-15></a><span class=c1># Downsample the number of cells per identity class</span>
<a id=__codelineno-16-16 name=__codelineno-16-16 href=#__codelineno-16-16></a><span class=nf>subset</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>downsample</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>100</span><span class=p>)</span>
</code></pre></div> <h2 id=saving-and-loading-a-seurat-object>Saving and loading a Seurat object<a class=headerlink href=#saving-and-loading-a-seurat-object title="Permanent link">&para;</a></h2> <p>You can save the Seurat object using the H5Seurat file format:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-17-1 name=__codelineno-17-1 href=#__codelineno-17-1></a><span class=nf>SaveH5Seurat</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>filename</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;sample_2.h5Seurat&quot;</span><span class=p>)</span>
</code></pre></div> <p>If you plan to work with <code>scanpy</code>, you might want to convert H5Seurat into h5ad format. This is easily done using the <code>Convert()</code> function. You can, of course, transform an h5ad file into H5Seurat using the same function</p> <div class=highlight><pre><span></span><code><a id=__codelineno-18-1 name=__codelineno-18-1 href=#__codelineno-18-1></a><span class=nf>Convert</span><span class=p>(</span><span class=s>&quot;sample_2.h5Seurat&quot;</span><span class=p>,</span><span class=w> </span><span class=n>dest</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;h5ad&quot;</span><span class=p>)</span>
<a id=__codelineno-18-2 name=__codelineno-18-2 href=#__codelineno-18-2></a><span class=nf>Convert</span><span class=p>(</span><span class=s>&quot;sample_2.h5ad&quot;</span><span class=p>,</span><span class=w> </span><span class=n>dest</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;h5seurat&quot;</span><span class=p>,</span><span class=w> </span><span class=n>overwrite</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span>
</code></pre></div> <p>To load a H5Seurat object, use the following function <code>LoadH5Seurat()</code>:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-19-1 name=__codelineno-19-1 href=#__codelineno-19-1></a><span class=nf>LoadH5Seurat</span><span class=p>(</span><span class=s>&quot;sample_2.h5seurat&quot;</span><span class=p>)</span>
</code></pre></div> <p>Then you can use it in python!</p> <div class=highlight><pre><span></span><code><a id=__codelineno-20-1 name=__codelineno-20-1 href=#__codelineno-20-1></a><span class=kn>import</span> <span class=nn>scanpy</span>
<a id=__codelineno-20-2 name=__codelineno-20-2 href=#__codelineno-20-2></a>
<a id=__codelineno-20-3 name=__codelineno-20-3 href=#__codelineno-20-3></a><span class=n>adata</span> <span class=o>=</span> <span class=n>scanpy</span><span class=o>.</span><span class=n>read_h5ad</span><span class=p>(</span><span class=s2>&quot;sample_2.h5ad&quot;</span><span class=p>)</span>
<a id=__codelineno-20-4 name=__codelineno-20-4 href=#__codelineno-20-4></a><span class=n>adata</span>
</code></pre></div> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../python/Part02_filtering_sample1.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Filtering a low quality sample" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Filtering a low quality sample </div> </div> </a> <a href=Part02_filtering_sample2.html class="md-footer__link md-footer__link--next" aria-label="Next: Preprocessing" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Preprocessing </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox/ target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "..", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.footer", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "../assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script> <script src=../assets/javascripts/bundle.5a2dcb6a.min.js></script> <script src=../javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> <script src=../javascripts/embed-amd.js></script> </body> </html>