<!doctype html><html lang=en class=no-js> <head><script type=application/ld+json>
    {
      "@context": "https://schema.org",
      "@id": "https://hds-sandbox.github.io/scRNASeq_course/",
      "@type": "LearningResource",
      "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
      "description": "This workshop contains a basic tutorial on how to approach scRNAseq experiments, starting from fastq files out of the sequencer. Thus, the workshop does not include any information of laboratory protocols, library preparation or any wet-lab related procedures.",
      "keywords": [
        "bioinformatics, RNAseq, single cell, NGS, python, R"
      ],
      "learningResourceType": [
        "e-learning"
      ],
      "license": [
        {
          "@id": "https://creativecommons.org/licenses/by-sa/4.0/legalcode",
          "@type": "CreativeWork"
        }
      ],
      "mentions": [
        {
          "@id": "10.5281/zenodo.7660225",
          "@type": "Thing"
        },
        {
          "@id": "10.5281/zenodo.7565997",
          "@type": "Thing"
        }
      ],
      "name": "Introduction to single cell RNAseq analysis workshop",
      "educationalLevel": "Beginner",
      "audience": "PhD, postdocs, MSc, Researchers"
    }
    </script><title>Cell clustering and Differential Expression (DE) - Single cell sequencing data analysis - an introductiory but comprehensive course</title><link rel=stylesheet href=../assets/stylesheets/main.240905d7.min.css><link rel=stylesheet href=../assets/stylesheets/palette.a0c5b2b5.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../stylesheets/neoteroi-mkdocs.css><link rel=stylesheet href=../stylesheets/extra.css><script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#cell-clustering-and-differential-expression-de class=md-skip> Skip to content </a> </div> <div data-md-component=announce> <aside class=md-banner> <div class="md-banner__inner md-grid md-typeset"> Page under construction - course not finalized </div> </aside> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Single cell sequencing data analysis - an introductiory but comprehensive course" class="md-header__button md-logo" aria-label="Single cell sequencing data analysis - an introductiory but comprehensive course" data-md-component=logo> <img src=../img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Single cell sequencing data analysis - an introductiory but comprehensive course </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Cell clustering and Differential Expression (DE) </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/scRNASeq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/scRNASeq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=../python/installation.html class=md-tabs__link> Python </a> </li> <li class=md-tabs__item> <a href=Part01_read_the_data.html class="md-tabs__link md-tabs__link--active"> R </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Single cell sequencing data analysis - an introductiory but comprehensive course" class="md-nav__button md-logo" aria-label="Single cell sequencing data analysis - an introductiory but comprehensive course" data-md-component=logo> <img src=../img/logo.png alt=logo> </a> Single cell sequencing data analysis - an introductiory but comprehensive course </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/scRNASeq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/scRNASeq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../index.html class=md-nav__link> Course Introduction </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> Python <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Python </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../python/installation.html class=md-nav__link> Setup and installation </a> </li> <li class=md-nav__item> <a href=../python/Part01_read_the_data.html class=md-nav__link> Read the data </a> </li> <li class=md-nav__item> <a href=../python/Part02_filtering_sample2.html class=md-nav__link> Quality Control (QC) and filtering </a> </li> <li class=md-nav__item> <a href=../python/Part03_normalize_and_integrate.html class=md-nav__link> Normalize and Integrate </a> </li> <li class=md-nav__item> <a href=../python/Part04_clustering.html class=md-nav__link> Cell clustering and Differential Expression (DE) </a> </li> <li class=md-nav__item> <a href=../python/Part05_cell_fate.html class=md-nav__link> Pseudotimes and cell fates </a> </li> <li class=md-nav__item> <a href=../python/Part06_Condition_analysis.html class=md-nav__link> Cross-data analysis </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8> <label class=md-nav__link for=__nav_2_8 id=__nav_2_8_label tabindex=0> Extra <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_8_label aria-expanded=false> <label class=md-nav__title for=__nav_2_8> <span class="md-nav__icon md-icon"></span> Extra </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../python/Part02_filtering_sample3.html class=md-nav__link> Filtering another sample </a> </li> <li class=md-nav__item> <a href=../python/Part02_filtering_sample1.html class=md-nav__link> Filtering a low quality sample </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3 checked> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> R <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=true> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> R </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=Part01_read_the_data.html class=md-nav__link> Read the data </a> </li> <li class=md-nav__item> <a href=Part02_filtering_sample2.html class=md-nav__link> Preprocessing </a> </li> <li class=md-nav__item> <a href=Part03_normalize_and_dim_reduc.html class=md-nav__link> Normalization </a> </li> <li class=md-nav__item> <a href=Part04_integration.html class=md-nav__link> Integration </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Clustering and Differential Expression <span class="md-nav__icon md-icon"></span> </label> <a href=Part05_clustering.html class="md-nav__link md-nav__link--active"> Clustering and Differential Expression </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#setup class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=#cell-clustering class=md-nav__link> Cell clustering </a> <nav class=md-nav aria-label="Cell clustering"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#visualization class=md-nav__link> Visualization </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#finding-differentially-expressed-markers class=md-nav__link> Finding differentially expressed markers </a> <nav class=md-nav aria-label="Finding differentially expressed markers"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#alternate-testing-for-differential-expression class=md-nav__link> Alternate testing for differential expression </a> </li> <li class=md-nav__item> <a href=#visualization-of-marker-genes class=md-nav__link> Visualization of marker genes </a> </li> <li class=md-nav__item> <a href=#cluster-annotation-through-known-markers class=md-nav__link> Cluster annotation through known markers </a> </li> <li class=md-nav__item> <a href=#calculating-cell-scores class=md-nav__link> Calculating cell scores </a> </li> <li class=md-nav__item> <a href=#assigning-names-to-clusters class=md-nav__link> Assigning names to clusters </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#selecting-and-subclustering class=md-nav__link> Selecting and subclustering </a> <nav class=md-nav aria-label="Selecting and subclustering"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#subsetting-the-data class=md-nav__link> Subsetting the data </a> </li> <li class=md-nav__item> <a href=#recalculate-clusters class=md-nav__link> Recalculate clusters </a> </li> <li class=md-nav__item> <a href=#differential-expression class=md-nav__link> Differential expression </a> </li> <li class=md-nav__item> <a href=#reintegrating-the-subclusters class=md-nav__link> Reintegrating the subclusters </a> </li> <li class=md-nav__item> <a href=#cluster-correlations class=md-nav__link> Cluster correlations </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#save-our-clustered-data class=md-nav__link> Save our clustered data </a> </li> <li class=md-nav__item> <a href=#wrapping-up class=md-nav__link> Wrapping up </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=Part06_pseudotime.html class=md-nav__link> Pseudotimes </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3_7> <label class=md-nav__link for=__nav_3_7 id=__nav_3_7_label tabindex=0> Extra <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_3_7_label aria-expanded=false> <label class=md-nav__title for=__nav_3_7> <span class="md-nav__icon md-icon"></span> Extra </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=Part02_filtering_sample3.html class=md-nav__link> Preprocessing another sample </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#setup class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=#cell-clustering class=md-nav__link> Cell clustering </a> <nav class=md-nav aria-label="Cell clustering"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#visualization class=md-nav__link> Visualization </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#finding-differentially-expressed-markers class=md-nav__link> Finding differentially expressed markers </a> <nav class=md-nav aria-label="Finding differentially expressed markers"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#alternate-testing-for-differential-expression class=md-nav__link> Alternate testing for differential expression </a> </li> <li class=md-nav__item> <a href=#visualization-of-marker-genes class=md-nav__link> Visualization of marker genes </a> </li> <li class=md-nav__item> <a href=#cluster-annotation-through-known-markers class=md-nav__link> Cluster annotation through known markers </a> </li> <li class=md-nav__item> <a href=#calculating-cell-scores class=md-nav__link> Calculating cell scores </a> </li> <li class=md-nav__item> <a href=#assigning-names-to-clusters class=md-nav__link> Assigning names to clusters </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#selecting-and-subclustering class=md-nav__link> Selecting and subclustering </a> <nav class=md-nav aria-label="Selecting and subclustering"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#subsetting-the-data class=md-nav__link> Subsetting the data </a> </li> <li class=md-nav__item> <a href=#recalculate-clusters class=md-nav__link> Recalculate clusters </a> </li> <li class=md-nav__item> <a href=#differential-expression class=md-nav__link> Differential expression </a> </li> <li class=md-nav__item> <a href=#reintegrating-the-subclusters class=md-nav__link> Reintegrating the subclusters </a> </li> <li class=md-nav__item> <a href=#cluster-correlations class=md-nav__link> Cluster correlations </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#save-our-clustered-data class=md-nav__link> Save our clustered data </a> </li> <li class=md-nav__item> <a href=#wrapping-up class=md-nav__link> Wrapping up </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <div><h1 id=cell-clustering-and-differential-expression-de>Cell clustering and Differential Expression (DE)<a class=headerlink href=#cell-clustering-and-differential-expression-de title="Permanent link">¶</a></h1> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>🕰 <strong>Time Estimation:</strong> 45-60 minutes</p> <p>💬 <strong>Learning Objectives:</strong> </p> <ol> <li>Perform unsupervised clustering on the dataset</li> <li>Identify potential cell clusters by visualizing marker genes on the UMAP plot</li> <li>Understanding and applying differential gene expression analysis to verify cluster identities</li> <li>Performing an analysis of subclusters in the dataset</li> </ol> </div> <p>Spermatogenesis goes through different stages, starting from <em>SpermatogoniaA</em> cells, going into clonal expansion while keeping cells connected through cytoplasmic bridges (<em>SpermatogoniaB</em>), and then continuing with the meiotic process (<em>Spermatocites I and II</em>). Finally, cells become <em>Round spermatids</em>, which then elongate to become <em>Elongated spermatids and sperm</em>.</p> <p><img alt src=img/Part05_clustering/spermatogenesis.png></p> <p>Detecting those cell types is essential to answer biological questions such as: - which genes are most expressed for each cell type (beyond well known ones)? - in which proportion is every cell type present? - are there unknown cell types that I can identify?</p> <h2 id=setup>Setup<a class=headerlink href=#setup title="Permanent link">¶</a></h2> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=nf>library</span><span class=p>(</span><span class=n>tidyverse</span><span class=p>)</span>
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a><span class=nf>library</span><span class=p>(</span><span class=n>patchwork</span><span class=p>)</span>
<a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a><span class=nf>library</span><span class=p>(</span><span class=n>Seurat</span><span class=p>)</span>
<a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a><span class=nf>library</span><span class=p>(</span><span class=n>SeuratDisk</span><span class=p>)</span>
<a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a>
<a id=__codelineno-0-6 name=__codelineno-0-6 href=#__codelineno-0-6></a><span class=nf>LoadH5Seurat</span><span class=p>(</span><span class=s>"../../Data/notebooks_data/sample_123.filt.norm.red.h5Seurat"</span><span class=p>)</span>
</code></pre></div> <h2 id=cell-clustering>Cell clustering<a class=headerlink href=#cell-clustering title="Permanent link">¶</a></h2> <p>Seurat applies a graph-based clustering approach, building upon initial strategies in (<a href=http://www.cell.com/abstract/S0092-8674(15)00549-8>Macosko <em>et al</em></a>). Importantly, the <em>distance metric</em> which drives the clustering analysis (based on previously identified PCs) remains the same. However, Seurat’s approach to partitioning the cellular distance matrix into clusters is heavily inspired by graph-based clustering approaches to scRNA-seq data <a href=http://bioinformatics.oxfordjournals.org/content/early/2015/02/10/bioinformatics.btv088.abstract>[SNN-Cliq, Xu and Su, Bioinformatics, 2015]</a> and CyTOF data <a href=http://www.ncbi.nlm.nih.gov/pubmed/26095251>[PhenoGraph, Levine <em>et al</em>., Cell, 2015]</a>. Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected ‘quasi-cliques’ or ‘communities’.</p> <p>As in PhenoGraph, Seurat first constructs a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the <code>FindNeighbors()</code> function, and takes as input the previously defined dimensionality of the dataset (first 10 PCs).</p> <p>To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM <a href=http://dx.doi.org/10.1088/1742-5468/2008/10/P10008>[SLM, Blondel <em>et al</em>., Journal of Statistical Mechanics]</a>, to iteratively group cells together, with the goal of optimizing the standard modularity function. The <code>FindClusters()</code> function implements this procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters.</p> <p>It is recommended to set this parameter between 0.4-1.2, as it typically returns good results for single-cell datasets of around 3K cells.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=n>sample.combined.sct</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindNeighbors</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>10</span><span class=p>)</span>
<a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a><span class=n>sample.combined.sct</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindClusters</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>resolution</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.5</span><span class=p>)</span>
</code></pre></div> <p>The clusters can be found using the <code>Idents()</code> function or in the metadata of the assay.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=nf>head</span><span class=p>(</span><span class=nf>Idents</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>),</span><span class=w> </span><span class=m>5</span><span class=p>)</span>
<a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a><span class=nf>head</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=o>@</span><span class=n>meta.data</span><span class=p>)</span>
<a id=__codelineno-2-3 name=__codelineno-2-3 href=#__codelineno-2-3></a><span class=nf>head</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=o>$</span><span class=n>seurat.clusters</span><span class=p>)</span>
</code></pre></div> <h3 id=visualization>Visualization<a class=headerlink href=#visualization title="Permanent link">¶</a></h3> <p>Finally we can visualize our clusters in the data using the <code>DimPlot()</code> function.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=nf>DimPlot</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>'umap'</span><span class=p>)</span>
</code></pre></div> <div class="admonition question"> <p class=admonition-title>Exercise</p> <p>Play with the <code>resolution</code> argument of FindClusters. How is the parameter related to the number of clusters identified? <em>Note:</em> The new clusters are saved in a new column of the metadata dataframe.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=n>sample.combined.sct</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindClusters</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>resolution</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=o>??</span><span class=p>)</span>
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a><span class=nf>DimPlot</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>'umap'</span><span class=p>)</span>
</code></pre></div> </div> <h2 id=finding-differentially-expressed-markers>Finding differentially expressed markers<a class=headerlink href=#finding-differentially-expressed-markers title="Permanent link">¶</a></h2> <p>Seurat can help you find markers that define clusters via differential expression using the <code>FindAllMarkers()</code> function. By default, it identifies positive and negative markers of a single cluster (specified in <code>ident.1</code>), compared to all other cells. <code>FindAllMarkers()</code> automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells using the <code>FindMarkers()</code> function. <code>FindMarkers()</code> function is specially useful to identify specific markers for subclusters of cells.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=c1># Find all markers</span>
<a id=__codelineno-5-2 name=__codelineno-5-2 href=#__codelineno-5-2></a><span class=n>all.markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindAllMarkers</span><span class=p>(</span><span class=n>pbmc</span><span class=p>)</span>
<a id=__codelineno-5-3 name=__codelineno-5-3 href=#__codelineno-5-3></a>
<a id=__codelineno-5-4 name=__codelineno-5-4 href=#__codelineno-5-4></a><span class=c1># A cluster vs another</span>
<a id=__codelineno-5-5 name=__codelineno-5-5 href=#__codelineno-5-5></a><span class=n>cluster0v1.markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindMarkers</span><span class=p>(</span><span class=n>pbmc</span><span class=p>,</span><span class=w> </span><span class=n>ident.1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0</span><span class=p>,</span><span class=w> </span><span class=n>ident.2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=m>1</span><span class=p>))</span>
<a id=__codelineno-5-6 name=__codelineno-5-6 href=#__codelineno-5-6></a>
<a id=__codelineno-5-7 name=__codelineno-5-7 href=#__codelineno-5-7></a><span class=c1># A cluster vs other clusters</span>
<a id=__codelineno-5-8 name=__codelineno-5-8 href=#__codelineno-5-8></a><span class=n>cluster0v1_2.markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindMarkers</span><span class=p>(</span><span class=n>pbmc</span><span class=p>,</span><span class=w> </span><span class=n>ident.1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0</span><span class=p>,</span><span class=w> </span><span class=n>ident.2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=m>1</span><span class=p>,</span><span class=m>2</span><span class=p>))</span>
</code></pre></div> <p>The results dataframe looks like this:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=nf>head</span><span class=p>(</span><span class=n>all.markers</span><span class=p>)</span>
</code></pre></div> <ul> <li>p_val : p_val of the test used for differential expression (unadjusted)</li> <li>avg_log2FC : log fold-change of the average expression between the two groups. Positive values indicate that the feature is more highly expressed in the first group.</li> <li>pct.1 : The percentage of cells where the feature is detected in the first group</li> <li>pct.2 : The percentage of cells where the feature is detected in the second group</li> <li>p_val_adj : Adjusted p-value, based on Bonferroni correction using all features in the dataset.</li> <li>gene : feature names will appear as row names using <code>FindMarkers()</code>, or as a column with <code>FindAllMarkers()</code></li> <li>cluster : using the <code>FindAllMarkers()</code> will return which cluster had each marker</li> </ul> <p>The <code>min.pct</code> argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the <code>logfc.threshold</code> argument requires a feature to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of features that are unlikely to be highly discriminatory. As another option to speed up these computations, <code>max.cells.per.ident</code> can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significant and the most highly differentially expressed features will likely still rise to the top.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=c1># find markers for every cluster compared to all remaining cells, report only the positive ones with a LFC above 0.25 and at least 25% of cells express that feature.</span>
<a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a><span class=n>all.markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindAllMarkers</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>only.pos</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>,</span><span class=w> </span><span class=n>min.pct</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.25</span><span class=p>,</span><span class=w> </span><span class=n>logfc.threshold</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.25</span><span class=p>)</span>
<a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a><span class=n>all.markers</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>group_by</span><span class=p>(</span><span class=n>cluster</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>slice_max</span><span class=p>(</span><span class=n>n</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>2</span><span class=p>,</span><span class=w> </span><span class=n>order_by</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>avg_log2FC</span><span class=p>)</span><span class=w> </span><span class=c1># Show top 2 markers for each cluster</span>
</code></pre></div> <p>You can easily save your marker dataframes using <code>write.table()</code></p> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=nf>write.table</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>all.markers</span><span class=p>,</span><span class=w> </span><span class=n>file</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"../../Data/results/all_markers.tsv"</span><span class=p>,</span><span class=w> </span><span class=n>col.names</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>,</span><span class=w> </span><span class=n>row.names</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>,</span><span class=w> </span><span class=n>sep</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"\t"</span><span class=p>,</span><span class=w> </span><span class=n>quote</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
</code></pre></div> <h3 id=alternate-testing-for-differential-expression>Alternate testing for differential expression<a class=headerlink href=#alternate-testing-for-differential-expression title="Permanent link">¶</a></h3> <p>As a default, Seurat performs differential expression based on the non-parametric Wilcoxon rank sum test. Nontheless, the following differential expression tests are currently supported:</p> <ul> <li>"wilcox" : Wilcoxon rank sum test (default)</li> <li>"bimod" : Likelihood-ratio test for single cell feature expression, (McDavid et al., Bioinformatics, 2013)</li> <li>"roc" : Standard AUC classifier</li> <li>"t" : Student’s t-test</li> <li>"poisson" : Likelihood ratio test assuming an underlying negative binomial distribution. Use only for UMI-based datasets</li> <li>"negbinom" : Likelihood ratio test assuming an underlying negative binomial distribution. Use only for UMI-based datasets</li> <li>"LR" : Uses a logistic regression framework to determine differentially expressed genes. Constructs a logistic regression model predicting group membership based on each feature individually and compares this to a null model with a likelihood ratio test.</li> <li>"MAST" : GLM-framework that treates cellular detection rate as a covariate (Finak et al, Genome Biology, 2015) (Installation instructions)</li> <li>"DESeq2" : DE based on a model using the negative binomial distribution (Love et al, Genome Biology, 2014) (Installation instructions)</li> </ul> <p>For MAST and DESeq2, please ensure that these packages are installed separately in order to use them as part of Seurat. Once installed, use the <code>test.use</code> parameter can be used to specify which DE test to use. For example, the ROC test returns the ‘classification power’ for any individual marker (ranging from 0 - random, to 1 - perfect).</p> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=n>cluster0.markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindMarkers</span><span class=p>(</span><span class=n>pbmc</span><span class=p>,</span><span class=w> </span><span class=n>ident.1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0</span><span class=p>,</span><span class=w> </span><span class=n>logfc.threshold</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.25</span><span class=p>,</span><span class=w> </span><span class=n>test.use</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"roc"</span><span class=p>,</span><span class=w> </span><span class=n>only.pos</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span>
</code></pre></div> <h3 id=visualization-of-marker-genes>Visualization of marker genes<a class=headerlink href=#visualization-of-marker-genes title="Permanent link">¶</a></h3> <p>Seurat includes several tools for visualizing marker expression. <code>VlnPlot()</code> (shows expression probability distributions across clusters), and <code>FeaturePlot()</code> (visualizes feature expression on a UMAP or PCA plot) are some of the most commonly used visualizations. It is also suggested that you exploring <code>RidgePlot()</code>, and <code>DotPlot()</code> as additional methods to view your dataset.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=c1># You can visualize several markers at the same time using the features argument</span>
<a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a><span class=nf>VlnPlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>"ID4"</span><span class=p>,</span><span class=w> </span><span class=s>"MKI67"</span><span class=p>,</span><span class=s>"CD14"</span><span class=p>))</span>
<a id=__codelineno-10-3 name=__codelineno-10-3 href=#__codelineno-10-3></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>"ID4"</span><span class=p>,</span><span class=w> </span><span class=s>"MKI67"</span><span class=p>,</span><span class=s>"CD14"</span><span class=p>))</span>
<a id=__codelineno-10-4 name=__codelineno-10-4 href=#__codelineno-10-4></a><span class=nf>RidgePlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>"ID4"</span><span class=p>,</span><span class=w> </span><span class=s>"MKI67"</span><span class=p>,</span><span class=s>"CD14"</span><span class=p>))</span>
<a id=__codelineno-10-5 name=__codelineno-10-5 href=#__codelineno-10-5></a><span class=nf>DotPlot</span><span class=p>(</span><span class=n>object</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>"ID4"</span><span class=p>,</span><span class=w> </span><span class=s>"MKI67"</span><span class=p>,</span><span class=s>"CD14"</span><span class=p>))</span>
</code></pre></div> <p><code>DoHeatmap()</code> generates an expression heatmap for given cells and features. In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=n>all.markers</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a><span class=w>    </span><span class=nf>group_by</span><span class=p>(</span><span class=n>cluster</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a><span class=w>    </span><span class=nf>top_n</span><span class=p>(</span><span class=n>n</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>10</span><span class=p>,</span><span class=w> </span><span class=n>wt</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>avg_log2FC</span><span class=p>)</span><span class=w> </span><span class=o>-&gt;</span><span class=w> </span><span class=n>top10</span>
<a id=__codelineno-11-4 name=__codelineno-11-4 href=#__codelineno-11-4></a><span class=nf>DoHeatmap</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>top10</span><span class=o>$</span><span class=n>gene</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>NoLegend</span><span class=p>()</span>
</code></pre></div> <h3 id=cluster-annotation-through-known-markers>Cluster annotation through known markers<a class=headerlink href=#cluster-annotation-through-known-markers title="Permanent link">¶</a></h3> <p>We try to identify clusters of cells by looking at the expression of relevant marker genes. This requires a previous biological knowledge of those cell types, such that we can input the markers. Below, we define a list of markers, where for each cell type we define a list of markers. Then we will plot every list of markers on the UMAP plot:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=n>markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>list</span><span class=p>()</span><span class=w> </span><span class=c1>#make an empty dictionary</span>
<a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a><span class=c1>### SPERMATOCYTOGENESIS</span>
<a id=__codelineno-12-3 name=__codelineno-12-3 href=#__codelineno-12-3></a><span class=n>markers</span><span class=p>[</span><span class=s>'SpermatogoniaA'</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'ID4'</span><span class=p>)</span>
<a id=__codelineno-12-4 name=__codelineno-12-4 href=#__codelineno-12-4></a><span class=n>markers</span><span class=p>[</span><span class=s>'SpermatogoniaB'</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'MKI67'</span><span class=p>,</span><span class=s>'DMRT1'</span><span class=p>,</span><span class=s>'STRA8'</span><span class=p>)</span>
<a id=__codelineno-12-5 name=__codelineno-12-5 href=#__codelineno-12-5></a><span class=n>markers</span><span class=p>[</span><span class=s>'SpermatocytesI'</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'MEIOB'</span><span class=p>,</span><span class=s>'SYCP1'</span><span class=p>,</span><span class=s>'TEX101'</span><span class=p>)</span>
<a id=__codelineno-12-6 name=__codelineno-12-6 href=#__codelineno-12-6></a><span class=n>markers</span><span class=p>[</span><span class=s>'SpermatocytesII'</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'PIWIL1'</span><span class=p>,</span><span class=s>'SPATA16'</span><span class=p>,</span><span class=s>'CLGN'</span><span class=p>)</span>
<a id=__codelineno-12-7 name=__codelineno-12-7 href=#__codelineno-12-7></a><span class=c1>### SPERMIOGENESIS</span>
<a id=__codelineno-12-8 name=__codelineno-12-8 href=#__codelineno-12-8></a><span class=n>markers</span><span class=p>[</span><span class=s>'Round.Spt'</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'SPATA9'</span><span class=p>,</span><span class=s>'SPAM1'</span><span class=p>)</span><span class=w> </span><span class=c1>#Round spermatids</span>
<a id=__codelineno-12-9 name=__codelineno-12-9 href=#__codelineno-12-9></a><span class=n>markers</span><span class=p>[</span><span class=s>'Elong.Spt'</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'PRM1'</span><span class=p>,</span><span class=s>'PRM2'</span><span class=p>,</span><span class=s>'PRM3'</span><span class=p>,</span><span class=s>'AKAP4'</span><span class=p>)</span><span class=w> </span><span class=c1>#Elongated spermatids</span>
<a id=__codelineno-12-10 name=__codelineno-12-10 href=#__codelineno-12-10></a><span class=c1>### SOMATIC CELLS</span>
<a id=__codelineno-12-11 name=__codelineno-12-11 href=#__codelineno-12-11></a><span class=n>markers</span><span class=p>[</span><span class=s>'Sertoli'</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'VIM'</span><span class=p>,</span><span class=s>'CTSL'</span><span class=p>)</span>
<a id=__codelineno-12-12 name=__codelineno-12-12 href=#__codelineno-12-12></a><span class=n>markers</span><span class=p>[</span><span class=s>'Macroph'</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'CD14'</span><span class=p>)</span>
<a id=__codelineno-12-13 name=__codelineno-12-13 href=#__codelineno-12-13></a><span class=n>markers</span><span class=p>[</span><span class=s>'Leydig'</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'CFD'</span><span class=p>)</span>
<a id=__codelineno-12-14 name=__codelineno-12-14 href=#__codelineno-12-14></a><span class=n>markers</span><span class=p>[</span><span class=s>'Endothelial'</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'CD34'</span><span class=p>)</span>
<a id=__codelineno-12-15 name=__codelineno-12-15 href=#__codelineno-12-15></a><span class=n>markers</span><span class=p>[</span><span class=s>'Myoid'</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'ACTA2'</span><span class=p>)</span>
<a id=__codelineno-12-16 name=__codelineno-12-16 href=#__codelineno-12-16></a>
<a id=__codelineno-12-17 name=__codelineno-12-17 href=#__codelineno-12-17></a><span class=c1># remove markers missing in the dataset</span>
<a id=__codelineno-12-18 name=__codelineno-12-18 href=#__codelineno-12-18></a>
<a id=__codelineno-12-19 name=__codelineno-12-19 href=#__codelineno-12-19></a><span class=nf>for </span><span class=p>(</span><span class=n>i</span><span class=w> </span><span class=n>in</span><span class=w> </span><span class=nf>names</span><span class=p>(</span><span class=n>markers</span><span class=p>)){</span>
<a id=__codelineno-12-20 name=__codelineno-12-20 href=#__codelineno-12-20></a><span class=w>  </span><span class=n>markers</span><span class=p>[</span><span class=n>i</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>intersect</span><span class=p>(</span><span class=n>markers</span><span class=p>[</span><span class=n>i</span><span class=p>],</span><span class=w> </span><span class=nf>rownames</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>))</span>
<a id=__codelineno-12-21 name=__codelineno-12-21 href=#__codelineno-12-21></a><span class=p>}</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=s>'SpermatogoniaA'</span><span class=p>])</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=s>'SpermatogoniaB'</span><span class=p>])</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=s>'SpermatocytesI'</span><span class=p>])</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-16-1 name=__codelineno-16-1 href=#__codelineno-16-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=s>'SpermatocytesII'</span><span class=p>])</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-17-1 name=__codelineno-17-1 href=#__codelineno-17-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=s>'Round.Spt'</span><span class=p>])</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-18-1 name=__codelineno-18-1 href=#__codelineno-18-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=s>'Elong.Spt'</span><span class=p>])</span>
</code></pre></div> <p>Sertoli are often not possible to identify. They are big in size, meaning they are often not isolated successfully. Many of their markers are in common with other somatic cells. Also, their function as nurse cells for germ cells of the testis means that their marker genes are also expressed. We can see that CTSL is expressed in some germ cells, but not in other clusters, while VIM is expressed in a likely somatic cluster (but it is common to other somatic cell types).</p> <div class=highlight><pre><span></span><code><a id=__codelineno-19-1 name=__codelineno-19-1 href=#__codelineno-19-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=s>'Sertoli'</span><span class=p>])</span>
</code></pre></div> <p>Macrophage cells seem to be absent.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-20-1 name=__codelineno-20-1 href=#__codelineno-20-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=s>'Macroph'</span><span class=p>])</span>
</code></pre></div> <p>Leydig cells are likely to be missing as well.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-21-1 name=__codelineno-21-1 href=#__codelineno-21-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=s>'Leydig'</span><span class=p>])</span>
</code></pre></div> <p>There is a little endothelial cluster.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-22-1 name=__codelineno-22-1 href=#__codelineno-22-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=s>'Endothelial'</span><span class=p>])</span>
</code></pre></div> <p>And also a myoid cluster.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-23-1 name=__codelineno-23-1 href=#__codelineno-23-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=s>'Myoid'</span><span class=p>])</span>
</code></pre></div> <h3 id=calculating-cell-scores>Calculating cell scores<a class=headerlink href=#calculating-cell-scores title="Permanent link">¶</a></h3> <p>Using the <code>AddModuleScore()</code> function, you can use the previous list of different cell markers to give a "score" to each cell, instead of exploring every gene from your list of markers. This is extremely useful when you want to assign cell types if you have biological knowledge (markers). Nonetheless, you will have to make sure the list of markers will be specific to those clusters, otherwise there is a chance you cannot properly identify clusters using this list.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-24-1 name=__codelineno-24-1 href=#__codelineno-24-1></a><span class=nf>for </span><span class=p>(</span><span class=n>i</span><span class=w> </span><span class=n>in</span><span class=w> </span><span class=nf>names</span><span class=p>(</span><span class=n>markers</span><span class=p>)){</span>
<a id=__codelineno-24-2 name=__codelineno-24-2 href=#__codelineno-24-2></a><span class=w>  </span><span class=n>pbmc3k</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>AddModuleScore</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=n>i</span><span class=p>],</span><span class=w> </span><span class=n>name</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>i</span><span class=p>)</span>
<a id=__codelineno-24-3 name=__codelineno-24-3 href=#__codelineno-24-3></a><span class=p>}</span>
</code></pre></div> <p>We can visualize the scores either in a <code>FeaturePlot()</code> or a <code>VlnPlot()</code>. For example, let’s take a look at the SpermatogoniaB cell score:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-25-1 name=__codelineno-25-1 href=#__codelineno-25-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"umap"</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"SpermatogoniaB1"</span><span class=p>)</span>
<a id=__codelineno-25-2 name=__codelineno-25-2 href=#__codelineno-25-2></a><span class=nf>VlnPlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"SpermatogoniaB1"</span><span class=p>,</span><span class=w> </span><span class=n>group.by</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"seurat_clusters"</span><span class=p>)</span>
</code></pre></div> <p>It looks pretty good! However, lets check the SpermatogoniaA</p> <div class=highlight><pre><span></span><code><a id=__codelineno-26-1 name=__codelineno-26-1 href=#__codelineno-26-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"umap"</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"SpermatogoniaA1"</span><span class=p>)</span>
<a id=__codelineno-26-2 name=__codelineno-26-2 href=#__codelineno-26-2></a><span class=nf>VlnPlot</span><span class=p>(</span><span class=n>pbmc3k</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"SpermatogoniaA1"</span><span class=p>,</span><span class=w> </span><span class=n>group.by</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"seurat_clusters"</span><span class=p>)</span>
</code></pre></div> <p>As we can see, the set of features used to calculate module scores may not be specific enough to accurately depict a cell type. Some clusters seem to have just enough expression of those genes to have positive scores, and single cell dropouts makes it very difficult to score a cell based only on a single gene. However, it works very well when selecting marker genes from the <code>FindMarkers()</code> function.</p> <h3 id=assigning-names-to-clusters>Assigning names to clusters<a class=headerlink href=#assigning-names-to-clusters title="Permanent link">¶</a></h3> <p>Now that we know what kind of cell types our clusters are, we can name them properly. Write the names in the list <code>new_names</code>. You should be able to <strong>give a name for each cell type</strong>. Below is an example, but the names are not in the right position. If there is more than one cluster with same cell type, just write the name followed by a dot <code>.</code> and a number. For example, by writing for example <code>.1</code> and <code>.2</code> at the end of the names. We will remove the numbers afterwords.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-27-1 name=__codelineno-27-1 href=#__codelineno-27-1></a><span class=n>new_names</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span>
<a id=__codelineno-27-2 name=__codelineno-27-2 href=#__codelineno-27-2></a><span class=w>    </span><span class=s>'0'</span><span class=o>=</span><span class=s>'SpermatocitesII'</span><span class=p>,</span>
<a id=__codelineno-27-3 name=__codelineno-27-3 href=#__codelineno-27-3></a><span class=w>    </span><span class=s>'1'</span><span class=o>=</span><span class=s>'SpermatogoniaA'</span><span class=p>,</span>
<a id=__codelineno-27-4 name=__codelineno-27-4 href=#__codelineno-27-4></a><span class=w>    </span><span class=s>'2'</span><span class=o>=</span><span class=s>'RoundSpermatids'</span><span class=p>,</span>
<a id=__codelineno-27-5 name=__codelineno-27-5 href=#__codelineno-27-5></a><span class=w>    </span><span class=s>'3'</span><span class=o>=</span><span class=s>'RoundSpermatids'</span><span class=p>,</span>
<a id=__codelineno-27-6 name=__codelineno-27-6 href=#__codelineno-27-6></a><span class=w>    </span><span class=s>'4'</span><span class=o>=</span><span class=s>'RoundSpermatids'</span><span class=p>,</span>
<a id=__codelineno-27-7 name=__codelineno-27-7 href=#__codelineno-27-7></a><span class=w>    </span><span class=s>'5'</span><span class=o>=</span><span class=s>'ElongSpermatids'</span><span class=p>,</span>
<a id=__codelineno-27-8 name=__codelineno-27-8 href=#__codelineno-27-8></a><span class=w>    </span><span class=s>'6'</span><span class=o>=</span><span class=s>'SpermatocitesI'</span><span class=p>,</span>
<a id=__codelineno-27-9 name=__codelineno-27-9 href=#__codelineno-27-9></a><span class=w>    </span><span class=s>'7'</span><span class=o>=</span><span class=s>'RoundSpermatids'</span><span class=p>,</span>
<a id=__codelineno-27-10 name=__codelineno-27-10 href=#__codelineno-27-10></a><span class=w>    </span><span class=s>'8'</span><span class=o>=</span><span class=s>'ElongSpermatids'</span><span class=p>,</span>
<a id=__codelineno-27-11 name=__codelineno-27-11 href=#__codelineno-27-11></a><span class=w>    </span><span class=s>'9'</span><span class=o>=</span><span class=s>'SpermatogoniaB'</span><span class=p>,</span>
<a id=__codelineno-27-12 name=__codelineno-27-12 href=#__codelineno-27-12></a><span class=w>    </span><span class=s>'10'</span><span class=o>=</span><span class=s>'Somatic'</span><span class=p>,</span>
<a id=__codelineno-27-13 name=__codelineno-27-13 href=#__codelineno-27-13></a><span class=w>    </span><span class=s>'11'</span><span class=o>=</span><span class=s>'SpermatocitesII'</span>
<a id=__codelineno-27-14 name=__codelineno-27-14 href=#__codelineno-27-14></a><span class=p>)</span>
</code></pre></div> <p>Now we apply the new names and save them in a new meta.data variable called "cluster_names":</p> <div class=highlight><pre><span></span><code><a id=__codelineno-28-1 name=__codelineno-28-1 href=#__codelineno-28-1></a><span class=n>sample.combined.sct</span><span class=o>$</span><span class=n>cluster_names</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>recode</span><span class=p>(</span><span class=nf>Idents</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>),</span><span class=w> </span><span class=o>!!!</span><span class=n>new_names</span><span class=p>)</span>
<a id=__codelineno-28-2 name=__codelineno-28-2 href=#__codelineno-28-2></a>
<a id=__codelineno-28-3 name=__codelineno-28-3 href=#__codelineno-28-3></a><span class=nf>Idents</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>)</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=s>"cluster_names"</span>
</code></pre></div> <p>Finally, we visualize the new cluster names!</p> <div class=highlight><pre><span></span><code><a id=__codelineno-29-1 name=__codelineno-29-1 href=#__codelineno-29-1></a><span class=nf>DimPlot</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"umap"</span><span class=p>,</span><span class=w> </span><span class=n>label</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span>
<a id=__codelineno-29-2 name=__codelineno-29-2 href=#__codelineno-29-2></a><span class=nf>DoHeatmap</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>unlist</span><span class=p>(</span><span class=n>markers</span><span class=p>))</span>
</code></pre></div> <h2 id=selecting-and-subclustering>Selecting and subclustering<a class=headerlink href=#selecting-and-subclustering title="Permanent link">¶</a></h2> <p>If we suspect that a cluster might be composed of an heterogeneous population of cells, we can subset that group of cells and find subclusters of cells. This is the case for the cell types found in late spermatogonia and spermatocites.</p> <h3 id=subsetting-the-data>Subsetting the data<a class=headerlink href=#subsetting-the-data title="Permanent link">¶</a></h3> <p>We can subset our Seurat object using the <code>subset()</code> function and selecting our clusters of interest from the <code>idents</code> argument. Note that you have to have selected our "cluster_names" as your <code>Ident()</code>!</p> <div class=highlight><pre><span></span><code><a id=__codelineno-30-1 name=__codelineno-30-1 href=#__codelineno-30-1></a><span class=c1># Clusters of interest and their markers</span>
<a id=__codelineno-30-2 name=__codelineno-30-2 href=#__codelineno-30-2></a><span class=n>clusters</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'SpermatogoniaB'</span><span class=p>,</span><span class=s>'SpermatocitesI'</span><span class=p>,</span><span class=s>'SpermatocitesII'</span><span class=p>)</span>
<a id=__codelineno-30-3 name=__codelineno-30-3 href=#__codelineno-30-3></a><span class=n>sample_subset</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>subset</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>idents</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>clusters</span><span class=p>)</span>
<a id=__codelineno-30-4 name=__codelineno-30-4 href=#__codelineno-30-4></a>
<a id=__codelineno-30-5 name=__codelineno-30-5 href=#__codelineno-30-5></a><span class=n>markers</span><span class=p>[</span><span class=s>'Leptotene'</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'SYCE2'</span><span class=p>,</span><span class=s>'SCML1'</span><span class=p>)</span>
<a id=__codelineno-30-6 name=__codelineno-30-6 href=#__codelineno-30-6></a><span class=n>markers</span><span class=p>[</span><span class=s>'Zygotene'</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'LY6K'</span><span class=p>,</span><span class=w> </span><span class=s>'SYCP1'</span><span class=p>)</span>
<a id=__codelineno-30-7 name=__codelineno-30-7 href=#__codelineno-30-7></a><span class=n>markers</span><span class=p>[</span><span class=s>'Pachytene'</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'PIWIL1'</span><span class=p>,</span><span class=s>'CCDC112'</span><span class=p>)</span>
<a id=__codelineno-30-8 name=__codelineno-30-8 href=#__codelineno-30-8></a><span class=n>markers</span><span class=p>[</span><span class=s>'Diplotene'</span><span class=p>]</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>'OVOL2'</span><span class=p>,</span><span class=s>'CCNA1'</span><span class=p>,</span><span class=w> </span><span class=s>'CDK1'</span><span class=p>,</span><span class=s>'AURKA'</span><span class=p>)</span>
</code></pre></div> <p>Let’s look at the marker plots of these specific clusters:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-31-1 name=__codelineno-31-1 href=#__codelineno-31-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>sample_subset</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=s>'Leptotene'</span><span class=p>]</span><span class=w> </span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-32-1 name=__codelineno-32-1 href=#__codelineno-32-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>sample_subset</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=s>'Zygotene'</span><span class=p>]</span><span class=w> </span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-33-1 name=__codelineno-33-1 href=#__codelineno-33-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>sample_subset</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=s>'Pachytene'</span><span class=p>]</span><span class=w> </span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-34-1 name=__codelineno-34-1 href=#__codelineno-34-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>sample_subset</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>markers</span><span class=p>[</span><span class=s>'Diplotene'</span><span class=p>]</span><span class=w> </span><span class=p>)</span>
</code></pre></div> <h3 id=recalculate-clusters>Recalculate clusters<a class=headerlink href=#recalculate-clusters title="Permanent link">¶</a></h3> <p>Now we can recalculate our neighbours and the clusters. What should be the <code>resolution</code> argument?</p> <div class=highlight><pre><span></span><code><a id=__codelineno-35-1 name=__codelineno-35-1 href=#__codelineno-35-1></a><span class=n>sample_subset</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindNeighbors</span><span class=p>(</span><span class=n>sample_subset</span><span class=p>)</span>
<a id=__codelineno-35-2 name=__codelineno-35-2 href=#__codelineno-35-2></a><span class=n>sample_subset</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindClusters</span><span class=p>(</span><span class=n>sample_subset</span><span class=p>,</span><span class=w> </span><span class=n>resolution</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.4</span><span class=p>)</span>
</code></pre></div> <p>And we can visualize our subclustering using <code>DimPlot()</code>:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-36-1 name=__codelineno-36-1 href=#__codelineno-36-1></a><span class=nf>DimPlot</span><span class=p>(</span><span class=n>sample_subset</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"umap"</span><span class=p>)</span>
</code></pre></div> <p>It would be more convenient to rename our clusters. We should be able to do it using the same code snippet as before!</p> <div class=highlight><pre><span></span><code><a id=__codelineno-37-1 name=__codelineno-37-1 href=#__codelineno-37-1></a><span class=n>new_names</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span>
<a id=__codelineno-37-2 name=__codelineno-37-2 href=#__codelineno-37-2></a><span class=w>    </span><span class=s>'0'</span><span class=o>=</span><span class=s>'SpermatogoniaB'</span><span class=p>,</span>
<a id=__codelineno-37-3 name=__codelineno-37-3 href=#__codelineno-37-3></a><span class=w>    </span><span class=s>'1'</span><span class=o>=</span><span class=s>'Zygotene'</span><span class=p>,</span>
<a id=__codelineno-37-4 name=__codelineno-37-4 href=#__codelineno-37-4></a><span class=w>    </span><span class=s>'2'</span><span class=o>=</span><span class=s>'Diplotene'</span><span class=p>,</span>
<a id=__codelineno-37-5 name=__codelineno-37-5 href=#__codelineno-37-5></a><span class=w>    </span><span class=s>'3'</span><span class=o>=</span><span class=s>'Diplotene'</span><span class=p>,</span>
<a id=__codelineno-37-6 name=__codelineno-37-6 href=#__codelineno-37-6></a><span class=w>    </span><span class=s>'4'</span><span class=o>=</span><span class=s>'Diplotene'</span><span class=p>,</span>
<a id=__codelineno-37-7 name=__codelineno-37-7 href=#__codelineno-37-7></a><span class=w>    </span><span class=s>'5'</span><span class=o>=</span><span class=s>'Leptotene'</span><span class=p>,</span>
<a id=__codelineno-37-8 name=__codelineno-37-8 href=#__codelineno-37-8></a><span class=w>    </span><span class=s>'6'</span><span class=o>=</span><span class=s>'Pachytene'</span><span class=p>,</span>
<a id=__codelineno-37-9 name=__codelineno-37-9 href=#__codelineno-37-9></a><span class=w>    </span><span class=s>'7'</span><span class=o>=</span><span class=s>'Pachytene'</span><span class=p>,</span>
<a id=__codelineno-37-10 name=__codelineno-37-10 href=#__codelineno-37-10></a><span class=w>    </span><span class=s>'8'</span><span class=o>=</span><span class=s>'Diplotene'</span><span class=p>)</span>
<a id=__codelineno-37-11 name=__codelineno-37-11 href=#__codelineno-37-11></a>
<a id=__codelineno-37-12 name=__codelineno-37-12 href=#__codelineno-37-12></a><span class=n>sample_subset</span><span class=o>$</span><span class=n>cluster_names</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>recode</span><span class=p>(</span><span class=nf>Idents</span><span class=p>(</span><span class=n>sample_subset</span><span class=p>),</span><span class=w> </span><span class=o>!!!</span><span class=n>new_names</span><span class=p>)</span>
</code></pre></div> <h3 id=differential-expression>Differential expression<a class=headerlink href=#differential-expression title="Permanent link">¶</a></h3> <p>We can find specific differences between the subclusters found in here:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-38-1 name=__codelineno-38-1 href=#__codelineno-38-1></a><span class=n>subset_markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindAllMarkers</span><span class=p>(</span><span class=n>sample_subset</span><span class=p>,</span><span class=w> </span><span class=n>logfc.threshold</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.25</span><span class=p>,</span><span class=w> </span><span class=n>only.pos</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span>
<a id=__codelineno-38-2 name=__codelineno-38-2 href=#__codelineno-38-2></a>
<a id=__codelineno-38-3 name=__codelineno-38-3 href=#__codelineno-38-3></a><span class=n>subset_markers</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-38-4 name=__codelineno-38-4 href=#__codelineno-38-4></a><span class=nf>group_by</span><span class=p>(</span><span class=n>cluster</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-38-5 name=__codelineno-38-5 href=#__codelineno-38-5></a><span class=w>    </span><span class=nf>top_n</span><span class=p>(</span><span class=n>n</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>10</span><span class=p>,</span><span class=w> </span><span class=n>wt</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>avg_log2FC</span><span class=p>)</span><span class=w> </span><span class=o>-&gt;</span><span class=w> </span><span class=n>subset_top10</span>
<a id=__codelineno-38-6 name=__codelineno-38-6 href=#__codelineno-38-6></a>
<a id=__codelineno-38-7 name=__codelineno-38-7 href=#__codelineno-38-7></a><span class=nf>head</span><span class=p>(</span><span class=n>subset_top10</span><span class=p>)</span>
</code></pre></div> <p>We should be able to find at least some of the markers used in the plots. However, it can be that those do not appear because there are many other coexpressed genes with high expression values.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-39-1 name=__codelineno-39-1 href=#__codelineno-39-1></a><span class=nf>DoHeatmap</span><span class=p>(</span><span class=n>sample_subset</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>subset_top10</span><span class=o>$</span><span class=n>gene</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>NoLegend</span><span class=p>()</span>
</code></pre></div> <p>We can find some of the marker genes. For example <code>SCML1</code> for leptotene cells, <code>SYCP1</code> and <code>LY6K</code> for zygotene, <code>CCDC112</code> and <code>PIWIL1</code> for pachitene, <code>CCNA1</code> and <code>AURKA</code> for zygotene.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-40-1 name=__codelineno-40-1 href=#__codelineno-40-1></a><span class=nf>DoHeatmap</span><span class=p>(</span><span class=n>sample_subset</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>unlist</span><span class=p>(</span><span class=n>markers</span><span class=p>[</span><span class=nf>c</span><span class=p>(</span><span class=s>"Leptotene"</span><span class=p>,</span><span class=s>"Zygotene"</span><span class=p>,</span><span class=s>"Pachytene"</span><span class=p>,</span><span class=s>"Diplotene"</span><span class=p>)]))</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>NoLegend</span><span class=p>()</span>
</code></pre></div> <h3 id=reintegrating-the-subclusters>Reintegrating the subclusters<a class=headerlink href=#reintegrating-the-subclusters title="Permanent link">¶</a></h3> <p>Since we have created a subset of the original data, how can we integrate the new subclusters back into the dataset? We will need to do some data wrangling</p> <p>First we create a new metadata variable in the original Seurat object.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-41-1 name=__codelineno-41-1 href=#__codelineno-41-1></a><span class=n>sample.combined.sct</span><span class=o>$</span><span class=n>subclusters</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>sample.combined.sct</span><span class=o>$</span><span class=n>cluster_names</span>
</code></pre></div> <p>The new subclusters are found in the subset metadata. We only need to get the cell names and change the original dataset accordingly.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-42-1 name=__codelineno-42-1 href=#__codelineno-42-1></a><span class=n>subset_cells</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>rownames</span><span class=p>(</span><span class=n>sample_subset</span><span class=o>@</span><span class=n>meta.data</span><span class=p>)</span>
<a id=__codelineno-42-2 name=__codelineno-42-2 href=#__codelineno-42-2></a><span class=n>sample.combined.sct</span><span class=o>$</span><span class=n>subclusters</span><span class=o>&lt;-</span><span class=w> </span><span class=n>sample_subset</span><span class=o>$</span><span class=n>seurat_clusters</span>
</code></pre></div> <p>We can visualize the before and after very easily now:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-43-1 name=__codelineno-43-1 href=#__codelineno-43-1></a><span class=n>p1</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DimPlot</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"umap"</span><span class=p>,</span><span class=w> </span><span class=n>group.by</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"cluster_names"</span><span class=p>,</span><span class=w> </span><span class=n>label</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>T</span><span class=p>)</span>
<a id=__codelineno-43-2 name=__codelineno-43-2 href=#__codelineno-43-2></a><span class=n>p2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DimPlot</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"umap"</span><span class=p>,</span><span class=w> </span><span class=n>group.by</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"subclusters"</span><span class=p>,</span><span class=w> </span><span class=n>label</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>T</span><span class=p>)</span>
<a id=__codelineno-43-3 name=__codelineno-43-3 href=#__codelineno-43-3></a>
<a id=__codelineno-43-4 name=__codelineno-43-4 href=#__codelineno-43-4></a><span class=n>p1</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=n>p2</span>
</code></pre></div> <h3 id=cluster-correlations>Cluster correlations<a class=headerlink href=#cluster-correlations title="Permanent link">¶</a></h3> <p>An useful method we can use to check the quality of our clusters is to compute cluster correlations. Unfortunately, Seurat does not provide a straight forward method, but it is fairly simple to calculate. First we make use of the <code>AverageExpression()</code> function, which, as the name implies, will compute the average expression of the groups of cells (clusters or other types of groups).</p> <div class=highlight><pre><span></span><code><a id=__codelineno-44-1 name=__codelineno-44-1 href=#__codelineno-44-1></a><span class=n>av.exp</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>AverageExpression</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>assay</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"SCT"</span><span class=p>,</span><span class=w> </span><span class=n>slot</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"data"</span><span class=p>)</span>
<a id=__codelineno-44-2 name=__codelineno-44-2 href=#__codelineno-44-2></a>
<a id=__codelineno-44-3 name=__codelineno-44-3 href=#__codelineno-44-3></a><span class=nf>head</span><span class=p>(</span><span class=n>av.exp</span><span class=p>)</span>
</code></pre></div> <p>Then we calculate the correlations between the columns (clusters) and plot it using the <code>pheatmap()</code> package.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-45-1 name=__codelineno-45-1 href=#__codelineno-45-1></a><span class=nf>pheatmap</span><span class=p>(</span><span class=nf>cor</span><span class=p>(</span><span class=n>av.exp</span><span class=p>),</span><span class=w> </span><span class=n>treeheight_col</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0</span><span class=p>,</span><span class=w> </span><span class=n>breaks</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>seq</span><span class=p>(</span><span class=m>-1</span><span class=p>,</span><span class=m>1</span><span class=p>,</span><span class=n>length.out</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>100</span><span class=p>),</span><span class=w> </span><span class=n>display_numbers</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>T</span><span class=p>)</span>
</code></pre></div> <h2 id=save-our-clustered-data>Save our clustered data<a class=headerlink href=#save-our-clustered-data title="Permanent link">¶</a></h2> <div class=highlight><pre><span></span><code><a id=__codelineno-46-1 name=__codelineno-46-1 href=#__codelineno-46-1></a><span class=nf>SaveH5Seurat</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>filename</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"../../Data/notebooks_data/sample_123.filt.norm.red.clust.h5Seurat"</span><span class=p>)</span>
</code></pre></div> <h2 id=wrapping-up>Wrapping up<a class=headerlink href=#wrapping-up title="Permanent link">¶</a></h2> <p>We have been showing how to identify potential cell clusters. In this dataset, cells change from one type to another in a continuous process, so such a hard clustering may not completely reflect biological reality. However, it is a good approximation, as it is illustrated by the differentially expressed genes that we checked in each cluster.</p> <p>We performed differential expression and what are the useful values that we get from it (p-value of the test, magnitude of the gene expression compared to all other clusters). Finally, we subsetted the data and subject it to a more fine grained cell identification.</p></div> </article> </div> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=Part04_integration.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Integration" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Integration </div> </div> </a> <a href=Part06_pseudotime.html class="md-footer__link md-footer__link--next" aria-label="Next: Pseudotimes" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Pseudotimes </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> CC-BY-SA 4.0 license </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox/ target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "..", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.footer", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=../assets/javascripts/bundle.19047be9.min.js></script> <script src=../javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> <script src=../javascripts/embed-amd.js></script> </body> </html>