<!doctype html><html lang=en class=no-js> <head><script type=application/ld+json>
    {
      "@context": "https://schema.org",
      "@id": "https://hds-sandbox.github.io/scRNASeq_course/",
      "@type": "LearningResource",
      "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
      "description": "This workshop contains a basic tutorial on how to approach scRNAseq experiments, starting from fastq files out of the sequencer. Thus, the workshop does not include any information of laboratory protocols, library preparation or any wet-lab related procedures.",
      "keywords": [
        "bioinformatics, RNAseq, single cell, NGS, python, R"
      ],
      "learningResourceType": [
        "e-learning"
      ],
      "license": [
        {
          "@id": "https://creativecommons.org/licenses/by-sa/4.0/legalcode",
          "@type": "CreativeWork"
        }
      ],
      "mentions": [
        {
          "@id": "10.5281/zenodo.7660225",
          "@type": "Thing"
        },
        {
          "@id": "10.5281/zenodo.7565997",
          "@type": "Thing"
        }
      ],
      "name": "Introduction to single cell RNAseq analysis workshop",
      "educationalLevel": "Beginner",
      "audience": "PhD, postdocs, MSc, Researchers"
    }
    </script><title>Datasets integration - Single cell sequencing data analysis - an introductiory but comprehensive course</title><link rel=stylesheet href=../assets/stylesheets/main.240905d7.min.css><link rel=stylesheet href=../assets/stylesheets/palette.a0c5b2b5.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../stylesheets/neoteroi-mkdocs.css><link rel=stylesheet href=../stylesheets/extra.css><script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#integration-of-datasets class=md-skip> Skip to content </a> </div> <div data-md-component=announce> <aside class=md-banner> <div class="md-banner__inner md-grid md-typeset"> Page under construction - course not finalized </div> </aside> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Single cell sequencing data analysis - an introductiory but comprehensive course" class="md-header__button md-logo" aria-label="Single cell sequencing data analysis - an introductiory but comprehensive course" data-md-component=logo> <img src=../img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Single cell sequencing data analysis - an introductiory but comprehensive course </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Datasets integration </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/scRNASeq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/scRNASeq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=../python/installation.html class=md-tabs__link> Python </a> </li> <li class=md-tabs__item> <a href=Part01_read_the_data.html class="md-tabs__link md-tabs__link--active"> R </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Single cell sequencing data analysis - an introductiory but comprehensive course" class="md-nav__button md-logo" aria-label="Single cell sequencing data analysis - an introductiory but comprehensive course" data-md-component=logo> <img src=../img/logo.png alt=logo> </a> Single cell sequencing data analysis - an introductiory but comprehensive course </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/scRNASeq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/scRNASeq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../index.html class=md-nav__link> Course Introduction </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> Python <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Python </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../python/installation.html class=md-nav__link> Setup and installation </a> </li> <li class=md-nav__item> <a href=../python/Part01_read_the_data.html class=md-nav__link> Read the data </a> </li> <li class=md-nav__item> <a href=../python/Part02_filtering_sample2.html class=md-nav__link> Quality Control (QC) and filtering </a> </li> <li class=md-nav__item> <a href=../python/Part03_normalize_and_integrate.html class=md-nav__link> Normalize and Integrate </a> </li> <li class=md-nav__item> <a href=../python/Part04_clustering.html class=md-nav__link> Cell clustering and Differential Expression (DE) </a> </li> <li class=md-nav__item> <a href=../python/Part05_cell_fate.html class=md-nav__link> Pseudotimes and cell fates </a> </li> <li class=md-nav__item> <a href=../python/Part06_Condition_analysis.html class=md-nav__link> Cross-data analysis </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8> <label class=md-nav__link for=__nav_2_8 id=__nav_2_8_label tabindex=0> Extra <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_8_label aria-expanded=false> <label class=md-nav__title for=__nav_2_8> <span class="md-nav__icon md-icon"></span> Extra </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../python/Part02_filtering_sample3.html class=md-nav__link> Filtering another sample </a> </li> <li class=md-nav__item> <a href=../python/Part02_filtering_sample1.html class=md-nav__link> Filtering a low quality sample </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3 checked> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> R <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=true> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> R </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=Part01_read_the_data.html class=md-nav__link> Read the data </a> </li> <li class=md-nav__item> <a href=Part02_filtering_sample2.html class=md-nav__link> Preprocessing </a> </li> <li class=md-nav__item> <a href=Part03_normalize_and_dim_reduc.html class=md-nav__link> Normalization </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <a href=Part04_integration.html class="md-nav__link md-nav__link--active"> Integration </a> </li> <li class=md-nav__item> <a href=Part05_clustering.html class=md-nav__link> Clustering and Differential Expression </a> </li> <li class=md-nav__item> <a href=Part06_pseudotime.html class=md-nav__link> Pseudotimes </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3_7> <label class=md-nav__link for=__nav_3_7 id=__nav_3_7_label tabindex=0> Extra <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_3_7_label aria-expanded=false> <label class=md-nav__title for=__nav_3_7> <span class="md-nav__icon md-icon"></span> Extra </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=Part02_filtering_sample3.html class=md-nav__link> Preprocessing another sample </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <div><h1 id=integration-of-datasets>Integration of datasets<a class=headerlink href=#integration-of-datasets title="Permanent link">Â¶</a></h1> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>ðŸ•° <strong>Time Estimation:</strong> 40 minutes if using at least 2 cores (that is, at least 16 virtual-CPUs). Expect longer time with slower hardware, and a RAM usage of some 40-100GB.</p> <p>ðŸ’¬ <strong>Learning Objectives:</strong> </p> <ol> <li>Create an 'integrated' data assay for downstream analysis</li> <li>Identify cell types that are present in both datasets</li> <li>Assess integration success.</li> </ol> </div> <p>The joint analysis of two or more single-cell datasets poses unique challenges. In particular, identifying cell populations that are present across multiple datasets can be problematic under standard workflows. Seurat v4 includes a set of methods to match (or â€˜alignâ€™) shared cell populations across datasets. These methods first identify cross-dataset pairs of cells that are in a matched biological state (â€˜anchorsâ€™), can be used both to correct for technical differences between datasets (i.e.Â batch effect correction), and to perform comparative scRNA-seq analysis of across experimental conditions.</p> <p><img alt src=img/Part04_integration/integration_explained.jpeg> !!! info "Figure caption"</p> <div class=highlight><pre><span></span><code>(A) Representation of two datasets, reference and query, each of which originates from a separate single-cell experiment. The two datasets share cells from similar biological states, but the query dataset contains a unique population (in black).

(B) We perform canonical correlation analysis, followed by L2 normalization of the canonical correlation vectors, to project the datasets into a subspace defined by shared correlation structure across datasets.

(C) In the shared space, we identify pairs of MNNs across reference and query cells. These should represent cells in a shared biological state across datasets (gray lines) and serve as anchors to guide dataset integration. In principle, cells in unique populations should not participate in anchors, but in practice, we observe "incorrect" anchors at low frequency (red lines).

(D) For each anchor pair, we assign a score based on the consistency of anchors across the neighborhood structure of each dataset.

(E) We utilize anchors and their scores to compute "correction" vectors for each query cell, transforming its expression so it can be jointly analyzed as part of an integrated reference.
</code></pre></div> <p>Below, we demonstrate methods for scRNA-seq integration as described in <a href=https://www.cell.com/cell/fulltext/S0092-8674(19)30559-8>Stuart*, Butler* et al, 2019</a> to perform a comparative analysis of human immune cells (PBMC) in either a <a href=https://www.nature.com/articles/nbt.4042>resting or interferon-stimulated state</a>.</p> <h1 id=setup>Setup<a class=headerlink href=#setup title="Permanent link">Â¶</a></h1> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=nf>library</span><span class=p>(</span><span class=n>tidyverse</span><span class=p>)</span>
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a><span class=nf>library</span><span class=p>(</span><span class=n>patchwork</span><span class=p>)</span>
<a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a><span class=nf>library</span><span class=p>(</span><span class=n>Seurat</span><span class=p>)</span>
<a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a><span class=nf>library</span><span class=p>(</span><span class=n>SeuratDisk</span><span class=p>)</span>
<a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a><span class=nf>library</span><span class=p>(</span><span class=n>sctransform</span><span class=p>)</span>
<a id=__codelineno-0-6 name=__codelineno-0-6 href=#__codelineno-0-6></a><span class=nf>library</span><span class=p>(</span><span class=n>kBET</span><span class=p>)</span>
<a id=__codelineno-0-7 name=__codelineno-0-7 href=#__codelineno-0-7></a>
<a id=__codelineno-0-8 name=__codelineno-0-8 href=#__codelineno-0-8></a><span class=nf>LoadH5Seurat</span><span class=p>(</span><span class=s>"../../Data/notebooks_data/sample_2.filt.norm.h5Seurat"</span><span class=p>)</span>
<a id=__codelineno-0-9 name=__codelineno-0-9 href=#__codelineno-0-9></a><span class=nf>LoadH5Seurat</span><span class=p>(</span><span class=s>"../../Data/notebooks_data/sample_3.filt.norm.h5Seurat"</span><span class=p>)</span>
</code></pre></div> <p>First of all, we need to make a list of datasets.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=n>sample.list</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>list</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>sample_3</span><span class=p>)</span>
</code></pre></div> <div class="admonition warning"> <p class=admonition-title>Warning</p> <p>It is extremely important that the datasets have been filtered and normalized separately before integration. Otherwise we will be introducing artifacts to the datasets that will be very undesirable.</p> </div> <h2 id=default-integration>Default integration<a class=headerlink href=#default-integration title="Permanent link">Â¶</a></h2> <p>Seurat default integration workflow uses two algorithms to merge datasets: canonical correlation analysis and mutual nearest neighbours. The former, Canonical correlation analysis (CCA), is an algorithm that enables the identification of shared sources of variation between datasets, while the latter is able to identify biologically matched cells in a pair of datasets. Thus, CCA-based integration therefore enables integrative analysis when experimental conditions or disease states introduce very strong expression shifts, or when integrating datasets across modalities and species. We will see in this section how to perform CCA and integrate our datasets.</p> <h3 id=feature-selection>Feature selection<a class=headerlink href=#feature-selection title="Permanent link">Â¶</a></h3> <p>Before integrating the datasets, we will select features that are repeatedly variable across datasets for integration. As discussed in the previous lesson, we typically use 3,000 or more features for analyses downstream of <code>SCTransform</code>.</p> <p>Then we will use the function<code>PrepSCTIntegration()</code>. This function takes in our list of datasets normalized with <code>SCTransform</code> and performs the following steps:</p> <ul> <li>Ensures that the sctransform residuals for the features specified to anchor.features are present in each object in the list. This is necessary because the default behavior of <code>SCTransform</code> is to only store the residuals for the features determined to be variable.</li> <li>Subsets the <code>scale.data</code> slot to only contain the residuals for <code>anchor.features</code> for efficiency in downstream processing.</li> </ul> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=n>features</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>SelectIntegrationFeatures</span><span class=p>(</span><span class=n>object.list</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample.list</span><span class=p>,</span><span class=w> </span><span class=n>nfeatures</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>3000</span><span class=p>)</span>
<a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a><span class=n>sample.list</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>PrepSCTIntegration</span><span class=p>(</span><span class=n>object.list</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample.list</span><span class=p>,</span><span class=w> </span><span class=n>anchor.features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>features</span><span class=p>)</span>
</code></pre></div> <h3 id=selecting-anchors>Selecting anchors<a class=headerlink href=#selecting-anchors title="Permanent link">Â¶</a></h3> <p>Through the identification of cell pairwise correspondences between single cells across datasets, termed <strong>"anchors"</strong>, Seurat can transform datasets into a shared space, even in the presence of extensive technical and/or biological differences. This enables the construction of harmonized atlases at the tissue or organism scale. To identify anchors, we use the <code>FindIntegrationAnchors()</code> function, which takes our list of Seurat objects as input, and then we use these anchors to integrate the two datasets together with <code>IntegrateData()</code>.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=n>sample.anchors</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindIntegrationAnchors</span><span class=p>(</span><span class=n>object.list</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample.list</span><span class=p>,</span><span class=w> </span><span class=n>normalization.method</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>'SCT'</span><span class=p>,</span><span class=w> </span><span class=n>anchor.features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>features</span><span class=p>)</span>
<a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a><span class=n>sample.combined.sct</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>IntegrateData</span><span class=p>(</span><span class=n>anchorset</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample.anchors</span><span class=p>,</span><span class=w> </span><span class=n>normalization.method</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>'SCT'</span><span class=p>)</span>
</code></pre></div> <h3 id=combined-dimensionality-reduction>Combined dimensionality reduction<a class=headerlink href=#combined-dimensionality-reduction title="Permanent link">Â¶</a></h3> <p>Now we can continue with our PCA, select an adequate number of PCs (remember that we use an increased number of PCs due to sctransform!).</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=n>sample.combined.sct</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>RunPCA</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>verbose</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a><span class=n>sample.combined.sct</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>RunUMAP</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"pca"</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>30</span><span class=p>)</span>
</code></pre></div> <p>And finally we can visualize the integrated dataset.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=nf>DimPlot</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"umap"</span><span class=p>)</span>
</code></pre></div> <p>We should save our integrated object for further analysis.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=nf>SaveH5Seurat</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>filename</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"../../Data/notebooks_data/sample_123.filt.norm.red.h5Seurat"</span><span class=p>)</span>
</code></pre></div> <h2 id=reciprocal-pca>Reciprocal PCA<a class=headerlink href=#reciprocal-pca title="Permanent link">Â¶</a></h2> <p>CCA is well-suited for identifying anchors when cell types are conserved. However, CCA-based integration may also lead to overcorrection, especially when a large proportion of cells are non-overlapping across datasets, and it is very time-consuming for large datasets. On the other hand, RPCA-based integration runs significantly faster, and also represents a more conservative approach where cells in different biological states are less likely to â€˜alignâ€™ after integration. Therefore, it is recommend to use RPCA during integrative analysis where:</p> <ul> <li>A substantial fraction of cells in one dataset have no matching type in the other</li> <li>Datasets originate from the same platform (i.e.Â multiple lanes of 10x genomics)</li> <li>There are a large number of datasets or cells to integrate</li> </ul> <p>RPCA works by determining anchors between any two datasets and project each dataset into the others PCA space. Then it constrains the anchors by the same mutual neighborhood requirement as CCA. While the list of commands is nearly identical, this workflow requires users to run principal components analysis (PCA) -<code>RunPCA()</code> individually on each dataset prior to integration. Users should also set the <code>reduction</code> argument to â€˜rpcaâ€™, when running <code>FindIntegrationAnchors()</code>.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=c1># Reload the dataset list</span>
<a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a><span class=n>sample.list</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>list</span><span class=p>(</span><span class=n>sample_2</span><span class=p>,</span><span class=w> </span><span class=n>sample_3</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=n>features</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>SelectIntegrationFeatures</span><span class=p>(</span><span class=n>object.list</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample.list</span><span class=p>,</span><span class=w> </span><span class=n>nfeatures</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>3000</span><span class=p>)</span>
<a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a><span class=n>sample.list</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>PrepSCTIntegration</span><span class=p>(</span><span class=n>object.list</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample.list</span><span class=p>,</span><span class=w> </span><span class=n>anchor.features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>features</span><span class=p>)</span>
<a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a><span class=n>sample.list</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>lapply</span><span class=p>(</span><span class=n>X</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample.list</span><span class=p>,</span><span class=w> </span><span class=n>FUN</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>RunPCA</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>features</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=n>sample.anchors</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindIntegrationAnchors</span><span class=p>(</span><span class=n>object.list</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample.list</span><span class=p>,</span><span class=w> </span><span class=n>normalization.method</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"SCT"</span><span class=p>,</span>
<a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a><span class=w>    </span><span class=n>anchor.features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>features</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>30</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"rpca"</span><span class=p>,</span><span class=w> </span><span class=n>k.anchor</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>20</span><span class=p>)</span>
<a id=__codelineno-9-3 name=__codelineno-9-3 href=#__codelineno-9-3></a><span class=n>sample.combined.sct</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>IntegrateData</span><span class=p>(</span><span class=n>anchorset</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>sample.anchors</span><span class=p>,</span><span class=w> </span><span class=n>normalization.method</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"SCT"</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>30</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=n>sample.combined.sct</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>RunPCA</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>verbose</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
<a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a><span class=n>sample.combined.sct</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>RunUMAP</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"pca"</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>30</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=c1># Visualization</span>
<a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a><span class=nf>DimPlot</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>"umap"</span><span class=p>)</span>
</code></pre></div> <p>In general, we observe strikingly similar results between the standard workflow and the one demonstrated here, with substantial reduction in compute time and memory. However, if the datasets are highly divergent (for example, cross-modality mapping or cross-species mapping), where only a small subset of features can be used to facilitate integration, and you may observe superior results using CCA.</p> <h2 id=assessing-integration-quality>Assessing integration quality<a class=headerlink href=#assessing-integration-quality title="Permanent link">Â¶</a></h2> <p>How well went the integration? We can see if, for every point in the UMAP plot, the other most similar points are a mix from the various samples, or if they are dominated by only one of them. The <code>R</code> package <code>kbet</code> does this test for each datapoint. The test is rejected if a datapoint is in an area where cells from all samples are not mixed/overlapping. A perfect integration has around zero rejection (left side of the plot). In reality, this does not happen, but a good integration stays in general below a 50% rejection (right side of the plot).</p> <div class=highlight><pre><span></span><code><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=n>batch.estimate</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>kBET</span><span class=p>(</span><span class=n>sample.combined.sct</span><span class=p>[[</span><span class=s>"SCT"</span><span class=p>]]</span><span class=o>@</span><span class=n>data</span><span class=p>,</span><span class=w> </span><span class=n>sample.combined.sct</span><span class=o>$</span><span class=n>batch</span><span class=p>,</span><span class=w> </span><span class=n>plot</span><span class=o>=</span><span class=kc>FALSE</span><span class=p>)</span>
<a id=__codelineno-12-2 name=__codelineno-12-2 href=#__codelineno-12-2></a><span class=n>plot.data</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>data.frame</span><span class=p>(</span><span class=n>class</span><span class=o>=</span><span class=nf>rep</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=s>'observed'</span><span class=p>,</span><span class=w> </span><span class=s>'expected'</span><span class=p>),</span><span class=w> </span>
<a id=__codelineno-12-3 name=__codelineno-12-3 href=#__codelineno-12-3></a><span class=w>                                  </span><span class=n>each</span><span class=o>=</span><span class=nf>length</span><span class=p>(</span><span class=n>batch.estimate</span><span class=o>$</span><span class=n>stats</span><span class=o>$</span><span class=n>kBET.observed</span><span class=p>)),</span><span class=w> </span>
<a id=__codelineno-12-4 name=__codelineno-12-4 href=#__codelineno-12-4></a><span class=w>                        </span><span class=n>data</span><span class=w> </span><span class=o>=</span><span class=w>  </span><span class=nf>c</span><span class=p>(</span><span class=n>batch.estimate</span><span class=o>$</span><span class=n>stats</span><span class=o>$</span><span class=n>kBET.observed</span><span class=p>,</span>
<a id=__codelineno-12-5 name=__codelineno-12-5 href=#__codelineno-12-5></a><span class=w>                                  </span><span class=n>batch.estimate</span><span class=o>$</span><span class=n>stats</span><span class=o>$</span><span class=n>kBET.expected</span><span class=p>))</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=nf>ggplot</span><span class=p>(</span><span class=n>plot.data</span><span class=p>,</span><span class=w> </span><span class=nf>aes</span><span class=p>(</span><span class=n>class</span><span class=p>,</span><span class=w> </span><span class=n>data</span><span class=p>))</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>geom_boxplot</span><span class=p>()</span><span class=w> </span><span class=o>+</span><span class=w> </span>
<a id=__codelineno-13-2 name=__codelineno-13-2 href=#__codelineno-13-2></a><span class=w>     </span><span class=nf>labs</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=s>'Test'</span><span class=p>,</span><span class=w> </span><span class=n>y</span><span class=o>=</span><span class=s>'Rejection rate'</span><span class=p>,</span><span class=n>title</span><span class=o>=</span><span class=s>'kBET test results'</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-13-3 name=__codelineno-13-3 href=#__codelineno-13-3></a><span class=w>     </span><span class=nf>theme_bw</span><span class=p>()</span><span class=w> </span><span class=o>+</span><span class=w>  </span>
<a id=__codelineno-13-4 name=__codelineno-13-4 href=#__codelineno-13-4></a><span class=w>     </span><span class=nf>scale_y_continuous</span><span class=p>(</span><span class=n>limits</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=m>0</span><span class=p>,</span><span class=m>1</span><span class=p>))</span>
</code></pre></div> <h2 id=wrapping-up>Wrapping up<a class=headerlink href=#wrapping-up title="Permanent link">Â¶</a></h2> <p>This notebook completes the integration of the datasets. We have both used the standard PCA and then a version of the PCA that can model transcript distribution and takes batches into account. Finally, we used a statistical test to check that the integration has been successful.</p></div> </article> </div> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=Part03_normalize_and_dim_reduc.html class="md-footer__link md-footer__link--prev" aria-label="Previous: Normalization" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Normalization </div> </div> </a> <a href=Part05_clustering.html class="md-footer__link md-footer__link--next" aria-label="Next: Clustering and Differential Expression" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Clustering and Differential Expression </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> CC-BY-SA 4.0 license </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox/ target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.3.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "..", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.footer", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=../assets/javascripts/bundle.19047be9.min.js></script> <script src=../javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> <script src=../javascripts/embed-amd.js></script> </body> </html>