<!doctype html><html lang=en class=no-js> <head><script type=application/ld+json>
{
  "@context": "https://schema.org",
  "@id": "https://hds-sandbox.github.io/scRNASeq_course/",
  "@type": "Course",
  "coursePrerequisites": [
    "Knowledge of R, Rstudio and Rmarkdown or Python and Jupyter Notebooks",
    "Basic knowledge of scRNAseq technology",
    "Basic knowledge of data science and statistics such as PCA, clustering and statistical testing"
  ],
  "dct:conformsTo": "https://bioschemas.org/profiles/Course/1.0-RELEASE",
  "description": "This workshop contains a basic tutorial on how to approach scRNAseq experiments, starting from fastq files out of the sequencer. Thus, the workshop does not include any information of laboratory protocols, library preparation or any wet-lab related procedures.",
  "isBasedOn": [
    "https://nf-co.re/scrnaseq",
    "https://satijalab.org/seurat/",
    "https://cran.r-project.org/web/packages/gprofiler2/vignettes/gprofiler2.html",
    "https://bioconductor.org/packages/release/bioc/vignettes/slingshot/inst/doc/vignette.html#constructing-smooth-curves-and-ordering-cells"
  ],
  "keywords": "single cell, RNAseq, python, R",
  "license": {
    "@id": "http://www.apache.org/licenses/",
    "@type": "CreativeWork"
  },
  "mentions": [
    {
      "@id": "#DOI of zenodo repo",
      "@type": "Thing"
    },
    {
      "@id": "#DOI of zenodo data and materials",
      "@type": "Thing"
    }
  ],
  "name": "single-cell RNAseq workshop"
}
</script><title>scRNAseq analysis for Deng et al. (2014) - Single cell sequencing data analysis - an introductiory but comprehensive course</title><link rel=stylesheet href=../../../assets/stylesheets/main.975780f9.min.css><link rel=stylesheet href=../../../assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../../stylesheets/extra.css><script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script><script>"undefined"!=typeof __md_analytics&&__md_analytics()</script></head> <body dir=ltr data-md-color-scheme=brightness data-md-color-primary data-md-color-accent> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#introduction class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=http://hds-sandbox.github.io/ title="Single cell sequencing data analysis - an introductiory but comprehensive course" class="md-header__button md-logo" aria-label="Single cell sequencing data analysis - an introductiory but comprehensive course" data-md-component=logo> <img src=../../../img/logo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Single cell sequencing data analysis - an introductiory but comprehensive course </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> scRNAseq analysis for Deng et al. (2014) </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=brightness data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/hds-sandbox/scRNASeq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/scRNASeq_course </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../../../index.html class=md-tabs__link> Course Introduction </a> </li> <li class=md-tabs__item> <a href=../../../python/installation.html class=md-tabs__link> Python </a> </li> <li class=md-tabs__item> <a href=../../Part01_read_the_data.html class=md-tabs__link> R </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=http://hds-sandbox.github.io/ title="Single cell sequencing data analysis - an introductiory but comprehensive course" class="md-nav__button md-logo" aria-label="Single cell sequencing data analysis - an introductiory but comprehensive course" data-md-component=logo> <img src=../../../img/logo.png alt=logo> </a> Single cell sequencing data analysis - an introductiory but comprehensive course </label> <div class=md-nav__source> <a href=https://github.com/hds-sandbox/scRNASeq_course title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> hds-sandbox/scRNASeq_course </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../index.html class=md-nav__link> Course Introduction </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_2 type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2> Python <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Python data-md-level=1> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Python </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../python/installation.html class=md-nav__link> Setup and installation </a> </li> <li class=md-nav__item> <a href=../../../python/Part01_read_the_data.html class=md-nav__link> Read the data </a> </li> <li class=md-nav__item> <a href=../../../python/Part02_filtering_sample2.html class=md-nav__link> Quality Control (QC) and filtering </a> </li> <li class=md-nav__item> <a href=../../../python/Part03_normalize_and_integrate.html class=md-nav__link> Normalize and Integrate </a> </li> <li class=md-nav__item> <a href=../../../python/Part04_clustering.html class=md-nav__link> Cell clustering and Differential Expression (DE) </a> </li> <li class=md-nav__item> <a href=../../../python/Part05_cell_fate.html class=md-nav__link> Pseudotimes and cell fates </a> </li> <li class=md-nav__item> <a href=../../../python/Part06_Condition_analysis.html class=md-nav__link> Cross-data analysis </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_2_8 type=checkbox id=__nav_2_8> <label class=md-nav__link for=__nav_2_8> Extra <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Extra data-md-level=2> <label class=md-nav__title for=__nav_2_8> <span class="md-nav__icon md-icon"></span> Extra </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../python/Part02_filtering_sample3.html class=md-nav__link> Filtering another sample </a> </li> <li class=md-nav__item> <a href=../../../python/Part02_filtering_sample1.html class=md-nav__link> Filtering a low quality sample </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_3 type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3> R <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=R data-md-level=1> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> R </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../Part01_read_the_data.html class=md-nav__link> Read the data </a> </li> <li class=md-nav__item> <a href=../../Part02_filtering_sample2.html class=md-nav__link> Preprocessing </a> </li> <li class=md-nav__item> <a href=../../Part03_normalize_and_dim_reduc.html class=md-nav__link> Normalization </a> </li> <li class=md-nav__item> <a href=../../Part04_integration.html class=md-nav__link> Integration </a> </li> <li class=md-nav__item> <a href=../../Part05_clustering.html class=md-nav__link> Clustering and Differential Expression </a> </li> <li class=md-nav__item> <a href=../../Part06_pseudotime.html class=md-nav__link> Pseudotimes </a> </li> <li class="md-nav__item md-nav__item--section md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_3_7 type=checkbox id=__nav_3_7> <label class=md-nav__link for=__nav_3_7> Extra <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Extra data-md-level=2> <label class=md-nav__title for=__nav_3_7> <span class="md-nav__icon md-icon"></span> Extra </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../Part02_filtering_sample3.html class=md-nav__link> Preprocessing another sample </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/hds-sandbox/scRNASeq_course/edit/master/docs/R/rmd/old/scRNAseq_Seurat.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1 id=introduction>Introduction<a class=headerlink href=#introduction title="Permanent link">&para;</a></h1> <div class="admonition note"> <p class=admonition-title>Section Overview</p> <p>&#128368; <strong>Time Estimation:</strong> 100 minutes </p> <p>&#128172; <strong>Learning Objectives:</strong> </p> <ol> <li>Load scRNAseq data into R using the Seurat package</li> <li>QC of scRNAseq data</li> <li>Normalization of scRNAseq data</li> <li>Dimensionality reduction and visualization</li> <li>Cell clustering</li> <li>Differential Expression Analysis</li> <li>Functional Analysis with the gprofiler2 package</li> <li>Pseudotime with the slingshot package</li> </ol> </div> <p>scRNAseq data analysis of <a href=https://www.science.org/doi/10.1126/science.1245316>Deng et al. (2014)</a>. This study follows the early embryonic development of mouse cells from zygote to late blastocists (and some adult cells).</p> <p><img alt="Mouse Early Development <https://doi.org/10.1093/molehr/gaw002>" src=img/scRNAseq_Seurat/mouse_development.jpeg></p> <h1 id=load-libraries>Load libraries<a class=headerlink href=#load-libraries title="Permanent link">&para;</a></h1> <div class=highlight><pre><span></span><code><a id=__codelineno-0-1 name=__codelineno-0-1 href=#__codelineno-0-1></a><span class=nf>library</span><span class=p>(</span><span class=n>rhdf5</span><span class=p>)</span>
<a id=__codelineno-0-2 name=__codelineno-0-2 href=#__codelineno-0-2></a><span class=nf>library</span><span class=p>(</span><span class=n>tidyverse</span><span class=p>)</span>
<a id=__codelineno-0-3 name=__codelineno-0-3 href=#__codelineno-0-3></a><span class=nf>library</span><span class=p>(</span><span class=n>Seurat</span><span class=p>)</span>
<a id=__codelineno-0-4 name=__codelineno-0-4 href=#__codelineno-0-4></a><span class=nf>library</span><span class=p>(</span><span class=n>SeuratWrappers</span><span class=p>)</span>
<a id=__codelineno-0-5 name=__codelineno-0-5 href=#__codelineno-0-5></a><span class=nf>library</span><span class=p>(</span><span class=n>patchwork</span><span class=p>)</span>
<a id=__codelineno-0-6 name=__codelineno-0-6 href=#__codelineno-0-6></a><span class=nf>library</span><span class=p>(</span><span class=n>biomaRt</span><span class=p>)</span>
<a id=__codelineno-0-7 name=__codelineno-0-7 href=#__codelineno-0-7></a><span class=nf>library</span><span class=p>(</span><span class=n>slingshot</span><span class=p>)</span>
<a id=__codelineno-0-8 name=__codelineno-0-8 href=#__codelineno-0-8></a><span class=nf>library</span><span class=p>(</span><span class=n>gprofiler2</span><span class=p>)</span>
</code></pre></div> <h1 id=load-dataset>Load dataset<a class=headerlink href=#load-dataset title="Permanent link">&para;</a></h1> <p>Data is a matrix of cells in columns and genes in rows. We specify that the row names (our genes) are in the first column.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-1-1 name=__codelineno-1-1 href=#__codelineno-1-1></a><span class=n>raw_data</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>read.delim</span><span class=p>(</span><span class=s>&quot;../../Data/example_data/GSE45719_cts.txt&quot;</span><span class=p>,</span><span class=w> </span><span class=n>sep</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;\t&quot;</span><span class=p>,</span><span class=w> </span><span class=n>row.names</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=p>)</span>
<a id=__codelineno-1-2 name=__codelineno-1-2 href=#__codelineno-1-2></a><span class=nf>head</span><span class=p>(</span><span class=n>raw_data</span><span class=p>,</span><span class=w> </span><span class=n>n</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>5</span><span class=p>)</span>
</code></pre></div> <p>Here is the metadata. The column names of <code>raw_data</code> are the same as the row names of metadata.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-2-1 name=__codelineno-2-1 href=#__codelineno-2-1></a><span class=n>metadata</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>read.delim</span><span class=p>(</span><span class=s>&quot;../../Data/example_data/GSE45719_metadata.tsv&quot;</span><span class=p>,</span><span class=w> </span><span class=n>header</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>T</span><span class=p>,</span><span class=w> </span><span class=n>sep</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;\t&quot;</span><span class=p>,</span><span class=w> </span><span class=n>quote</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;&quot;</span><span class=p>,</span><span class=w> </span><span class=n>stringsAsFactors</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>F</span><span class=p>,</span><span class=w> </span><span class=n>row.names</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=p>)</span>
<a id=__codelineno-2-2 name=__codelineno-2-2 href=#__codelineno-2-2></a><span class=nf>head</span><span class=p>(</span><span class=n>metadata</span><span class=p>,</span><span class=w> </span><span class=n>n</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>5</span><span class=p>)</span>
</code></pre></div> <p>For the purposes of this workshop, the metadata of this project has been simplified. The most interesting variable is <strong>Cell_type</strong>, which shows the developmental stage of our cells. Cell type is not ordered by its differentiation stage, but we can do that now:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-3-1 name=__codelineno-3-1 href=#__codelineno-3-1></a><span class=n>metadata</span><span class=o>$</span><span class=n>Cell_type</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>factor</span><span class=p>(</span><span class=n>metadata</span><span class=o>$</span><span class=n>Cell_type</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-3-2 name=__codelineno-3-2 href=#__codelineno-3-2></a><span class=w>                             </span><span class=n>levels</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;MII Oocyte&quot;</span><span class=p>,</span><span class=s>&quot;Zygote&quot;</span><span class=p>,</span>
<a id=__codelineno-3-3 name=__codelineno-3-3 href=#__codelineno-3-3></a><span class=w>                                        </span><span class=s>&quot;Early 2-cell&quot;</span><span class=p>,</span><span class=s>&quot;Mid 2-cell&quot;</span><span class=p>,</span><span class=s>&quot;Late 2-cell&quot;</span><span class=p>,</span>
<a id=__codelineno-3-4 name=__codelineno-3-4 href=#__codelineno-3-4></a><span class=w>                                        </span><span class=s>&quot;4-cell&quot;</span><span class=p>,</span><span class=s>&quot;8-cell&quot;</span><span class=p>,</span><span class=s>&quot;16-cell&quot;</span><span class=p>,</span>
<a id=__codelineno-3-5 name=__codelineno-3-5 href=#__codelineno-3-5></a><span class=w>                                        </span><span class=s>&quot;Early blastocyst&quot;</span><span class=p>,</span><span class=s>&quot;Mid blastocyst&quot;</span><span class=p>,</span><span class=s>&quot;Late blastocyst&quot;</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-3-6 name=__codelineno-3-6 href=#__codelineno-3-6></a><span class=w>                                        </span><span class=s>&quot;Fibroblast&quot;</span><span class=p>,</span><span class=s>&quot;Adult&quot;</span><span class=p>))</span>
</code></pre></div> <p>There are some cell types that we are not really interested in, and will be removed in the next step below.</p> <h2 id=seurat-object>Seurat object<a class=headerlink href=#seurat-object title="Permanent link">&para;</a></h2> <p>Now that we have our data loaded, we can create a <strong>Seurat Object</strong>, from which we will perform our analysis. The function <code>CreateSeuratObject()</code> takes as arguments our count matrix, metadata and several options for filtering the data:</p> <ul> <li><code>min.cells</code> will keep features (genes) that are detected in at least this many cells.</li> <li><code>min.features</code> will keep cells with at least this many features detected.</li> </ul> <p>This will prevent from keeping cells and genes with an immense majority of 0’s as values. In addition, we are removing some cell types we do not want to analyse.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-4-1 name=__codelineno-4-1 href=#__codelineno-4-1></a><span class=n>raw_ann</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>CreateSeuratObject</span><span class=p>(</span><span class=n>counts</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>raw_data</span><span class=p>,</span><span class=w> </span><span class=n>meta.data</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>metadata</span><span class=p>,</span><span class=w> </span><span class=n>min.cells</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>3</span><span class=p>,</span><span class=w> </span><span class=n>min.features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>200</span><span class=p>,)</span>
<a id=__codelineno-4-2 name=__codelineno-4-2 href=#__codelineno-4-2></a><span class=n>raw_ann</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>subset</span><span class=p>(</span><span class=n>raw_ann</span><span class=p>,</span><span class=w> </span><span class=n>cells</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>colnames</span><span class=p>(</span><span class=n>raw_ann</span><span class=p>)[</span><span class=o>!</span><span class=n>raw_ann</span><span class=o>$</span><span class=n>Cell_type</span><span class=w> </span><span class=o>%in%</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;MII Oocyte&quot;</span><span class=p>,</span><span class=s>&quot;Fibroblast&quot;</span><span class=p>,</span><span class=s>&quot;Adult&quot;</span><span class=p>)])</span>
</code></pre></div> <h1 id=quality-control>Quality Control<a class=headerlink href=#quality-control title="Permanent link">&para;</a></h1> <p>The Seurat object initialization step above only considered cells that expressed at least 300 genes and genes detected in at least 3 cells. Here is how many cells and genes we start with:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-5-1 name=__codelineno-5-1 href=#__codelineno-5-1></a><span class=nf>print</span><span class=p>(</span><span class=nf>paste0</span><span class=p>(</span><span class=s>&quot;Before filtering: &quot;</span><span class=p>,</span><span class=w> </span><span class=nf>dim</span><span class=p>(</span><span class=n>raw_ann</span><span class=p>)[</span><span class=m>2</span><span class=p>],</span><span class=w> </span><span class=s>&quot; cells &quot;</span><span class=p>,</span><span class=w>  </span><span class=nf>dim</span><span class=p>(</span><span class=n>raw_ann</span><span class=p>)[</span><span class=m>1</span><span class=p>],</span><span class=w> </span><span class=s>&quot; genes&quot;</span><span class=p>))</span>
</code></pre></div> <p>Additionally, we would like to exclude cells that are damaged. A common metric to judge this (although by no means the only one) is the relative expression of mitochondrially derived genes. When the cells apoptose due to stress, their mitochondria becomes leaky and there is widespread RNA degradation. Thus a relative enrichment of mitochondrially derived genes can be a tell-tale sign of cell stress. Here, we compute the proportion of transcripts that are of mitochondrial origin for every cell (<em>percent.mito</em>), and visualize its distribution as a violin plot. We also use the <code>GenePlot()</code> function to observe how percent.mito correlates with other metrics.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-6-1 name=__codelineno-6-1 href=#__codelineno-6-1></a><span class=n>raw_ann</span><span class=p>[[</span><span class=s>&#39;percent.mito&#39;</span><span class=p>]]</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>PercentageFeatureSet</span><span class=p>(</span><span class=n>raw_ann</span><span class=p>,</span><span class=w> </span><span class=n>pattern</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;^mt-&quot;</span><span class=p>)</span>
</code></pre></div> <p>Now, sometimes, mitochondrial genes are a bit tricky to find, specially if your genes are not gene names, but gene IDs. You might have already a collection of gene names you want to use, but it is not always the case. Therefore, it might very useful to have some extra annotation on our genes which will help to select mitochondrial genes (or ERCC genes and ribosomal genes). Since this dataset was aligned using the mouse genome version <em>mm9</em>, we will use the mm9 annotation from <a href=http://www.biomart.org/ >Biomart</a>. The package <strong>biomaRt</strong> will do the job!</p> <div class=highlight><pre><span></span><code><a id=__codelineno-7-1 name=__codelineno-7-1 href=#__codelineno-7-1></a><span class=n>ensembl67</span><span class=o>=</span><span class=nf>useMart</span><span class=p>(</span><span class=n>host</span><span class=o>=</span><span class=s>&#39;http://feb2014.archive.ensembl.org/&#39;</span><span class=p>,</span><span class=w> </span><span class=c1># select latest version of mm9 genome annotation</span>
<a id=__codelineno-7-2 name=__codelineno-7-2 href=#__codelineno-7-2></a><span class=w>                  </span><span class=n>biomart</span><span class=o>=</span><span class=s>&#39;ENSEMBL_MART_ENSEMBL&#39;</span><span class=p>,</span><span class=w> </span><span class=n>dataset</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;mmusculus_gene_ensembl&quot;</span><span class=p>)</span>
<a id=__codelineno-7-3 name=__codelineno-7-3 href=#__codelineno-7-3></a>
<a id=__codelineno-7-4 name=__codelineno-7-4 href=#__codelineno-7-4></a><span class=c1># we get a list of annotations we would like to fetch</span>
<a id=__codelineno-7-5 name=__codelineno-7-5 href=#__codelineno-7-5></a><span class=n>mm9.gene.annotations</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>biomaRt</span><span class=o>::</span><span class=nf>getBM</span><span class=p>(</span><span class=n>mart</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>ensembl67</span><span class=p>,</span><span class=w> </span><span class=n>attributes</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;ensembl_gene_id&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;external_gene_id&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;description&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;chromosome_name&quot;</span><span class=p>))</span>
<a id=__codelineno-7-6 name=__codelineno-7-6 href=#__codelineno-7-6></a><span class=nf>head</span><span class=p>(</span><span class=n>mm9.gene.annotations</span><span class=p>)</span>
</code></pre></div> <p>We select that mitochondrial genes are those with the <strong>chromosome_name</strong> annotation equal to "MT".</p> <div class=highlight><pre><span></span><code><a id=__codelineno-8-1 name=__codelineno-8-1 href=#__codelineno-8-1></a><span class=n>mt_genes</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>mm9.gene.annotations</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>filter</span><span class=p>(</span><span class=n>chromosome_name</span><span class=w> </span><span class=o>==</span><span class=w> </span><span class=s>&quot;MT&quot;</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>pull</span><span class=p>(</span><span class=n>external_gene_id</span><span class=p>)</span>
<a id=__codelineno-8-2 name=__codelineno-8-2 href=#__codelineno-8-2></a><span class=n>mt_genes</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=n>mt_genes</span><span class=w> </span><span class=o>%in%</span><span class=w> </span><span class=nf>rownames</span><span class=p>(</span><span class=n>raw_ann</span><span class=p>)</span>
<a id=__codelineno-8-3 name=__codelineno-8-3 href=#__codelineno-8-3></a>
<a id=__codelineno-8-4 name=__codelineno-8-4 href=#__codelineno-8-4></a><span class=n>raw_ann</span><span class=p>[[</span><span class=s>&#39;percent.mito&#39;</span><span class=p>]]</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>PercentageFeatureSet</span><span class=p>(</span><span class=n>raw_ann</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>mt_genes</span><span class=p>)</span>
</code></pre></div> <p>If your data contains spike-ins (ERCC genes), you can also compute the proportion of transcripts belonging to them. Furthermore, it might be useful to calculate the percentage of reads aligned to ribosomal genes, since it has been shown that they can skew the data due to their high variability.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-9-1 name=__codelineno-9-1 href=#__codelineno-9-1></a><span class=n>raw_ann</span><span class=p>[[</span><span class=s>&#39;percent.ercc&#39;</span><span class=p>]]</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>PercentageFeatureSet</span><span class=p>(</span><span class=n>raw_ann</span><span class=p>,</span><span class=w> </span><span class=n>pattern</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;^ERCC-&quot;</span><span class=p>)</span>
<a id=__codelineno-9-2 name=__codelineno-9-2 href=#__codelineno-9-2></a><span class=n>raw_ann</span><span class=p>[[</span><span class=s>&#39;percent.ribo&#39;</span><span class=p>]]</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>PercentageFeatureSet</span><span class=p>(</span><span class=n>raw_ann</span><span class=p>,</span><span class=w> </span><span class=n>pattern</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;^Rp[ls]&quot;</span><span class=p>)</span>
</code></pre></div> <p>Nonetheless, in this experiment there aren’t ERCC genes or mitochondrial genes:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-10-1 name=__codelineno-10-1 href=#__codelineno-10-1></a><span class=nf>sum</span><span class=p>(</span><span class=n>raw_ann</span><span class=o>$</span><span class=n>percent.ercc</span><span class=p>)</span>
<a id=__codelineno-10-2 name=__codelineno-10-2 href=#__codelineno-10-2></a><span class=nf>sum</span><span class=p>(</span><span class=n>raw_ann</span><span class=o>$</span><span class=n>percent.mito</span><span class=p>)</span>
</code></pre></div> <h2 id=visualizations>Visualizations<a class=headerlink href=#visualizations title="Permanent link">&para;</a></h2> <p>It is extremely useful to visualize QC measurements calculated so far. Violin plots (fancy boxplots) are a great way to check the distribution of values of all our QC measurements. When initializing the Seurat Object, Seurat calculates also the number of genes detected and the total library size per cell.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-11-1 name=__codelineno-11-1 href=#__codelineno-11-1></a><span class=nf>VlnPlot</span><span class=p>(</span><span class=n>raw_ann</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-11-2 name=__codelineno-11-2 href=#__codelineno-11-2></a><span class=w>        </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;nFeature_RNA&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;nCount_RNA&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;percent.ribo&quot;</span><span class=p>),</span>
<a id=__codelineno-11-3 name=__codelineno-11-3 href=#__codelineno-11-3></a><span class=w>        </span><span class=n>ncol</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>4</span><span class=p>,</span><span class=w> </span><span class=n>group.by</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;Cell_type&quot;</span><span class=p>)</span>
</code></pre></div> <p>It seems that the late 2-cells have a very high total number of reads. We should not filter them out, there are very few! In addition, our percentage of ribosomal counts is also quite low (maximum is ~5%).</p> <p>We can check the relationship between library size and number of genes detected:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-12-1 name=__codelineno-12-1 href=#__codelineno-12-1></a><span class=nf>FeatureScatter</span><span class=p>(</span><span class=n>raw_ann</span><span class=p>,</span><span class=w> </span><span class=n>feature1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;nCount_RNA&quot;</span><span class=p>,</span><span class=w> </span><span class=n>feature2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;nFeature_RNA&quot;</span><span class=p>,</span><span class=w> </span><span class=n>group.by</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;Cell_type&quot;</span><span class=p>)</span>
</code></pre></div> <p>In other experiment with more cells, we might consider removing cells with an unexpectedly high number of detected genes and reads. These cells might be doublets, that is, two cells that were sequenced together!</p> <h2 id=filtering>Filtering<a class=headerlink href=#filtering title="Permanent link">&para;</a></h2> <p>Some automatic filtering can be made using quantiles! Using the extreme upper and lower quantiles (0.99 and 0.01) we can make sure that outliers are removed. In this case, we will remove cells with very low library size and genes detected (lower than quantile 0.01).</p> <div class=highlight><pre><span></span><code><a id=__codelineno-13-1 name=__codelineno-13-1 href=#__codelineno-13-1></a><span class=n>feature_min</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>quantile</span><span class=p>(</span><span class=n>raw_ann</span><span class=o>$</span><span class=n>nFeature_RNA</span><span class=p>,</span><span class=w> </span><span class=n>probs</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.01</span><span class=p>)</span>
<a id=__codelineno-13-2 name=__codelineno-13-2 href=#__codelineno-13-2></a><span class=n>count_min</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>quantile</span><span class=p>(</span><span class=n>raw_ann</span><span class=o>$</span><span class=n>nCount_RNA</span><span class=p>,</span><span class=w> </span><span class=n>probs</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.01</span><span class=p>)</span>
</code></pre></div> <p>We can subset our dataset using the function <code>subset()</code>.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-14-1 name=__codelineno-14-1 href=#__codelineno-14-1></a><span class=n>adata</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>subset</span><span class=p>(</span><span class=n>raw_ann</span><span class=p>,</span><span class=w> </span><span class=n>subset</span><span class=w> </span><span class=o>=</span><span class=w> </span>
<a id=__codelineno-14-2 name=__codelineno-14-2 href=#__codelineno-14-2></a><span class=w>                    </span><span class=n>nFeature_RNA</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=n>feature_min</span><span class=w>  </span><span class=o>&amp;</span><span class=w> </span>
<a id=__codelineno-14-3 name=__codelineno-14-3 href=#__codelineno-14-3></a><span class=w>                    </span><span class=n>nCount_RNA</span><span class=w> </span><span class=o>&gt;</span><span class=w> </span><span class=n>count_min</span><span class=p>)</span>
<a id=__codelineno-14-4 name=__codelineno-14-4 href=#__codelineno-14-4></a>
<a id=__codelineno-14-5 name=__codelineno-14-5 href=#__codelineno-14-5></a><span class=nf>rm</span><span class=p>(</span><span class=n>raw_ann</span><span class=p>)</span><span class=w> </span><span class=c1># we remove the initial unfiltered dataset to reduce computational resources, this is not necessary!</span>
</code></pre></div> <p>Finally, this is how many cells and genes we have after filtering:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-15-1 name=__codelineno-15-1 href=#__codelineno-15-1></a><span class=nf>print</span><span class=p>(</span><span class=nf>paste0</span><span class=p>(</span><span class=s>&quot;After filtering: &quot;</span><span class=p>,</span><span class=w> </span><span class=nf>dim</span><span class=p>(</span><span class=n>adata</span><span class=p>)[</span><span class=m>2</span><span class=p>],</span><span class=w> </span><span class=s>&quot; cells &quot;</span><span class=p>,</span><span class=w>  </span><span class=nf>dim</span><span class=p>(</span><span class=n>adata</span><span class=p>)[</span><span class=m>1</span><span class=p>],</span><span class=w> </span><span class=s>&quot; genes&quot;</span><span class=p>))</span>
</code></pre></div> <p>And this is how our filtered data looks like:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-16-1 name=__codelineno-16-1 href=#__codelineno-16-1></a><span class=nf>VlnPlot</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-16-2 name=__codelineno-16-2 href=#__codelineno-16-2></a><span class=w>        </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;nFeature_RNA&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;nCount_RNA&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;percent.ribo&quot;</span><span class=p>),</span>
<a id=__codelineno-16-3 name=__codelineno-16-3 href=#__codelineno-16-3></a><span class=w>        </span><span class=n>ncol</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>4</span><span class=p>,</span><span class=w> </span><span class=n>group.by</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;Cell_type&quot;</span><span class=p>)</span>
</code></pre></div> <h1 id=exploratory-analysis>Exploratory analysis<a class=headerlink href=#exploratory-analysis title="Permanent link">&para;</a></h1> <h2 id=normalization>Normalization<a class=headerlink href=#normalization title="Permanent link">&para;</a></h2> <p>Now that the data is filtered, we can proceed to normalize our count matrix. Seurat normalizes the gene expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. There have been many methods to normalize the data, but this is the simplest and the most intuitive. The division by total expression is done to change all expression counts to a relative measure, since experience has suggested that technical factors (e.g. capture rate, efficiency of reverse transcription) are largely responsible for the variation in the number of molecules per cell, although genuine biological factors (e.g. cell cycle stage, cell size) also play a smaller, but non-negligible role. The log-transformation is a commonly used transformation that has many desirable properties, such as variance stabilization (can you think of others?).</p> <div class=highlight><pre><span></span><code><a id=__codelineno-17-1 name=__codelineno-17-1 href=#__codelineno-17-1></a><span class=n>adata</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>NormalizeData</span><span class=p>(</span><span class=n>adata</span><span class=p>)</span>
<a id=__codelineno-17-2 name=__codelineno-17-2 href=#__codelineno-17-2></a><span class=n>adata</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindVariableFeatures</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>selection.method</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;vst&quot;</span><span class=p>,</span><span class=w> </span><span class=n>nfeatures</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>2000</span><span class=p>)</span>
</code></pre></div> <h2 id=identification-of-variable-genes>Identification of Variable Genes<a class=headerlink href=#identification-of-variable-genes title="Permanent link">&para;</a></h2> <p>Identify most variable genes and label top 5 most highly variable.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-18-1 name=__codelineno-18-1 href=#__codelineno-18-1></a><span class=n>top10</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>head</span><span class=p>(</span><span class=nf>VariableFeatures</span><span class=p>(</span><span class=n>adata</span><span class=p>),</span><span class=w> </span><span class=m>5</span><span class=p>)</span>
<a id=__codelineno-18-2 name=__codelineno-18-2 href=#__codelineno-18-2></a><span class=nf>LabelPoints</span><span class=p>(</span><span class=n>plot</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>VariableFeaturePlot</span><span class=p>(</span><span class=n>adata</span><span class=p>),</span><span class=w> </span><span class=n>points</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>top10</span><span class=p>,</span><span class=w> </span><span class=n>repel</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>T</span><span class=p>)</span>
</code></pre></div> <h2 id=scale>Scale<a class=headerlink href=#scale title="Permanent link">&para;</a></h2> <p>Gene expression scaling is necessary for proper clustering of our cells. Since genes may be expressed in very different orders of magnitude, extreme expression levels may drive the separation between cells and bias the results.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-19-1 name=__codelineno-19-1 href=#__codelineno-19-1></a><span class=n>adata</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>ScaleData</span><span class=p>(</span><span class=n>adata</span><span class=p>)</span>
</code></pre></div> <h2 id=pca>PCA<a class=headerlink href=#pca title="Permanent link">&para;</a></h2> <p>We can perform a Principal Component analysis using the following function. PCA can be used for visualization of our cells, as well as clustering and other dimensionality reduction methods such as t-SNE or UMAP.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-20-1 name=__codelineno-20-1 href=#__codelineno-20-1></a><span class=n>adata</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>RunPCA</span><span class=p>(</span><span class=n>adata</span><span class=p>)</span>
</code></pre></div> <p>We can also calculate and visualize the importance of each gene for each Principal Component.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-21-1 name=__codelineno-21-1 href=#__codelineno-21-1></a><span class=n>pcalod_1</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>VizDimLoadings</span><span class=p>(</span><span class=n>object</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>theme</span><span class=p>(</span><span class=n>axis.text.y</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>element_text</span><span class=p>(</span><span class=n>size</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>8</span><span class=p>))</span><span class=w> </span>
<a id=__codelineno-21-2 name=__codelineno-21-2 href=#__codelineno-21-2></a><span class=n>pcalod_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>VizDimLoadings</span><span class=p>(</span><span class=n>object</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>2</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>theme</span><span class=p>(</span><span class=n>axis.text.y</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>element_text</span><span class=p>(</span><span class=n>size</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>8</span><span class=p>))</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-22-1 name=__codelineno-22-1 href=#__codelineno-22-1></a><span class=nf>CombinePlots</span><span class=p>(</span><span class=n>plots</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>list</span><span class=p>(</span><span class=n>pcalod_1</span><span class=p>,</span><span class=w> </span><span class=n>pcalod_2</span><span class=p>),</span><span class=w> </span><span class=n>ncol</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>2</span><span class=p>)</span>
</code></pre></div> <p>Visualization only show negative loadings cause there are many more than positives. We can see a balanced plot using <em>balance = TRUE</em>:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-23-1 name=__codelineno-23-1 href=#__codelineno-23-1></a><span class=n>pcalod_1</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>VizDimLoadings</span><span class=p>(</span><span class=n>object</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=p>,</span><span class=w> </span><span class=n>balanced</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>theme</span><span class=p>(</span><span class=n>axis.text.y</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>element_text</span><span class=p>(</span><span class=n>size</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>8</span><span class=p>))</span><span class=w> </span>
<a id=__codelineno-23-2 name=__codelineno-23-2 href=#__codelineno-23-2></a><span class=n>pcalod_2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>VizDimLoadings</span><span class=p>(</span><span class=n>object</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>2</span><span class=p>,</span><span class=w> </span><span class=n>balanced</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>theme</span><span class=p>(</span><span class=n>axis.text.y</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>element_text</span><span class=p>(</span><span class=n>size</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>8</span><span class=p>))</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-24-1 name=__codelineno-24-1 href=#__codelineno-24-1></a><span class=nf>CombinePlots</span><span class=p>(</span><span class=n>plots</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>list</span><span class=p>(</span><span class=n>pcalod_1</span><span class=p>,</span><span class=w> </span><span class=n>pcalod_2</span><span class=p>),</span><span class=w> </span><span class=n>ncol</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>2</span><span class=p>)</span>
</code></pre></div> <h2 id=cell-clustering>Cell clustering<a class=headerlink href=#cell-clustering title="Permanent link">&para;</a></h2> <p>Now that we have calculated our components, we can proceed to select the number of PCs necessary to perform clustering. There are two methods that we can use to determine the proper number of dimensions:</p> <h3 id=pc-selection>PC selection<a class=headerlink href=#pc-selection title="Permanent link">&para;</a></h3> <p><strong>Jackstraw method</strong></p> <p>The Jackstraw method randomly permutes a subset of data, and calculates projected PCA scores for these ‘random’ genes. Then compares the PCA scores for the ‘random’ genes with the observed PCA scores to determine statistical significance. End result is a p-value for each gene’s association with each principal component.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-25-1 name=__codelineno-25-1 href=#__codelineno-25-1></a><span class=n>adata</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>JackStraw</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>num.replicate</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>100</span><span class=p>)</span>
<a id=__codelineno-25-2 name=__codelineno-25-2 href=#__codelineno-25-2></a><span class=n>adata</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>ScoreJackStraw</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>20</span><span class=p>)</span>
<a id=__codelineno-25-3 name=__codelineno-25-3 href=#__codelineno-25-3></a><span class=nf>JackStrawPlot</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>20</span><span class=p>)</span>
</code></pre></div> <p>As you can see in the plot above, there is a change in the orders of magnitude of the PCs’ p-values. Suggesting that we may cut off around PC8.</p> <p><strong>Elbow method</strong></p> <p>The elbow method allows us to explore the explained variation of each of the Principal Components. The plot usually looks like an "elbow", where adding more PCs does not really contribute to the amount of explained variation. We can see again that we reach a plateau around PC6 or PC8.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-26-1 name=__codelineno-26-1 href=#__codelineno-26-1></a><span class=nf>ElbowPlot</span><span class=p>(</span><span class=n>adata</span><span class=p>)</span>
</code></pre></div> <h3 id=clustering>Clustering<a class=headerlink href=#clustering title="Permanent link">&para;</a></h3> <p>We could include 6 PCs, but it does not hurt to use more; there are very few cells in this experiment and we would like to include as much information as possible. Use the <code>resolution</code> argument of the <code>FindClusters()</code> function to fine-tune the number of clusters to find. The larger the number the less clusters it will find.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-27-1 name=__codelineno-27-1 href=#__codelineno-27-1></a><span class=n>adata</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindNeighbors</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>20</span><span class=p>)</span>
<a id=__codelineno-27-2 name=__codelineno-27-2 href=#__codelineno-27-2></a><span class=n>adata</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindClusters</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>resolution</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.8</span><span class=p>)</span>
</code></pre></div> <h1 id=visualization>Visualization<a class=headerlink href=#visualization title="Permanent link">&para;</a></h1> <div class=highlight><pre><span></span><code><a id=__codelineno-28-1 name=__codelineno-28-1 href=#__codelineno-28-1></a><span class=n>adata</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>RunTSNE</span><span class=p>(</span><span class=n>adata</span><span class=p>)</span>
<a id=__codelineno-28-2 name=__codelineno-28-2 href=#__codelineno-28-2></a><span class=n>adata</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>RunUMAP</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>dims</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>1</span><span class=o>:</span><span class=m>20</span><span class=p>)</span>
</code></pre></div> <h2 id=pca_1>PCA<a class=headerlink href=#pca_1 title="Permanent link">&para;</a></h2> <div class=highlight><pre><span></span><code><a id=__codelineno-29-1 name=__codelineno-29-1 href=#__codelineno-29-1></a><span class=nf>DimPlot</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;pca&quot;</span><span class=p>,</span><span class=w> </span><span class=n>group.by</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&#39;Cell_type&#39;</span><span class=p>))</span><span class=w> </span><span class=o>+</span><span class=w> </span>
<a id=__codelineno-29-2 name=__codelineno-29-2 href=#__codelineno-29-2></a><span class=w>  </span><span class=nf>DimPlot</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;pca&quot;</span><span class=p>,</span><span class=w> </span><span class=n>group.by</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&#39;seurat_clusters&#39;</span><span class=p>))</span>
</code></pre></div> <h2 id=umap>UMAP<a class=headerlink href=#umap title="Permanent link">&para;</a></h2> <div class=highlight><pre><span></span><code><a id=__codelineno-30-1 name=__codelineno-30-1 href=#__codelineno-30-1></a><span class=n>p1</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DimPlot</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;umap&quot;</span><span class=p>,</span><span class=w> </span><span class=n>group.by</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&#39;Cell_type&#39;</span><span class=p>)</span>
<a id=__codelineno-30-2 name=__codelineno-30-2 href=#__codelineno-30-2></a><span class=n>p2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DimPlot</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;umap&quot;</span><span class=p>)</span>
<a id=__codelineno-30-3 name=__codelineno-30-3 href=#__codelineno-30-3></a><span class=n>p1</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=n>p2</span>
</code></pre></div> <h2 id=tsne>TSNE<a class=headerlink href=#tsne title="Permanent link">&para;</a></h2> <div class=highlight><pre><span></span><code><a id=__codelineno-31-1 name=__codelineno-31-1 href=#__codelineno-31-1></a><span class=n>p3</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DimPlot</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;tsne&quot;</span><span class=p>,</span><span class=w> </span><span class=n>group.by</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&#39;Cell_type&#39;</span><span class=p>)</span>
<a id=__codelineno-31-2 name=__codelineno-31-2 href=#__codelineno-31-2></a><span class=n>p4</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>DimPlot</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;tsne&quot;</span><span class=p>)</span>
<a id=__codelineno-31-3 name=__codelineno-31-3 href=#__codelineno-31-3></a><span class=n>p3</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=n>p4</span>
</code></pre></div> <h1 id=differential-expression-analysis>Differential Expression Analysis<a class=headerlink href=#differential-expression-analysis title="Permanent link">&para;</a></h1> <p>Differential testing works a bit different than in bulk RNA-Seq analysis. Usually, your experiment will include dozens or hundreds of cells per cluster/condition. We can make use of Wilcox tests and multiple testing correction to identify statistically significant genes using the <code>FindAllMarkers()</code> function. This function will gather all "markers" for each of your <em>Ident</em> variable values (your conditions or identified clusters).</p> <div class=highlight><pre><span></span><code><a id=__codelineno-32-1 name=__codelineno-32-1 href=#__codelineno-32-1></a><span class=n>markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindAllMarkers</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=n>logfc.threshold</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.5</span><span class=p>,</span><span class=w> </span><span class=n>only.pos</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>T</span><span class=p>)</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-33-1 name=__codelineno-33-1 href=#__codelineno-33-1></a><span class=n>markers</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-33-2 name=__codelineno-33-2 href=#__codelineno-33-2></a><span class=w>    </span><span class=nf>group_by</span><span class=p>(</span><span class=n>cluster</span><span class=p>)</span><span class=w> </span><span class=o>%&gt;%</span>
<a id=__codelineno-33-3 name=__codelineno-33-3 href=#__codelineno-33-3></a><span class=w>    </span><span class=nf>top_n</span><span class=p>(</span><span class=n>n</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>10</span><span class=p>,</span><span class=w> </span><span class=n>wt</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>avg_log2FC</span><span class=p>)</span><span class=w> </span><span class=o>-&gt;</span><span class=w> </span><span class=n>top10</span>
<a id=__codelineno-33-4 name=__codelineno-33-4 href=#__codelineno-33-4></a>
<a id=__codelineno-33-5 name=__codelineno-33-5 href=#__codelineno-33-5></a><span class=nf>head</span><span class=p>(</span><span class=n>top10</span><span class=p>)</span>
</code></pre></div> <p>On the other hand, if you want to make a specific comparison, you can use the <code>FindMarkers()</code> function:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-34-1 name=__codelineno-34-1 href=#__codelineno-34-1></a><span class=n>cluster2.markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindMarkers</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>ident.1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>2</span><span class=p>,</span><span class=w> </span><span class=n>min.pct</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.5</span><span class=p>,</span><span class=w> </span><span class=n>only.pos</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>T</span><span class=p>)</span><span class=w> </span><span class=c1># Only cluster 2 markers</span>
<a id=__codelineno-34-2 name=__codelineno-34-2 href=#__codelineno-34-2></a><span class=n>cluster2.markers</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>arrange</span><span class=p>(</span><span class=nf>desc</span><span class=p>(</span><span class=n>avg_log2FC</span><span class=p>))</span><span class=w> </span><span class=o>%&gt;%</span><span class=w> </span><span class=nf>head</span><span class=p>(</span><span class=m>10</span><span class=p>)</span>
</code></pre></div> <p>Even compare specific clusters against others by selecting <code>ident.1</code> and <code>ident.2</code>:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-35-1 name=__codelineno-35-1 href=#__codelineno-35-1></a><span class=n>cluster5.markers</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>FindMarkers</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>ident.1</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>5</span><span class=p>,</span><span class=w> </span><span class=n>ident.2</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=m>0</span><span class=p>,</span><span class=w> </span><span class=m>3</span><span class=p>),</span><span class=w> </span><span class=c1># Differences between cluster 5 and clusters 0 and 3</span>
<a id=__codelineno-35-2 name=__codelineno-35-2 href=#__codelineno-35-2></a><span class=w>                                </span><span class=n>min.pct</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.5</span><span class=p>,</span><span class=w> </span><span class=n>only.pos</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=bp>T</span><span class=p>)</span><span class=w> </span>
</code></pre></div> <h2 id=visualizations_1>Visualizations<a class=headerlink href=#visualizations_1 title="Permanent link">&para;</a></h2> <p>We can visualize all markers in a heatmap just like this! To go easy on the visualization, we only use the top 10 markers.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-36-1 name=__codelineno-36-1 href=#__codelineno-36-1></a><span class=nf>DoHeatmap</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>top10</span><span class=o>$</span><span class=n>gene</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>NoLegend</span><span class=p>()</span>
</code></pre></div> <p>We can also plot the expression on our PCA, tSNE or UMAP. We only need to pass a vector of genes to the <code>features</code> argument of <code>FeaturePlot()</code>.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-37-1 name=__codelineno-37-1 href=#__codelineno-37-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>top10</span><span class=o>$</span><span class=n>gene</span><span class=p>[</span><span class=m>1</span><span class=o>:</span><span class=m>4</span><span class=p>],</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;pca&quot;</span><span class=p>)</span>
</code></pre></div> <p>Or as a violin plot, a ridge plot or a dot plot.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-38-1 name=__codelineno-38-1 href=#__codelineno-38-1></a><span class=nf>VlnPlot</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>top10</span><span class=o>$</span><span class=n>gene</span><span class=p>[</span><span class=m>1</span><span class=p>])</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>RidgePlot</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>top10</span><span class=o>$</span><span class=n>gene</span><span class=p>[</span><span class=m>1</span><span class=p>])</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-39-1 name=__codelineno-39-1 href=#__codelineno-39-1></a><span class=nf>DotPlot</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>top10</span><span class=o>$</span><span class=n>gene</span><span class=p>[</span><span class=m>1</span><span class=o>:</span><span class=m>10</span><span class=p>])</span>
</code></pre></div> <h1 id=functional-analysis-with-gprofiler2>Functional analysis with gprofiler2<a class=headerlink href=#functional-analysis-with-gprofiler2 title="Permanent link">&para;</a></h1> <p><code>gost()</code> function allows us to do functional profiling of gene lists, such as our differentially expressed genes. The function performs statistical enrichment analysis to find over-representation of terms from Gene Ontology, biological pathways like KEGG and Reactome, human disease annotations, etc. This is done by using hypergeometric tests that are corrected for multiple testing.</p> <h2 id=single-query>Single query<a class=headerlink href=#single-query title="Permanent link">&para;</a></h2> <p>A standard input of the <code>gost()</code> function is a (named) list of gene identifiers. The list can consist of mixed types of identifiers (proteins, transcripts, microarray IDs, etc), SNP IDs, chromosomal intervals or functional term IDs.</p> <p>The result is a named list where <em>result</em> is a data.frame with the enrichment analysis results and <em>meta</em> containing a named list with all the metadata for the query.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-40-1 name=__codelineno-40-1 href=#__codelineno-40-1></a><span class=n>gostres</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>gost</span><span class=p>(</span><span class=n>query</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>rownames</span><span class=p>(</span><span class=n>cluster2.markers</span><span class=p>),</span><span class=w> </span>
<a id=__codelineno-40-2 name=__codelineno-40-2 href=#__codelineno-40-2></a><span class=w>                </span><span class=n>organism</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;mmusculus&quot;</span><span class=p>,</span><span class=w> </span><span class=n>ordered_query</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-40-3 name=__codelineno-40-3 href=#__codelineno-40-3></a><span class=w>                </span><span class=n>multi_query</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>,</span><span class=w> </span><span class=n>significant</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>,</span><span class=w> </span><span class=n>exclude_iea</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-40-4 name=__codelineno-40-4 href=#__codelineno-40-4></a><span class=w>                </span><span class=n>measure_underrepresentation</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>,</span><span class=w> </span><span class=n>evcodes</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-40-5 name=__codelineno-40-5 href=#__codelineno-40-5></a><span class=w>                </span><span class=n>user_threshold</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>0.05</span><span class=p>,</span><span class=w> </span><span class=n>correction_method</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;g_SCS&quot;</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-40-6 name=__codelineno-40-6 href=#__codelineno-40-6></a><span class=w>                </span><span class=n>domain_scope</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;annotated&quot;</span><span class=p>,</span><span class=w> </span><span class=n>custom_bg</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>NULL</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-40-7 name=__codelineno-40-7 href=#__codelineno-40-7></a><span class=w>                </span><span class=n>numeric_ns</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;&quot;</span><span class=p>,</span><span class=w> </span><span class=n>sources</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>NULL</span><span class=p>,</span><span class=w> </span><span class=n>as_short_link</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
<a id=__codelineno-40-8 name=__codelineno-40-8 href=#__codelineno-40-8></a>
<a id=__codelineno-40-9 name=__codelineno-40-9 href=#__codelineno-40-9></a><span class=nf>head</span><span class=p>(</span><span class=n>gostres</span><span class=o>$</span><span class=n>result</span><span class=p>)</span>
</code></pre></div> <p>The result data.frame contains the following columns:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-41-1 name=__codelineno-41-1 href=#__codelineno-41-1></a><span class=nf>names</span><span class=p>(</span><span class=n>gostres</span><span class=o>$</span><span class=n>meta</span><span class=p>)</span>
</code></pre></div> <h2 id=multiple-queries>Multiple queries<a class=headerlink href=#multiple-queries title="Permanent link">&para;</a></h2> <p>The function <code>gost()</code> also allows to perform enrichment on multiple input gene lists. Multiple queries are automatically detected if the input query is a list of vectors with gene identifiers and the results are combined into identical data.frame as in case of single query.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-42-1 name=__codelineno-42-1 href=#__codelineno-42-1></a><span class=n>multi_gostres1</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>gost</span><span class=p>(</span><span class=n>query</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>list</span><span class=p>(</span><span class=s>&quot;chromX&quot;</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;X:1000:1000000&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;rs17396340&quot;</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-42-2 name=__codelineno-42-2 href=#__codelineno-42-2></a><span class=w>                                                 </span><span class=s>&quot;GO:0005005&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;ENSG00000156103&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;NLRP1&quot;</span><span class=p>),</span>
<a id=__codelineno-42-3 name=__codelineno-42-3 href=#__codelineno-42-3></a><span class=w>                             </span><span class=s>&quot;chromY&quot;</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;Y:1:10000000&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;rs17396340&quot;</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-42-4 name=__codelineno-42-4 href=#__codelineno-42-4></a><span class=w>                                          </span><span class=s>&quot;GO:0005005&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;ENSG00000156103&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;NLRP1&quot;</span><span class=p>)),</span><span class=w> </span>
<a id=__codelineno-42-5 name=__codelineno-42-5 href=#__codelineno-42-5></a><span class=w>                       </span><span class=n>multi_query</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
<a id=__codelineno-42-6 name=__codelineno-42-6 href=#__codelineno-42-6></a>
<a id=__codelineno-42-7 name=__codelineno-42-7 href=#__codelineno-42-7></a><span class=nf>head</span><span class=p>(</span><span class=n>multi_gostres1</span><span class=o>$</span><span class=n>result</span><span class=p>,</span><span class=w> </span><span class=m>3</span><span class=p>)</span>
</code></pre></div> <p>The column "query" in the result dataframe will now contain the corresponding name for the query. If no name is specified, then the query name is defined as the order of query with the prefix "query_." Another option for multiple gene lists is setting the parameter <code>multiquery = TRUE</code>. Then the results from all of the input queries are grouped according to term IDs for better comparison.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-43-1 name=__codelineno-43-1 href=#__codelineno-43-1></a><span class=n>multi_gostres2</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>gost</span><span class=p>(</span><span class=n>query</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>list</span><span class=p>(</span><span class=s>&quot;chromX&quot;</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;X:1000:1000000&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;rs17396340&quot;</span><span class=p>,</span>
<a id=__codelineno-43-2 name=__codelineno-43-2 href=#__codelineno-43-2></a><span class=w>                                                 </span><span class=s>&quot;GO:0005005&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;ENSG00000156103&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;NLRP1&quot;</span><span class=p>),</span>
<a id=__codelineno-43-3 name=__codelineno-43-3 href=#__codelineno-43-3></a><span class=w>                             </span><span class=s>&quot;chromY&quot;</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;Y:1:10000000&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;rs17396340&quot;</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-43-4 name=__codelineno-43-4 href=#__codelineno-43-4></a><span class=w>                                          </span><span class=s>&quot;GO:0005005&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;ENSG00000156103&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;NLRP1&quot;</span><span class=p>)),</span><span class=w> </span>
<a id=__codelineno-43-5 name=__codelineno-43-5 href=#__codelineno-43-5></a><span class=w>                       </span><span class=n>multi_query</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>)</span>
<a id=__codelineno-43-6 name=__codelineno-43-6 href=#__codelineno-43-6></a>
<a id=__codelineno-43-7 name=__codelineno-43-7 href=#__codelineno-43-7></a><span class=nf>head</span><span class=p>(</span><span class=n>multi_gostres2</span><span class=o>$</span><span class=n>result</span><span class=p>,</span><span class=w> </span><span class=m>3</span><span class=p>)</span>
</code></pre></div> <h2 id=visualization_1>Visualization<a class=headerlink href=#visualization_1 title="Permanent link">&para;</a></h2> <p>The enrichment results are visualized with a Manhattan-like-plot using the function <code>gostplot()</code> and the previously found gost results <em>gostres</em>:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-44-1 name=__codelineno-44-1 href=#__codelineno-44-1></a><span class=c1>#gostplot(gostres, capped = TRUE, interactive = TRUE)</span>
</code></pre></div> <p>The function <code>publish_gostplot()</code> takes the static plot object as an input and enables to highlight a selection of interesting terms from the results with numbers and table of results. These can be set with parameter highlight_terms listing the term IDs in a vector or as a data.frame with column "term_id" such as a subset of the result dataframe.</p> <p>First we create the static plot:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-45-1 name=__codelineno-45-1 href=#__codelineno-45-1></a><span class=n>p</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>gostplot</span><span class=p>(</span><span class=n>gostres</span><span class=p>,</span><span class=w> </span><span class=n>capped</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>,</span><span class=w> </span><span class=n>interactive</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span>
<a id=__codelineno-45-2 name=__codelineno-45-2 href=#__codelineno-45-2></a><span class=n>p</span>
</code></pre></div> <p>Then we make it in high quality. We can add highlighted terms if we want with the <em>highlight_terms</em> argument.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-46-1 name=__codelineno-46-1 href=#__codelineno-46-1></a><span class=n>pp</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>publish_gostplot</span><span class=p>(</span><span class=n>p</span><span class=p>,</span><span class=w> </span><span class=n>highlight_terms</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;CORUM:3047&quot;</span><span class=p>),</span>
<a id=__codelineno-46-2 name=__codelineno-46-2 href=#__codelineno-46-2></a><span class=w>                       </span><span class=n>width</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>NA</span><span class=p>,</span><span class=w> </span><span class=n>height</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>NA</span><span class=p>,</span><span class=w> </span><span class=n>filename</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>NULL</span><span class=w> </span><span class=p>)</span>
</code></pre></div> <p>The gost results can also be visualized with a table. The <code>publish_gosttable()</code> function will create a nice-looking table with the result statistics for the highlight_terms from the result data.frame. The highlight_terms can be a vector of term IDs or a subset of the results.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-47-1 name=__codelineno-47-1 href=#__codelineno-47-1></a><span class=nf>publish_gosttable</span><span class=p>(</span><span class=n>gostres</span><span class=p>,</span><span class=w> </span><span class=n>highlight_terms</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>gostres</span><span class=o>$</span><span class=n>result</span><span class=p>[</span><span class=nf>c</span><span class=p>(</span><span class=m>1</span><span class=o>:</span><span class=m>10</span><span class=p>),],</span>
<a id=__codelineno-47-2 name=__codelineno-47-2 href=#__codelineno-47-2></a><span class=w>                        </span><span class=n>use_colors</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>TRUE</span><span class=p>,</span><span class=w> </span>
<a id=__codelineno-47-3 name=__codelineno-47-3 href=#__codelineno-47-3></a><span class=w>                        </span><span class=n>show_columns</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;source&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;term_name&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;term_size&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;intersection_size&quot;</span><span class=p>),</span>
<a id=__codelineno-47-4 name=__codelineno-47-4 href=#__codelineno-47-4></a><span class=w>                        </span><span class=n>filename</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>NULL</span><span class=p>)</span><span class=w> </span>
</code></pre></div> <h1 id=pseudotime>Pseudotime<a class=headerlink href=#pseudotime title="Permanent link">&para;</a></h1> <p>We will use the package <a href=https://bioconductor.org/packages/release/bioc/vignettes/slingshot/inst/doc/vignette.html>Slignshot</a>. Slingshot was designed to model developmental trajectories in single-cell RNA sequencing data and serve as a component in an analysis pipeline after dimensionality reduction and clustering.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-48-1 name=__codelineno-48-1 href=#__codelineno-48-1></a><span class=n>adata_SCE</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>as.SingleCellExperiment</span><span class=p>(</span><span class=n>adata</span><span class=p>)</span>
<a id=__codelineno-48-2 name=__codelineno-48-2 href=#__codelineno-48-2></a><span class=n>sde</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>slingshot</span><span class=p>(</span><span class=n>adata_SCE</span><span class=p>,</span><span class=w> </span><span class=n>reducedDim</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&#39;PCA&#39;</span><span class=p>)</span>
</code></pre></div> <p>The using the <code>slingshot</code> package it is possible to plot our components and the calculated trajectory. Unfortunately, it does not work with ggplot.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-49-1 name=__codelineno-49-1 href=#__codelineno-49-1></a><span class=n>cols</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>colorRampPalette</span><span class=p>(</span><span class=n>viridis</span><span class=o>::</span><span class=nf>viridis</span><span class=p>(</span><span class=m>3</span><span class=p>))</span>
<a id=__codelineno-49-2 name=__codelineno-49-2 href=#__codelineno-49-2></a><span class=n>cols</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>cols</span><span class=p>(</span><span class=m>50</span><span class=p>)[</span><span class=nf>cut</span><span class=p>(</span><span class=n>sde</span><span class=o>$</span><span class=n>slingPseudotime_1</span><span class=p>,</span><span class=w> </span><span class=n>breaks</span><span class=o>=</span><span class=m>50</span><span class=p>)]</span>
<a id=__codelineno-49-3 name=__codelineno-49-3 href=#__codelineno-49-3></a><span class=nf>plot</span><span class=p>(</span><span class=nf>reducedDim</span><span class=p>(</span><span class=n>adata_SCE</span><span class=p>,</span><span class=w> </span><span class=s>&quot;PCA&quot;</span><span class=p>),</span><span class=w> </span><span class=n>pch</span><span class=o>=</span><span class=m>16</span><span class=p>,</span><span class=w> </span><span class=n>col</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>cols</span><span class=p>)</span><span class=w> </span><span class=c1># adata_SCE contains, PC1 and PC2, and colored by pseudotime</span>
<a id=__codelineno-49-4 name=__codelineno-49-4 href=#__codelineno-49-4></a><span class=nf>lines</span><span class=p>(</span><span class=n>slingshot</span><span class=o>::</span><span class=nf>SlingshotDataSet</span><span class=p>(</span><span class=n>sde</span><span class=p>),</span><span class=w> </span><span class=n>lwd</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>2</span><span class=p>)</span><span class=w> </span><span class=c1># here we plot the trajectory</span>
</code></pre></div> <p>It is possible to create a <code>FeaturePlot()</code> with both pseudotime and the trajectory by extracting the information from our <code>slingshot</code> object and add it to our <code>Seurat</code> object, but it is a bit inconvenient. First, pseudotime can be extracted with the function <code>slingPseudotime()</code>. The trajectory curves are stored inside the slingshot object <code>sde</code>; in this case, it is easier since there is only one trajectory.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-50-1 name=__codelineno-50-1 href=#__codelineno-50-1></a><span class=n>adata</span><span class=o>$</span><span class=n>pseudotime</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>slingPseudotime</span><span class=p>(</span><span class=n>sde</span><span class=p>)</span>
<a id=__codelineno-50-2 name=__codelineno-50-2 href=#__codelineno-50-2></a><span class=n>trajectory</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>data.frame</span><span class=p>(</span><span class=nf>SlingshotDataSet</span><span class=p>(</span><span class=n>sde</span><span class=p>)</span><span class=o>@</span><span class=n>curves</span><span class=p>[[</span><span class=s>&quot;Lineage1&quot;</span><span class=p>]][[</span><span class=s>&quot;s&quot;</span><span class=p>]])</span>
<a id=__codelineno-50-3 name=__codelineno-50-3 href=#__codelineno-50-3></a><span class=nf>head</span><span class=p>(</span><span class=n>trajectory</span><span class=p>)</span>
</code></pre></div> <p>Now we can use Seurat functions to plot the pseudotime and the calculated trajectory!</p> <div class=highlight><pre><span></span><code><a id=__codelineno-51-1 name=__codelineno-51-1 href=#__codelineno-51-1></a><span class=nf>FeaturePlot</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>features</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;pseudotime&quot;</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;pca&quot;</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>scale_color_viridis_c</span><span class=p>()</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>geom_path</span><span class=p>(</span><span class=n>data</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>trajectory</span><span class=p>,</span><span class=w> </span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>PC_1</span><span class=p>,</span><span class=w> </span><span class=n>y</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>PC_2</span><span class=p>))</span>
</code></pre></div> <div class=highlight><pre><span></span><code><a id=__codelineno-52-1 name=__codelineno-52-1 href=#__codelineno-52-1></a><span class=nf>DimPlot</span><span class=p>(</span><span class=n>adata</span><span class=p>,</span><span class=w> </span><span class=n>group.by</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;Cell_type&quot;</span><span class=p>,</span><span class=w> </span><span class=n>reduction</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;pca&quot;</span><span class=p>)</span>
</code></pre></div> <p>We can create a plot that sorts our cells by pseudotime and check that it corresponds to their cell type:</p> <div class=highlight><pre><span></span><code><a id=__codelineno-53-1 name=__codelineno-53-1 href=#__codelineno-53-1></a><span class=n>adata_SCE</span><span class=o>$</span><span class=n>pseudotime</span><span class=w> </span><span class=o>&lt;-</span><span class=w> </span><span class=nf>as.numeric</span><span class=p>(</span><span class=nf>slingPseudotime</span><span class=p>(</span><span class=n>sde</span><span class=p>))</span>
<a id=__codelineno-53-2 name=__codelineno-53-2 href=#__codelineno-53-2></a><span class=nf>ggplot</span><span class=p>(</span><span class=nf>as.data.frame</span><span class=p>(</span><span class=nf>colData</span><span class=p>(</span><span class=n>adata_SCE</span><span class=p>)),</span><span class=w> </span><span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>pseudotime</span><span class=p>,</span>
<a id=__codelineno-53-3 name=__codelineno-53-3 href=#__codelineno-53-3></a><span class=w>                                             </span><span class=n>y</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>Cell_type</span><span class=p>,</span>
<a id=__codelineno-53-4 name=__codelineno-53-4 href=#__codelineno-53-4></a><span class=w>                                             </span><span class=n>colour</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>Cell_type</span><span class=p>))</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-53-5 name=__codelineno-53-5 href=#__codelineno-53-5></a><span class=w>  </span><span class=n>ggbeeswarm</span><span class=o>::</span><span class=nf>geom_quasirandom</span><span class=p>(</span><span class=n>groupOnX</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=kc>FALSE</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-53-6 name=__codelineno-53-6 href=#__codelineno-53-6></a><span class=w>  </span><span class=nf>theme_classic</span><span class=p>()</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-53-7 name=__codelineno-53-7 href=#__codelineno-53-7></a><span class=w>  </span><span class=nf>xlab</span><span class=p>(</span><span class=s>&quot;Slingshot pseudotime&quot;</span><span class=p>)</span><span class=w> </span><span class=o>+</span><span class=w> </span><span class=nf>ylab</span><span class=p>(</span><span class=s>&quot;Timepoint&quot;</span><span class=p>)</span><span class=w> </span><span class=o>+</span>
<a id=__codelineno-53-8 name=__codelineno-53-8 href=#__codelineno-53-8></a><span class=w>  </span><span class=nf>ggtitle</span><span class=p>(</span><span class=s>&quot;Cells ordered by Slingshot pseudotime&quot;</span><span class=p>)</span>
</code></pre></div> <h1 id=session-info>Session info<a class=headerlink href=#session-info title="Permanent link">&para;</a></h1> <p>Finally, we create a <code>session_info()</code> table that will allow anyone to check what versions of R and packages are we using for reproducibility purposes.</p> <div class=highlight><pre><span></span><code><a id=__codelineno-54-1 name=__codelineno-54-1 href=#__codelineno-54-1></a><span class=n>devtools</span><span class=o>::</span><span class=nf>session_info</span><span class=p>()</span>
</code></pre></div> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/hds-sandbox/ target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://twitter.com/ucph_heads target=_blank rel=noopener title=twitter.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 512 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../../..", "features": ["content.code.annotate", "content.tooltips", "navigation.indexes", "navigation.sections", "navigation.footer", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow"], "search": "../../../assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script> <script src=../../../assets/javascripts/bundle.5a2dcb6a.min.js></script> <script src=../../../javascripts/mathjax.js></script> <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script src=https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js></script> <script src=../../../javascripts/embed-amd.js></script> </body> </html>